---
title: "DESeq2 Analysis of Placenta Sorted Cell Types"
subtitle: "Kyle Campbell - March 14th, 2019"
output:
  html_document:
   self_contained: true
---

```{r setup, include=FALSE}
# CRAN Packages
library(ashr)
library(edgeR)
library(ggplot2)
library(ggrepel)
library(gprofiler2)
library(ggpubr)
library(here)
library(knitr)
library(pheatmap)
library(RColorBrewer)
library(tidyverse)

# Bioconductor Packages
library(BiocManager)
library(biomaRt)
library(DESeq2)
library(GO.db)
library(mygene)
library(org.Hs.eg.db)

knitr::opts_chunk$set(echo=TRUE)
```

Note: we deduced we had two sample swaps within biological replicates (Sample_103903 -> Sample_103906 & Sample_103911 -> Sample_103914) based on marker gene expression and estimated cell type proportions based on the scRNA-seq dataset reference

## Entering Data, Preproccessing, and Model Fitting

```{r DESeq2_object}
counts <- readRDS(file =  here("data", "bulk", "DESeq2_collapsed_count_data.Rda"))
meta <- readRDS(file = here("data", "bulk", "DESeq2_collapsed_sample_metadata.Rda"))

# Must make meta rownames identical to counts colnames
rownames(meta) <- substr(rownames(meta), 8, 14)

# Creating the DESeq2 experiment object from count data and meta info
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = meta,
                              design = ~ 0 + sample.ID + cell.type)
```

Droppping .xx from ENSG IDs for symbol annotation: https://www.biostars.org/p/293965/
Explanation of ENSG vs. ENST: https://www.biostars.org/p/199073/

```{r AddGeneAnnotation, eval = F}
# Chop off the .xx from the ENSEMBL ID
row.names(dds) <- substr(x = row.names(dds), start = 1, stop = 15)
```

BiomaRt returns 58,069 of 58,381 gene features from DESeq2 results. Of the 58,069 results, 56,521 return unique external gene names and 58,024 return unique ENSEMBL gene IDs.

```{r biomaRt_DESeq2_results, eval = F}
# Setting up biomaRt
# Assigning Mart of interest, human ensembl here
mart = useMart("ensembl",dataset="hsapiens_gene_ensembl")

# Querying biomaRt using gene names to get back ENSEMBL IDs
DESeq2GeneIDs <- getBM(attributes = c('ensembl_gene_id', 'external_gene_name'),
             filters = c('ensembl_gene_id'),
             values = rownames(dds),
             mart = mart)
genes.tbl <- as_tibble(DESeq2GeneIDs)

#saveRDS(genes.tbl, here("data", "bulk", "deseq2_gene_symbols.rda"))
```

```{r load_biomaRt_DESeq2_results}
# Load gene ids
genes.tbl <- readRDS(here("data", "bulk", "deseq2_gene_symbols.rda"))
```

Save/load 
```{r, eval = F}
#saveRDS(DESeq2GeneIDs_tbl, "biomaRt_DESeq2_ensembl_ids.rda")
#DESeq2GeneIDs_tbl <- as_tibble(readRDS("biomaRt_DESeq2_ensembl_ids.rda"))
```

```{r SaveLoad, eval = F}
# Saved the fitted model object to avoid having to reestimate parameters
#saveRDS(dds, file = "./data/analysis_datasets/DESeq2_dds_noint_celltype_sampleid_nofiltering.rda")

# Load the annotated DESeq2 object - note this object has trimmed ensembl IDs
dds <- readRDS(here("data", "bulk", "DESeq2_dds_noint_celltype_sampleid_nofiltering.rda"))
```

Ideas for how to deal with multiple Ensembl IDs mapping to a single gene symbol
https://www.researchgate.net/post/How-to-deal-with-multiple-ensemble-IDs-mapping-to-one-gene-symbol-in-a-RNA-Seq-dataset

## Comparing cell type-specific gene expression to the average expression level across other cell types

Our first contrast of interest is how cell type-specific expression differs vs. the average expression across all other cell types. [This thread](https://support.bioconductor.org/p/118090/) recommends fitting a no-intercept model to answer this question. [Another thread](https://support.bioconductor.org/p/86347/) contains a brief discussion of how to code it. These contrasts will exclude the composite tissue group. I wonder if excluding the composite samples from normalization and filtering is appropriate for this particular set of comparisions. I think that also depends on whether we use the composite samples in any analyses.

Updated code (3/2021):
1. Include composite samples in model for gene variance moderation or drop? Leave in
  a. "If I have multiple groups, should I run all together or split into pairs of groups?" DESeq2 vignette recommends to do so unless within-group variability varies differently than other groups, the high variance can disrupt the genewise dispersion estimates in the other groups.
2. Include endothelial in the cell type comparison group?
  a. Empirically test
    i. Stable model either way (fewer hits than CT)
3. Handling Ensembl ID/HGNC symbol multimapping? Multimapping a very minor issue (see below), so leave as is
  a. previously handled at annotating results stage (when DESeq2 vignette does it)
    i. Currently, independent filtering handles most multimapping reads due to low counts or low variance, requiring no filtering
    ii. Less than 100 multimapping genes per contrast remain even after independent filtering and are dropped

Note: in DESeq2 alpha is the FDR-adjusted significance cutoff

4. Bimodal histograms are a little weird down below. Several threads (https://support.bioconductor.org/p/102313/, https://www.biostars.org/p/331711/, https://www.biostars.org/p/312792/), including Michael Love suggest pre-filtering as low counts may be causing the peaks on the p-value histogram (recall that a well-calibrated statistical test has a uniform p-value distribution with a spike at low p-values if there's a significant difference). 

Solution: Pre-filtering: normalized counts greater than 10 per gene and detected in at least three samples, as recommended in the DESeq2 paper to address bimodal p-value distribution.

```{r, eval = F}
binary <- as.matrix((counts(dds) > 0) + 0)
# Exclude 19048 transcripts that were not detected in at least 3 samples - 39,333 remaining
keep.three <- rowSums(binary) >= 3
dds <- dds[keep.three]
```

```{r, eval = F}
keep.min <- rowSums(counts(dds, normalized = TRUE)) >= 10
# Exclude 865 transcripts that did not have at least 10 normalized counts across all samples, leaving 38,468 for analysis
dds <- dds[keep.min, ]
```

```{r, eval = F}
dds <- DESeq(dds)
```

Save a prefiltered version of the dds for other analyses.
```{r, eval = F}
#saveRDS(dds, here("data", "bulk", paste0(Sys.Date(), "_dds_prefiltered.rda")))
```

Load the prefiltered version.
```{r, eval = F}
dds <- readRDS(here("data", "bulk", "2021-04-16_dds_prefiltered.rda"))
```

Format and export metadata table
```{r, eval = F}
meta.tb <- dds@colData %>% as.data.frame %>% as_tibble
levels(meta.tb$sample.ID) <- c("Sample 6", "Sample 7", "Sample 1", "Sample 8")
meta.tb$sample.ID <- relevel(meta.tb$sample.ID, "Sample 1")

meta.tb$title <- paste0(meta.tb$sample.ID, " - Bulk RNA-seq - ", str_to_title(meta.tb$cell.type))

#write.csv(meta.tb, here("data", "bulk", "bulk_RNA_seq_meta.csv"), row.names = FALSE)
```

Custom function to plot gene counts by sample and cell type
```{r, eval = F}
gene.lookup <- rowData(dds) %>% as.data.frame %>% dplyr::select(ensgene, symbol)
plot_gene_counts<- function(gene) {
  
  gene.id <- gene.lookup %>%
  filter(symbol == gene) %>%
  pull(ensgene)
  
  d <- plotCounts(dds, gene=gene.id, intgroup = "cell.type", returnData = TRUE)
  ggplot(d, aes(x = cell.type, y = count, color = cell.type)) + 
    geom_point(position=position_jitter(w = 0.1,h = 0)) +
    geom_text_repel(aes(label = rownames(d))) + 
    theme_bw() +
    ggtitle(gene) +
    theme(plot.title = element_text(hjust = 0.5))
}
```

```{r, eval = F}
genes <- c("PTPRC", "COL1A1", "HLA-G", "KDR", "KRT7", "CSH1")
lapply(genes, plot_gene_counts)
```

```{r, eval = F}
immune.markers <- c("CD163", "LYVE1", "CD14", "CD8A", "CD3D", "NKG7", "CD79A")
lapply(immune.markers, plot_gene_counts)
```

We deduced we had two sample swaps within biological replicates (Sample_103903 -> Sample_103906 & Sample_103911 -> Sample_103914) based on marker gene expression and estimated cell type proportions based on the scRNA-seq dataset reference
```{r, eval = F}
dds@colData[dds@colData$run.sample.ID == "Sample_103906", 1] <- "leukocyte"
dds@colData[dds@colData$run.sample.ID == "Sample_103903", 1] <- "fibroblast"

dds@colData[dds@colData$run.sample.ID == "Sample_103914", 1] <- "leukocyte"
dds@colData[dds@colData$run.sample.ID == "Sample_103911", 1] <- "fibroblast"
```

Format and export updated metadata table, including renaming samples
```{r, eval = F}
meta.tb <- dds@colData %>% as.data.frame %>% as_tibble
levels(meta.tb$sample.ID) <- c("Sorted 2", "Sorted 3", "Sorted 1", "Sorted 4")
meta.tb$sample.ID <- relevel(meta.tb$sample.ID, "Sorted 1")

meta.tb$title <- paste0(meta.tb$sample.ID, " - Bulk RNA-seq - ", str_to_title(meta.tb$cell.type))

dds@colData$sample.ID <- meta.tb$sample.ID

#write.csv(meta.tb, here("data", "bulk", "bulk_RNA_seq_meta.csv"), row.names = FALSE)
```

Re-run DESeq model with updated dds object
```{r, eval = F}
dds <- DESeq(dds)
```

```{r}
#saveRDS(dds, here("data", "analytic", "sorted", paste0(Sys.Date(), "DEseq2_object.rda")))
dds <- readRDS(here("data", "analytic", "sorted", "2022-05-14DEseq2_object.rda"))
```

## Data Quality Assessment by Clustering and Visualization

We'll begin by transforming the data to the log-fold-2 scale using the same shrinkage transformation employed in the DEX pipeline
```{r transformation}
#rld <- rlog(dds, blind = TRUE)
#saveRDS(rld, here("data", "bulk", paste0(Sys.Date(), "_DESeq2_rld.rda")))
#rld <- readRDS(here("data", "bulk", "2021-03-26DESeq2_rld.rda"))
rld <- readRDS(here("data", "bulk", "2022-05-14_DESeq2_rld.rda"))
```

103911 - leukocyte apperas to be an outlier on blind PCA analysis, will remove and rerun. PCA looks similar with 103911 - do not remove; after sample swap, no longer an outlier
```{r}
#dds_103911 <- dds[,colnames(dds) != "103911"]
#dds_103911 <- DESeq(dds_103911)
#rld <-rlog(dds_103911, blind = TRUE)
```

### Hierarchical Clustering Heatmap

```{r hc_heatmap}
# Calculate sample distances
sampleDists <- dist(t(assay(rld)))

# Hierarchical heatmap clustering - DESeq2 vignette
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$cell.type, rld$sample.ID, rld$is.Paired, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

### PCA Plot

```{r rld_pca}
# Customized PCA ggplot
pcaData <- plotPCA(rld, intgroup=c("cell.type", "sample.ID"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

# Modify labels for prettier graphing
cell.type.labels <- c("Whole Tissue", "Cytotrophoblast", "Endothelial", "Extravillous Trophoblast", "Fibroblast", "Hofbauer Cell", "Leukocyte", "Syncytiotrophoblast")
levels(pcaData$cell.type) <- cell.type.labels

# 6 corresponds to KC36, 7 to KC39, 1 to KC40 (the single-cell matched sample), and 8 to KC48
#sample.labels <- c(6, 7, 1, 8)
#levels(pcaData$sample.ID) <- sample.labels

ggplot(pcaData, aes(PC1, PC2, color=cell.type, shape=sample.ID)) +
  geom_point(size=3.5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  labs(color = "Cell Type", shape = "Sample")
  coord_fixed()
#ggsave(sprintf("%sSorted_PCA.png", local_output_path), device = "png")
#ggsave(here("results", "bulk", paste0(Sys.Date(), "_analytic_sorted_PCA.png")))
```

### Model

``` {r}
cell.types <-
    list(
      "cytotrophoblast",
      "endothelial",
      "extravilloustrophoblast",
      "fibroblast",
      "hofbauer",
      "leukocyte",
      "syncytiotrophoblast"
    )

run_deseq2 <- function(target, dds) {
  
  # Print current target
  print(target)
  
  # Set alpha
  alpha <- .05
  
  # Set logFC cutoff
  logFC.cutoff <- 1.2
  
  # List coefficients
    cell.types <-
    c(
      "cell.typecytotrophoblast",
      "cell.typeendothelial",
      "cell.typeextravilloustrophoblast",
      "cell.typefibroblast",
      "cell.typehofbauer",
      "cell.typeleukocyte",
      "cell.typesyncytiotrophoblast"
    )
  
  # Select contrast coefficients based on cell type input
  coef <- cell.types[grepl(target, cell.types)]
  inv.coef <- cell.types[!grepl(target, cell.types)]
  
  # For the DESeq2 results() call in this chunk:
  # The first vector is our cell type of interest
  # Create the contrast of interest
  # The following vector contains comparison group
  # listValues Specifies comparision against average expression across other groups
  # Pull contrast of interest
  results <- results(
    dds,
    contrast = list(coef,
                    inv.coef),
    listValues = c(1,-1 / 6),
    alpha = alpha
  )
  
  # Apply logFoldChange shrinkage
  resLFC <- lfcShrink(dds, contrast = list(coef,
                  inv.coef), type="ashr")
  
  # Summarize DESeq2 filtering
  summary <- summary(results)
  print(
    paste0(
      summary(is.na(results$padj))[3],
      " features dropped due to independent filtering, all zero counts, or extreme outlier status in DESeq2"
    )
  )
  
  # Coerce DESeq2 Results to df
  res.df <- as.data.frame(results)
  
  # Move gene ensembl IDs from row to table column
  res <- rownames_to_column(res.df, "ensembl_gene_id")
  
  # Remove dropped genes from results
  res <-
    res %>%
    filter(!is.na(padj))
  
  # Add gene annotation and drop duplicate gene names
  res <- left_join(res, genes.tbl) %>%
    dplyr::select(external_gene_name, everything()) %>%
    relocate(ensembl_gene_id, .after = last_col())
  
  # Print the small number of duplicated genes after DESeq2 auto filtering. Most are SNORNAs and many are vague Y_RNA, rRNA, metazoa RNA, etc.
  duplicated.genes <-
    res %>%
    group_by(external_gene_name) %>%
    tally() %>%
    filter(n > 1)
  print(paste0(dim(duplicated.genes)[1], " duplicated gene symbols dropped"))
  
  # Drop duplicated genes after DESeq2 auto filtering
  res <-
    res %>%
    group_by(external_gene_name) %>%
    filter(!(n() > 1)) %>%
    as_tibble()
  
  # Rename gene symbol column from "external_gene_name" to "gene_symbol"
  res <-
    res %>%
    mutate(gene_symbol = external_gene_name) %>%
    dplyr::select(-external_gene_name)
  
  res <-
    res %>%
    dplyr::select(gene_symbol,
                  baseMean,
                  log2FoldChange,
                  lfcSE,
                  stat,
                  pvalue,
                  padj,
                  ensembl_gene_id)
  
  # Get and arrange upreg
  upreg <-
    res %>%
    filter(padj < alpha) %>%
    filter(log2FoldChange > logFC.cutoff) %>%
    arrange(desc(log2FoldChange))
  
  # Get and arrange downreg
  downreg <-
    res %>%
    filter(padj < alpha) %>%
    filter(log2FoldChange < -logFC.cutoff) %>%
    arrange(desc(abs(log2FoldChange)))
  
  # Number of top downregulated and upregulated genes to label on Volcano plot
  num.to.plot <- 10
  
  # Option to manually label a vector of genes
  genes.of.interest <- NULL
  
  # Create label dummy to decide if gene labelled on Volcano plot
  top.upreg <- upreg$gene_symbol[1:num.to.plot]
  top.downreg <- downreg$gene_symbol[1:num.to.plot]
  genes.to.plot <- c(top.upreg, top.downreg, genes.of.interest)
  res$label[(res$gene_symbol %in% genes.to.plot)] <- T
  
  # Add significance indicator variable for plotting
  res <-
    res %>%
    mutate(sig = ifelse((padj < alpha) & (abs(log2FoldChange) > logFC.cutoff),
                        T,
                        F))
  # Label factor levels for legend purposes
  #res$sig <- factor(res$sig,
  #                  levels = c(FALSE, TRUE),
  #                  labels = c("FDR-adjusted p-value > 0.05", "FDR-adjusted p-value < 0.05"))
  
  res$sig <- factor(res$sig,
                    levels = c(FALSE, TRUE),
                    labels = c("Does not meet p-value and logFC criteria", paste0("FDR-adjusted p-value < ", alpha, " & log2FC > ", logFC.cutoff)))

  
  # Volcano plot
  plot <- ggplot(res) +
    geom_point(aes(
      x = log2FoldChange,
      y = -log10(padj),
      colour = sig
    )) +
    geom_text_repel(aes(
      x = log2FoldChange,
      y = -log10(padj),
      label = ifelse(label, gene_symbol, "")
    ),
    max.overlaps = Inf,
    size = 6) +
    geom_hline(yintercept = -log10(alpha), linetype = "dotted") +
    geom_vline(xintercept = logFC.cutoff, linetype = "dotted") +
    geom_vline(xintercept = -logFC.cutoff, linetype = "dotted") +
    ggtitle("Contrast of Interest") +
    xlab("log2 fold change") +
    ylab("-log10 adjusted p-value") +
    theme(
      legend.position = "bottom",
      legend.title = element_blank(),
      plot.title = element_text(size = rel(1.5), hjust = 0.5),
      axis.title = element_text(size = rel(1.25))
    )# +
    #labs(caption = paste0(
    #  dim(res)[1],
    #  " total genes (",
    #  summary(is.na(results$padj))[3],
    #  " dropped from contrast)")
    #  )
  
  # Create a label dataframe for geom_label annotation
  label.df <- data.frame(
    label = paste0(dim(upreg)[1], " upregulated genes\n", paste0(dim(downreg)[1], " downregulated genes")),
    x = Inf,
    y = Inf,
    hjustvar = 1,
    vjustvar = 1
  )
  
  # Add geom_label annotation
  plot <-
    plot + geom_label(data = label.df,
                      aes(
                        x = x,
                        y = y,
                        label = label,
                        hjust = hjustvar,
                        vjust = vjustvar
                      ),
                      size = 6,
                      label.padding = unit(0.5, "lines")
                      )
  plot

  returnList <- list("res" = res,
                     "volcano" = plot,
                     "duplicated.genes" = duplicated.genes,
                     "results" = results,
                     "shrunk" = resLFC)
  return(returnList)
}
```

``` {r}
mapped <- lapply(cell.types, run_deseq2, dds = dds)
names(mapped) <- cell.types
```

MA Plots, p-adj < .1 is highlighted by default
```{r}
cell.types
mas <- map(mapped, ~ DESeq2::plotMA((.x[["results"]])))
```

ashr-shrunk MA Plots. Not yet sure what the purpose of shrinkage is beyond "visualization and gene ranking" as explained in the DESeq2 vignette
```{r}
cell.types
mas <- map(mapped, ~ DESeq2::plotMA((.x[["shrunk"]])))
```

Making presentable Volcano plots
```{r}
# Define a base_size variable for the font size for the publication multipanel
base.size <- 24

# Input ggplot to get x and y limits for graph
get_coord_limits <- function(ggplot) {
  
  # Get build options
  build <- ggplot_build(ggplot)
  # Get x-limits
  x <- build$layout$panel_scales_x[[1]]$range$range
  # Get y-limits
  y <- build$layout$panel_scales_y[[1]]$range$range
  
  # Return x and y limits as list
  return(list(
    "x" = x,
    "y" = y
  ))
}

# Pull limits from each graph
limits <- map(mapped, ~ get_coord_limits(.x[["volcano"]]))

# Pull x-limits and convert to list of vectors
x_limits <- map(limits, ~ unlist(.x["x"]))
# Convert x-limits to vector
x_limits <- unlist(x_limits)
# Get x-min
x_min <- min(x_limits)
# Get x-max
x_max <- max(x_limits)

# Pull y-limits and convert to list of vectors
y_limits <- map(limits, ~ unlist(.x["y"]))
# Convert y-limits to vector
y_limits <- unlist(y_limits)
# Get y-min
y_min <- min(y_limits)
# Get y-max
# y-max for ST is 1e-60, going to use 2nd largest value instead for graph readibility
y_sorted <- sort(y_limits)
y_max <- y_sorted[length(y_sorted)-3] %>% as.numeric
#y_max <- max(y_limits)

names(mapped)

contrast.labels <- c("Cytotrophoblast" , "Endothelial", "Extravillous Trophoblast", "Fibroblast", "Hofbauer Cell", "Leukocyte", "Syncytiotrophoblast")

# Get presentable volcanos
# Add contrast title and set x and y limits without removing data points
volcanos.ind <- map2(mapped, contrast.labels,
                 ~ .x[["volcano"]] + ggtitle(.y) + coord_cartesian(xlim = c(x_min, x_max), ylim = c(y_min, y_max)))


# Update geom_points, which are in 5/14 ratio to text font size
#geom.size <- 5
#update_geom_defaults("text_repel", list(size = geom.size))
#update_geom_defaults("label", list(size = geom.size))

volcanos <- map2(mapped, contrast.labels,
                 ~ .x[["volcano"]] + ggtitle(.y) + coord_cartesian(xlim = c(x_min, x_max), ylim = c(y_min, y_max)) + theme_bw(base_size = base.size) + theme(legend.position = "none") + ylab("-log10 q-value"))

big.volcano.ct <- volcanos$cytotrophoblast + theme(legend.position = "right")

# Custom function to save ggplots
save_volcano_plot <- function(plot, filename) {
  print(plot)
  ggsave(
    here("results", "bulk", "volcanos", paste0(filename, "_volcano_", Sys.Date(), ".png"))
  )
}
# Save volcanos
#map2(volcanos.ind, names(mapped), ~ save_volcano_plot(.x, .y))
```

```{r}
#map(volcanos.ind, print)
```

```{r}
# Pull legend from a Volcano and make sure the title is removed
volcano.legend <- get_legend(big.volcano.ct + theme(legend.title = element_blank()) + guides(colour = guide_legend(override.aes = list(size=15))))

#volcano.legend <- get_legend(big.volcano.ct + theme(legend.title = element_blank(), legend.key.width = unit(4, "cm"), legend.key.height = unit(4, "cm")), position = "right")
```

```{r}
panel <- ggarrange(volcanos[[1]], volcanos[[2]], volcanos[[3]], volcanos[[4]], volcanos[[5]], volcanos[[6]], volcanos[[7]], volcano.legend, ncol = 4, nrow = 2, labels = "AUTO", font.label = list(size = 36, face = "bold", color = "black"))
#print(panel)
# Works okay with theme_bw(base_size = 20) and font.label size = 24 in ggarrange
#ggexport(panel, filename = here("results", "bulk", "volcanos", paste0(Sys.Date(), "_volcano_multipanel.png")), width = 1920, height = 1080)
```

```{r}
# Pull dex results
results <- map(mapped, ~ .x[["res"]])
```

```{r}
#View(results)
```

```{r}
logFC.cutoff <- 1.2
alpha <- 0.05

all <- plyr::ldply(results, rbind) %>%
  group_by(.id) %>%
  arrange(desc(stat)) %>% # Sort by value of test statistic, fine since we want to see the overexpressed hits
  dplyr::select(.id, gene_symbol, log2FoldChange, baseMean, pvalue, padj) %>% # Pick which columns to keep and in which order
  # Filter by logFC and pvalue
  filter(log2FoldChange > logFC.cutoff) %>%
  filter(padj < alpha) %>%
  mutate(log2FoldChange = round(log2FoldChange, digits = 2)) %>%
  mutate(baseMean = round(baseMean, digits = 2)) %>%
  mutate(pvalue = signif(pvalue, digits = 2)) %>%
  mutate(padj = signif(padj, digits = 2))
all$.id <- factor(all$.id)
levels(all$.id) <- contrast.labels
colnames(all) <- c("Cell Type", "Gene", "log2 Fold-change", "Base Mean", "P-value", "False-discovery rate-adjusted Q-value")

all.split <- all %>% group_split()
names(all.split) <- contrast.labels

# Write to multiple sheets
#write.xlsx(as.list(all.split), file = here("results", "bulk", paste0(Sys.Date(), "_all_sorted_dex_res.xlsx")))
```

Create a top10 table for brief supplemental table.
```{r, eval = F, include = F}
all <- plyr::ldply(results, rbind) %>%
  group_by(.id) %>%
  arrange(desc(stat)) %>% # Sort by value of test statistic, fine since we want to see the overexpressed hits
  dplyr::select(.id, gene_symbol, log2FoldChange, baseMean, pvalue, padj) %>% # Pick which columns to keep and in which order
  mutate(log2FoldChange = round(log2FoldChange, digits = 2)) %>%
  mutate(baseMean = round(baseMean, digits = 2)) %>%
  mutate(pvalue = signif(pvalue, digits = 2)) %>%
  mutate(padj = signif(padj, digits = 2))
all$.id <- factor(all$.id)
levels(all$.id) <- contrast.labels
colnames(all) <- c("Cell Type", "Gene", "log2 Fold-change", "Base Mean", "P-value", "False-discovery rate-adjusted Q-value")

all.grouped <- all %>% group_by(`Cell Type`) %>% arrange(`Cell Type`, `log2 Fold-change`)

all.grouped %>% filter(`False-discovery rate-adjusted Q-value` < 0.05) %>% tally()
library(psych)
all.grouped %>% filter(`False-discovery rate-adjusted Q-value` < 0.05) %>% filter(`log2 Fold-change` > 0) %>% tally() %>% describe()

all.top10 <- all %>%
  slice_head(n = 10)

# Save
#write_csv(all.grouped, here("results", "bulk", paste0(Sys.Date(), "_deseq2_all_dex.csv")))
#write_csv(all.top10, here("results", "bulk", paste0(Sys.Date(), "_deseq2_dex_top10.csv")))
```

PrerankGSEA does not resolve ties. Code to make GSEA input and diagnostic plots to ensure that there are not excessive ties. GSEA should always include all genes tested in differential expression. Perhaps subsetting to the most enriched pathways for interpretation. In RNA-seq, GSEA may suffer from very lowly expressed genes that may be artifactual and expressed in any samples.
```{r}
make_gsea_input <- function(res, title) {

  return <- res %>%
    arrange(desc(stat)) %>%
    dplyr::select(gene_symbol, stat)
  
  # Pull the stat associated with the largest p-value less than .05 for boundaries
  stat.lim <- res %>%
    filter(padj < .05) %>%
    arrange(desc(padj)) %>%
    slice_head(n = 1) %>%
    pull(stat)
  
  # Pull the associated padj
  padj.lim <- res %>%
    filter(padj < .05) %>%
    arrange(desc(padj)) %>%
    slice_head(n = 1) %>%
    pull(padj)
  
  # Get number of stat duplicates
  dup.stats.counts <- res %>%
  group_by(stat) %>%
  filter(n() > 1) %>%
  tally()
  
  # Graph stat duplicates
  print(ggplot(dup.stats.counts, aes(x = stat)) + geom_histogram() + ggtitle(paste0("Duplicate Test Statistics - ", title)) +
    geom_vline(xintercept = abs(stat.lim)) + geom_vline(xintercept = -abs(stat.lim)) +
    geom_label(
      data = NULL,
      x = abs(stat.lim),
      y = 50,
      label = paste0("padj = ", round(padj.lim, 2))
    ))
  
  # Save plots
  # Unsure why this code is producing an error about not being able to open .png device
  # ggsave(
  #   here("results", "GSEA", " sorted", "dup_stats_plots", paste0(title, "_dup_gsea_input_stats_", Sys.Date(), ".png"))
  # )
  
  # Write  .rnk file for GSEA input
  write_tsv(return, here("results", "GSEA", "sorted", paste0(title, "_all_cell_types_prefilter_gsea_input_stats_", Sys.Date(), ".rnk")), col_names = FALSE)
  
  return(return)
}

inputs <- map2(results, contrast.labels, ~ make_gsea_input(res = .x, title = .y))
```

```{r}
map2(results, contrast.labels, ~ ggplot(data = .x, mapping = aes(x = pvalue)) + geom_histogram() + ggtitle(.y))
```

```{r}
gost_deseq2 <- function(res, alpha) {
  
  # Filter results and rank by test statistic
  res <-
    res %>%
    filter(padj < alpha) %>%
    arrange(desc(stat)) %>%
    filter(log2FoldChange > 0)
  gost.input <- res$gene_symbol
  
  # gost from ggprofiler2 with ordered query and manually curated annotation background
  gostres <- gost(query = gost.input,
               organism = "hsapiens", ordered_query = TRUE,
               multi_query = FALSE, significant = FALSE, exclude_iea = TRUE,
               measure_underrepresentation = FALSE, evcodes = FALSE,
               user_threshold = alpha, correction_method = "g_SCS",
               domain_scope = "annotated", custom_bg = NULL,
               numeric_ns = "", sources = NULL, as_short_link = FALSE)
  # graph
  gostplot <- gostplot(gostres, capped = FALSE, interactive = FALSE)
  print(gostplot)
  
  results <- list(
    "gostres" = gostres,
    "gostplot" = gostplot
  )
  
  return(results)
}

gost.mapped <- map(results, ~ gost_deseq2(.x, alpha = .05))
```

Processing GSEA results
```{r}
# Find indices for the folders to keep
keep <- grep(pattern = "GseaPreranked", x = list.files(here("results", "GSEA", "sorted")))
folders <- list.files(here("results", "GSEA", "sorted"))[keep]

# Get path for folders that contain GSEA results
paths <- paste0(here("results", "GSEA", "sorted"), "/", folders)

# Custom function that takes a folder path and retrieves the postive .tsv GSEA results 
retrieve_gsea_tsv <- function(path) {
  
  name <- str_replace(path, pattern = "G:/My Drive/Placenta_Cell_Types/RNA/placenta_cell_types_rna/results/GSEA/sorted/", "")
  
  # List all files in the path
  files <- list.files(path)
  # Find indices of file list that correspond to the grep
  files.keep <- grep(pattern = "gsea_report_for_na_pos.*tsv", x = files)
  
  # Return the indexed files that are being searched for
  return <- paste0(path, "/", files[files.keep])
  #return <- files[files.keep]
  return(return)
}

res.gsea.files <- unlist(map(paths, ~ retrieve_gsea_tsv(.x)))

gsea <- map(res.gsea.files, ~ read_tsv(.x), col_names = TRUE)
names(gsea) <- contrast.labels

gsea.all <- plyr::ldply(gsea, rbind)
```

Formatting for publication table
```{r}
gsea.table <- gsea.all %>%
  dplyr::select(.id, NAME, SIZE, NES, 'NOM p-val', 'FDR q-val')

colnames(gsea.table) <- c("Cell Type", "Name", "Size", "nes", "pval", "qval")

# A couple dozen of the least significant terms in the Hofbauer contrast return NA values from GSEA and are present here as NA's; does not meaningfully affect results
gsea.table <- gsea.table %>%
  mutate(nes = as.numeric(nes)) %>%
  mutate(pval = as.numeric(pval)) %>%
  mutate(qval = as.numeric(qval)) %>%
  filter(qval < 0.05) %>%
  mutate(nes =  round(nes, digits = 3)) %>%
  mutate(pval = round(pval, digits = 3)) %>%
  mutate(qval = round(qval, digits = 3)) %>%
  mutate(pval = ifelse(pval == 0, "<.001", pval)) %>%
  mutate(qval = ifelse(qval == 0, "<.001", pval))

colnames(gsea.table) <- c("Cell Type", "Name", "Size", "Normalized Enrichment Score", "P-value", "False Discovery Rate-Controlled q-value")


cell.type.ontologies <- 
  gsea.table %>%
  group_by(`Cell Type`) %>%
  group_split()
names(cell.type.ontologies) <- contrast.labels


#write.xlsx(as.list(cell.type.ontologies), file = here("results", "bulk", paste0(Sys.Date(), "_all_sorted_gsea_res.xlsx")))

#gsea.table.top10 <- gsea.table %>%
#  group_by(`Cell Type`) %>%
#  slice_head(n = 10)

#write_csv(gsea.table, here("results", "GSEA", "sorted", paste0(Sys.Date(), "_gsea_hits.csv")))
#write_csv(gsea.table.top10, here("results", "GSEA", "sorted", paste0(Sys.Date(), "_gsea_hits_top10.csv")))
```

# Export for CIBERSORTx mixture deconvolution
```{r}
# CPM-normalize DESeq2 count matrix
cpm <- fpm((dds))

cpm.tbl <- 
  cpm %>%
  as.data.frame() %>%
  rownames_to_column("ensembl_gene_id") %>%
  left_join(genes.tbl)

# Dropped genes due to duplication or missing annotation
duplicated.genes <- 
  cpm.tbl %>%
  group_by(external_gene_name) %>%
  filter(n()>1) %>%
  dplyr::select(external_gene_name, everything())

# Reformat table
cpm.tbl.unique <-
  cpm.tbl %>%
  group_by(external_gene_name) %>%
  filter(n() == 1) %>%
  ungroup() %>%
  dplyr::select(-ensembl_gene_id) %>%
  dplyr::select(external_gene_name, everything())

# Create sample names
labels <- paste0(dds@colData$sample.ID, "_", str_to_title(dds@colData$cell.type))
labels <- str_replace(labels, " ", "_")

# Annotate sample names
colnames(cpm.tbl.unique)[1] <- "GeneSymbol"
colnames(cpm.tbl.unique)[2:length(colnames(cpm.tbl.unique))] <- labels

# Check for missingness
is.na(cpm.tbl.unique) %>% summary

# Write
#write_tsv(cpm.tbl.unique, here("data", "bulk", paste0(Sys.Date(), "_bulk_mixture_CIBERSORTx_input_fpm.txt")))
```

