---
title: "GSE75010_cibersort_bulk_prep"
author: "Kyle Campbell"
date: "11/2/2020"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(ggridges)

library(biomaRt)
library(DESeq2)

knitr::opts_chunk$set(echo = TRUE)

data_dir <- "C:/Users/k_cam/Downloads/scRNA_hpc_qc/"
```
Read in prepared GSE75010 from GSE75010_cibersort_sc_prep.
```{r}
GSE75010 <- read_tsv(paste0(data_dir, "GSE75010_input_mixture_2020-10-13.txt"))
```
Read in DESeq2 object that contains sorted cell type bulk data.
```{r}
dds <- readRDS("C:/Users/k_cam/Google Drive/Placenta_Cell_Types_Kyle/code/R code/local/placenta_sort/dex/data/analysis_datasets/DESeq2_dds_noint_celltype_sampleid_nofiltering.rda")
```
## Prepare sorted cell type RNA-seq for signature matrix input
Call biomaRt to get gene annotation information for the rld genes.
```{r call_biomaRt}
# Setting up biomaRt
# Assigning Mart of interest, human ensembl here
mart = useMart("ensembl",dataset="hsapiens_gene_ensembl")

# Querying biomaRt using gene names to get back ENSEMBL IDs
DESeq2GeneIDs <- getBM(attributes = c('ensembl_gene_id', 'external_gene_name'),
             filters = c('ensembl_gene_id'),
             values = rownames(dds),
             mart = mart)
DESeq2GeneIDs_tbl <- as_tibble(DESeq2GeneIDs)
# Clean-up environment
rm(DESeq2GeneIDs)
```
Summarize the results of the biomaRt call.
```{r summarize_biomaRt}
nrow.mart <- as.numeric(nrow(DESeq2GeneIDs_tbl))
nrow.rld <- as.numeric(nrow(dds))
n.unique.ids <- DESeq2GeneIDs_tbl %>%
  distinct(ensembl_gene_id) %>%
  tally()
n.unique.genes <- DESeq2GeneIDs_tbl %>%
  distinct(external_gene_name) %>%
  tally()
print(paste0("biomaRt returns ", nrow.mart, " of ", nrow.rld, " gene features from DESeq2 results. Of the ", nrow.mart, " results, ", n.unique.genes, " return unique external gene names and ", n.unique.ids, " return unique ENSEMBL gene IDs."))
# Clean-up environment
rm(nrow.mart, nrow.rld, n.unique.ids, n.unique.genes)
```
Summarize gene overlap between datasets.
```{r}
contains <- summary(GSE75010$GeneSymbol %in% DESeq2GeneIDs_tbl$external_gene_name)
nrow.microarray <- as.numeric(nrow(GSE75010))
print(paste0(as.numeric(contains[["TRUE"]]), " of ", nrow.microarray, " genes in the GSE75010 dataset contained in the bulk dataset."))
```
"Standard RNA-Seq expression quantification metrics, such as fragments per kilobase per million (FPKM) and transcripts per kilobase million (TPM), are suitable for use with CIBERSORT." Remove the composite samples and the lone epithelial sample from the dataset and drop that level.
```{r}
# Subset to non-composite samples
dds.sub <- dds[, dds$cell.type != "composite"]
dds.sub <- dds[, dds$cell.type != "endothelial"]
# Drop composite from cell.type factor levels
dds.sub$cell.type <- droplevels(dds.sub$cell.type)
# Recalculate size factors
dds.sub <- estimateSizeFactors(dds.sub)
# Get size-normalized dds counts
df <- as.data.frame(counts(dds.sub, normalized=TRUE))
# Convert to tibble for merging and further processing
tb <- as_tibble(rownames_to_column(df, var = colnames(DESeq2GeneIDs_tbl)[1]))
```
Merge normalized bulk data with gene symbol annotation.
```{r}
join <- left_join(tb, DESeq2GeneIDs_tbl, by = "ensembl_gene_id") %>%
  dplyr::select(external_gene_name, ensembl_gene_id, everything()) %>%
  dplyr::select(-ensembl_gene_id)
rm(tb)
```
Drop non-unique genes
```{r}
dropped <- join %>%
  group_by(external_gene_name) %>%
  filter((n() > 1)) %>%
  tally()
print(paste0("Dropping ", length(dropped$external_gene_name), " duplicated gene symbols:"))
print(dropped$external_gene_name)
join <- join %>%
  group_by(external_gene_name) %>%
  filter(!(n() > 1))
```
Keep only genes that are in both datasets.
```{r}
contained <- join$external_gene_name %in% GSE75010$GeneSymbol
join <- join[contained,]
```
Data reformatting for CibersortX.
```{r}
row1 <- c("genes", as.character(dds.sub$cell.type))
colnames(join) <- row1
tb <- tibble(join, .name_repair = "unique")
# Reorganize columns by cell type name alphabetically
# Not sure why this breaks here when running the whole file but works if you enter commands stepwise
tb <- tb %>%
  dplyr::select(genes, starts_with("cyto"), starts_with("endo"), starts_with("extra"), starts_with("fibro"), starts_with("hof"), starts_with("leuk"), starts_with("syn"))
# https://stackoverflow.com/questions/44021352/regex-for-three-dots requires double escape
colnames(tb) <- gsub(replacement = "", pattern = "\\.{3}.*", colnames(tb))
row.one <- colnames(tb)
colnames(tb) <- NULL
tb <- rbind(row.one, tb)
```
Convert reference input matrix to long form to check count distribution is negative binomial.
```{r}
tb.long <- tb
colnames(tb.long) <- tb[1, ]
tb.long <- tb.long [-1,]
tb.long <- tb.long %>%
  pivot_longer(!genes, names_to = "cell.type")
tb.long$value <- as.numeric(tb.long$value)
tb.long$cell.type <- as.factor(tb.long$cell.type)
```
It does appear to be negative binomially distributed. 
```{r}
a <- ggplot(tb.long, aes(value))
a + geom_histogram()
```
Appears similar across cell types.
```{r}
b <- ggplot(tb.long, aes(x = value, y = cell.type, group = cell.type))
b + geom_density_ridges() + coord_cartesian(xlim = c(0, 4000))
```
Write table.
```{r}
write.table(tb, file = paste0(data_dir, "kc_bulk_signature_matrix_input_GSE75010_", Sys.Date(), ".txt"), sep = "\t", quote = F, row.names = F, col.names = F)
```

```{r}
test <- read_tsv(paste0(data_dir, "kc_bulk_signature_matrix_classes_input_2020-11-03.txt"), col_names = FALSE)

#write.table(test, file = paste0(data_dir, "kc_bulk_signature_matrix_classes_input_GSE75010_", Sys.Date(), ".txt"), sep = "\t", quote = F, row.names = F, col.names = F)
```