---
title: "Prep for xMWAS"
subtitle: "Kyle Campbell - November 1st, 2020"
output: html_document
---

```{r setup, include=FALSE}
# Update R all installed R Packages
# update.packages(ask = FALSE)

library(knitr)
library(pheatmap)
library(RColorBrewer)
library(Seurat)
library(tidyverse)
library(devtools)
library(WGCNA)
library(snow)
library(igraph)
library(plyr)
library(plsgenomics)
library(shiny)
library(shinyBS)
library(visNetwork)

library(BiocManager)
library(biomaRt)
library(DESeq2)
library(EnhancedVolcano)
library(GO.db)
library(mygene)
library(org.Hs.eg.db)
library(topGO)
library(graph)
library(RBGL)
library(impute)
library(mixOmics) 
library(preprocessCore)

# Install xMWAS 
#install_github("kuppal2/xMWAS")
library(xMWAS)

knitr::opts_chunk$set(echo=TRUE)

#data_dir <- "C:/Users/k_cam/Google Drive/Placenta_Cell_Types/RNA/placenta_rna/data"
# Temp directory
data_dir <- "C:/Users/k_cam/Downloads/scRNA_hpc_qc/"
```
# xMWAS sorted cell type data preparation
This document is meant to prepare the placenta RNA project data for input into xMWAS. Installation instructions came from https://github.com/kuppal2/xMWAS/blob/master/xMWAS_installation_and_tutorial.pptx. File format guides for xMWAS are located at https://kuppal.shinyapps.io/xmwas/ under 'Help and Support'.
## Prepare the DESeq2 rlog-normalized data for xMWAS Dataset File Format
Read in the regularized log data from the DESeq2 analysis.
```{r saveload}
rld <- readRDS(file = "C:/Users/k_cam/Google Drive/Placenta_Cell_Types_Kyle/code/R code/local/placenta_sort/dex/data/analysis_datasets/DESeq2_rld.rda")
```
Call biomaRt to get gene annotation information for the rld genes.
```{r call_biomaRt}
# Setting up biomaRt
# Assigning Mart of interest, human ensembl here
mart = useMart("ensembl",dataset="hsapiens_gene_ensembl")

# Querying biomaRt using gene names to get back ENSEMBL IDs
DESeq2GeneIDs <- getBM(attributes = c('ensembl_gene_id', 'external_gene_name'),
             filters = c('ensembl_gene_id'),
             values = rownames(rld),
             mart = mart)
DESeq2GeneIDs_tbl <- as_tibble(DESeq2GeneIDs)
# Clean-up environment
rm(DESeq2GeneIDs)
```
Summarize the results of the biomaRt call.
```{r summarize_biomaRt}
nrow.mart <- as.numeric(nrow(DESeq2GeneIDs_tbl))
nrow.rld <- as.numeric(nrow(rld))
n.unique.ids <- DESeq2GeneIDs_tbl %>%
  distinct(ensembl_gene_id) %>%
  tally()
n.unique.genes <- DESeq2GeneIDs_tbl %>%
  distinct(external_gene_name) %>%
  tally()
print(paste0("biomaRt returns ", nrow.mart, " of ", nrow.rld, " gene features from DESeq2 results. Of the ", nrow.mart, " results, ", n.unique.genes, " return unique external gene names and ", n.unique.ids, " return unique ENSEMBL gene IDs."))
# Clean-up environment
rm(nrow.mart, nrow.rld, n.unique.ids, n.unique.genes)
```
Prepare rld data for merge.
```{r}
# Extract the count matrix from the DESeqTransform object rld using assay()
df <- as.data.frame(assay(rld))
# Pull the ensembl ids to a column and name it according to biomaRt return
tb <- as_tibble(rownames_to_column(df, var = colnames(DESeq2GeneIDs_tbl)[1]))
```
Merge rld data with gene symbol annotation.
```{r}
join <- left_join(tb, DESeq2GeneIDs_tbl, by = "ensembl_gene_id") %>%
  dplyr::select(external_gene_name, ensembl_gene_id, everything()) %>%
  dplyr::select(-ensembl_gene_id)
```
Mutate gene symbol names with '_mrna' to label assay type in xMWAS.
```{r}
xMWAS.mrna <- join %>%
  mutate(mrna_id = paste0(external_gene_name, "_mrna")) %>%
  dplyr::select(-external_gene_name) %>%
  dplyr::select(mrna_id, everything())
```
## Prepare the mRNA Class Label File Format (multiclass) for xMWAS input
Get class labels xMWAS input.
```{r}
xMWAS.mrna.labels <- tibble(
  FileName = colnames(rld),
  Class = rld$cell.type
)
```
# xMWAS single-cell data prep
## Prepare integrated and normalized Seurat scRNA-seq dataset (Campbell and Pique-Regi)
Read in Seurat-processed data and subset to the confidently labelled cell types.
```{r}
# Read in data
seu <- readRDS(paste0(data_dir,"tnl_combined_mnn_labelled.rda"))
# Subset to the confidently labelled cells
seu <- subset(x = seu, idents = c("Cytotrophoblasts", "Fibroblasts", "npiCytotrophoblasts", "Hofbauer Cells", "Extravillous Trophoblasts", "Synctyiotrophoblast", "Naive T Cells", "CD14+ Monocytes", "B Cells", "Activated T Cells", "Natural Killer", "pDC", "Endothelial Cells", "FCGR3A+ Monocytes"))
```
Extract the normalized count matrix (normalized using Seurat's default workflow).
```{r}
tb <- as_tibble(seu@assays$RNA@data)
xMWAS.scmrna <- tb %>%
  mutate(mrna_id = paste0(rownames(seu@assays$RNA@data), "_scmrna")) %>%
  dplyr::select(mrna_id, everything())
```
## Prepare the scRNA-seq Class Label File Format (multiclass) for xMWAS input
Get scRNA-seq class labels.
```{r}
xMWAS.scmrna.labels <- tibble(
  FileName = colnames(seu@assays$RNA@data),
  Class = Idents(seu)
)
```
## Merge class label files
Merge the class label files for each dataset.
```{r}
xMWAS.labels <- bind_rows(xMWAS.mrna.labels, xMWAS.scmrna.labels)
```
# Run xMWAS
```{r}
xmwas_res<-run_xmwas(
  Xome_data=xMWAS.mrna,
  Yome_data=xMWAS.scmrna,
  outloc=paste0(data_dir, "/xMWAS/"),
  classlabels=xMWAS.labels,
  xmwasmethod="pls",
  plsmode="regression",
  max_xvar=10000,
  max_yvar=10000,
  rsd.filt.thresh=1,corthresh=0.4,keepX=1000,keepY=1000,keepZ=1000,keepW=1000, pairedanalysis=FALSE,optselect=TRUE,rawPthresh=0.05,numcomps=10,net_edge_colors=c("blue","red"), net_node_colors=c("orange", "green","cyan","pink"),Xname="X",Yname="Y",Zname="Z",Wname="W", net_node_shape=c("square","circle","triangle","star"),all.missing.thresh=0,missing.val=0, seednum=100,label.cex=0.2,vertex.size=6,max_connections=NA, centrality_method="eigenvector",use.X.reference=FALSE,removeRda=TRUE,compare.classes=TRUE,class.comparison.allvar=TRUE)
suppressWarnings(try(sink(file=NULL),silent=TRUE))

```













