---
title: "Comparing placental bulk- and scRNA-seq"
output: html_document
---

```{r setup, include=FALSE}
library(here)
library(tidyverse)
library(corrplot)
library(formulaic)
library(ggplot2)
library(ggpubr)
library(ggridges)
library(ggrepel)
library(Seurat)
library(cowplot)
library(rgl)
library(EnvStats)
library(reticulate)
library(clustree)
library(BiocManager)
# Must install the following packages using BiocManager, e.g. > BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)
library(biomaRt)
library(DESeq2)
library(mygene)
knitr::opts_chunk$set(echo = TRUE)
```

Not really feasible to compare the two at all, https://www.biostars.org/p/399869/. Batch effect would be larger than biological effect and no way to control for that.

```{r ImportData}
seu <- readRDS(here("data", "analytic", "2022-05-01_all_merged_pruned_batch_mnn.rda"))
dds <- readRDS(here("data", "analytic", "sorted", "2022-05-14DEseq2_object.rda"))

# Import bulk data gene list already retrieved during the DESeq2 differential expression analysis; don't need
#dds.genes <- readRDS(here("data", "bulk", "deseq2_gene_symbols.rda"))
```

Normalized Seurat single-cell counts are stored in the "data" slot of the "RNA" field of the "assay" slot. Here we use biomaRt to collect ENSEMBL gene IDs using gene names stored in the Seurat object.

```{r biomaRtGeneIDs, eval = F}
# Setting up biomaRt
# Assigning Mart of interest, human ensembl here
mart <- useMart("ensembl",dataset="hsapiens_gene_ensembl")

# # Alternative because biomaRt connection unavailable above
mart <- useEnsembl(biomart = "ensembl", dataset = 'hsapiens_gene_ensembl', host = "uswest.ensembl.org")
 
# Querying biomaRt using gene names to get back ENSEMBL IDs
scGeneIDs <- getBM(attributes = c('ensembl_gene_id', 'external_gene_name'),
              filters = c('external_gene_name'),
              values = rownames(seu@assays$RNA@data),
              mart = mart)
genes.sc <- as_tibble(scGeneIDs)

#saveRDS(genes.sc, here("data", "bulk", "biomaRt_single-cell_2022-05-20.rda"))

# 36601 genes in single-cell data
length(rownames(seu@assays$RNA@data))
# biomaRt query returned 38886 gene IDs
dim(genes.sc)
```

Load ensembl gene annotation
```{r}
genes.sc <- readRDS(here("data", "bulk", "biomaRt_single-cell_2022-05-20.rda"))
```


No duplicate IDs
```{r}
genes.sc %>%
  group_by(ensembl_gene_id) %>%
  filter(n() > 1)
```

Check for duplicate gene names
```{r}
genes.sc %>%
  group_by(external_gene_name) %>%
  filter(n() > 1)
```

```{r}
# Function arguments
seuratObject <- seu
clusterIdentity <- "Fetal Hofbauer Cells"
biomaRtList <- genes.sc

# Helper code to write out the long strings required for RenameIdents() input
#base::paste(idents, collapse = " = Leukocytes, ")

seu <- SetIdent(seu, value = "cell.type")
levels(seu)

# Grouping Mesenchymal Stem Cells with Fibroblast since the Flow sorting for the bulk samples does not distinguish between the two based on single-cell expression data
seu <- RenameIdents(seu,
             "Fetal B Cells" = "Leukocytes",
             "Fetal CD14+ Monocytes" = "Leukocytes",
             "Fetal CD8+ Activated T Cells" = "Leukocytes",
             "Fetal Cytotrophoblasts" = "Cytotrophoblast",
             "Fetal Endothelial Cells" = "Endothelial",
             "Fetal Extravillous Trophoblasts" = "Extravillous Trophoblast",
             "Fetal Fibroblasts" = "Fibroblast",
             "Fetal GZMB+ Natural Killer" = "Leukocytes",
             "Fetal GZMK+ Natural Killer" = "Leukocytes",
             "Fetal Hofbauer Cells" = "Hofbauer Cell",
             "Fetal Memory CD4+ T Cells" = "Leukocytes",
             "Fetal Mesenchymal Stem Cells" = "Fibroblast",
             "Fetal Naive CD4+ T Cells" = "Leukocytes",
             "Fetal Naive CD8+ T Cells" = "Leukocytes",
             "Fetal Natural Killer T Cells" = "Leukocytes",
             "Fetal Nucleated Red Blood Cells" = "Fetal Nucleated Red Blood Cells",
             "Fetal Plasmacytoid Dendritic Cells" = "Leukocytes",
             "Fetal Proliferative Cytotrophoblasts" = "Cytotrophoblast",
             "Fetal Syncytiotrophoblast" = "Syncytiotrophoblast",
             "Maternal B Cells" = "Leukocytes",
             "Maternal CD14+ Monocytes" = "Leukocytes",
             "Maternal CD8+ Activated T Cells" = "Leukocytes",
             "Maternal FCGR3A+ Monocytes" = "Leukocytes",
             "Maternal Naive CD4+ T Cells" = "Leukocytes",
             "Maternal Naive CD8+ T Cells" = "Leukocytes",
             "Maternal Natural Killer Cells" = "Leukocytes",
             "Maternal Plasma Cells" = "Leukocytes")

# Stash collapse cell type groups under a variable called collapse 
seu <- StashIdent(seu, "collapse")

table <- table(Idents(seu), seu$ident)
table %>% sum

# Calculate AverageExpression by cell type group
# WARNING: by default, returns in non-log space
seu <- NormalizeData(seu)
gex.sc.average <- AverageExpression(seu, group.by = "collapse", slot = "data", return.seurat = T)

# Converting Seurat Normalized expression matrix to a dataframe for gene ID annotation
cluster.df <- GetAssayData(gex.sc.average, slot = "data") %>% as.data.frame

#cluster.df <- cluster.df %>%
#  mutate_all(log1p)

cluster.df <- rownames_to_column(cluster.df, var = "external_gene_name")
```

Double-check how featureCounts or DESeq2 handles gene IDs mapping to multiple haplotypes (could also just choose the standard haplotype but don't know of a way of doing without manually looking up IDs, perhaps someone has a list, going to ignore for now), noticing a lot of the same expression values with each multimapping gene symbol. Discussion of disambiguation https://www.researchgate.net/post/How-to-deal-with-multiple-ensemble-IDs-mapping-to-one-gene-symbol-in-a-RNA-Seq-dataset - taking the highest expression value seems the easiest. Another practical thread: https://www.biostars.org/p/389804/. Genes in haplotypic regions are the most likely reason for duplicate IDs mapping to one symbol https://www.biostars.org/p/119540/. 
```{r}
# Join sc expression matrix and gene IDs
gex.sc <-
  as_tibble(left_join(x = cluster.df, y = biomaRtList))

gex.sc %>%
  group_by(external_gene_name) %>%
  tally %>%
  filter (n > 1)

gex.sc %>%
  filter(is.na(external_gene_name))

dim(gex.sc)

# Take max expression value of the multiple ensembl gene IDs mapping to the same external gene name
gex.sc.unique <- 
  gex.sc %>%
  group_by(external_gene_name) %>%
  summarise(across(c(Leukocytes, Cytotrophoblast, Endothelial, `Extravillous Trophoblast`, Fibroblast, `Hofbauer Cell`, `Fetal Nucleated Red Blood Cells`, `Syncytiotrophoblast`),.fn = ~ max(.x))) # Operate colwise on all column names that contain "RNA" and take max of non-unique external_gene_names
dim(gex.sc.unique)

# Drop the Fetal Nucleated Red blood cells since we're not using that
gex.sc.unique <- gex.sc.unique %>%
  dplyr::select(!`Fetal Nucleated Red Blood Cells`)
```

Calculating normalized DESeq2 counts from the sorted data and subsetting to relevant cell type populations (https://support.bioconductor.org/p/66067/ for reference).

```{r BulkNormalizedCounts}
# Calculate normalized counts from bulk RNA-seq DESeq2 object (code from thread above)
bulk.normalized <- as.data.frame(counts(dds, normalized = T))
#bulk.normalized <- rownames_to_column(bulk.normalized, "ensembl_gene_id")
# Want to emulate normalization used in Seurat. Seurat performs library size normalization and then log1p transforms the result using the natural log.
bulk.norm.elog <- 
  bulk.normalized %>%
  mutate_all(log1p)
rownames(bulk.norm.elog) <- rownames(bulk.normalized)
bulk.normalized <- bulk.norm.elog
```

```{r AnnotateBulkNormalized, eval = F}
# Querying biomaRt using gene names to get back ENSEMBL IDs
bulkGeneIDs <- getBM(attributes = c('ensembl_gene_id', 'external_gene_name'),
              filters = c('ensembl_gene_id'),
              values = rownames(bulk.normalized),
              mart = mart)
genes.bulk <- as_tibble(bulkGeneIDs)
dim(genes.bulk)
#saveRDS(genes.bulk, here("data", "bulk", "biomaRt_bulk_2022-05-20.rda"))
```

Load bulk data annotation
```{r AnnotateBulkNormalized}
genes.bulk <- readRDS(here("data", "bulk", "biomaRt_bulk_2022-05-20.rda"))

bulk.gex <- rownames_to_column(bulk.normalized, var = "ensembl_gene_id")
bulk <- left_join(bulk.gex, genes.bulk, by = "ensembl_gene_id")
dim(bulk)

missing <- bulk %>%
  filter(is.na(external_gene_name)) %>%
  tally
print(paste0(missing$n, " missing external gene name annotations in bulk data dropped"))

bulk <- bulk %>%
  filter(!is.na(external_gene_name))
dim(bulk)
```

```{r}
# Take max expression value of the multiple ensembl gene IDs mapping to the same external gene name
bulk.unique <- 
  bulk %>%
  group_by(external_gene_name) %>%
  summarise(across(where(is.numeric), ~ max(.x))) # Operate colwise on all numeric cols and take max of non-unique
dim(bulk.unique)

# Return external_gene_name to rownames slot
bulk.unique <- column_to_rownames(bulk.unique, var = "external_gene_name")
```

```{r BulkSubset}
ct.bulk        <- bulk.unique[,c( "103905", "105816")]
composite.bulk <- bulk.unique[,c( "103900", "103908", "103916", "105819", "105820")]
# Need to add drop = FALSE to prevent conversion to vector
en.bulk        <- bulk.unique[,c( "103915"), drop = FALSE]
et.bulk        <- bulk.unique[,c( "103904", "103912", "105815")]
fb.bulk        <- bulk.unique[,c( "103903", "103911", "105817")]
hb.bulk        <- bulk.unique[,c( "103894", "103902", "103910", "105813")]
lk.bulk        <- bulk.unique[,c( "103895", "103906", "103914", "105814")]
st.bulk        <- bulk.unique[,c( "103893", "105812")]
```

```{r}
bulk <- 
  tibble(
  external_gene_name = rownames(bulk.unique),
  ct.bulk %>%
    rowwise() %>%
    summarize(`Sorted Cytotrophoblast` = mean(c_across(where(
      is.numeric
    )))),
  composite.bulk %>%
    rowwise() %>%
    summarize(`Whole Tissue` = mean(c_across(where(
      is.numeric
    )))),
  `Sorted Endothelial` = en.bulk$`103915`,
  et.bulk %>%
    rowwise() %>%
    summarize(`Sorted Extravillous Trophoblast` = mean(c_across(where(
      is.numeric
    )))),
  fb.bulk %>%
    rowwise() %>%
    summarize(`Sorted Fibroblast` = mean(c_across(where(
      is.numeric
    )))),
  hb.bulk %>%
    rowwise() %>%
    summarize(`Sorted Hofbauer Cell` = mean(c_across(where(
      is.numeric
    )))),
  lk.bulk %>%
    rowwise() %>%
    summarize(`Sorted Leukocytes` = mean(c_across(where(
      is.numeric
    )))),
  st.bulk %>%
    rowwise() %>%
    summarize(`Sorted Syncytiotrophoblast` = mean(c_across(where(
      is.numeric
    ))))
)
```

```{r}
gex.sc.unique %>% dim
bulk %>% dim
```

Join gene symbols that are in both datasets
```{r}
joint <- inner_join(gex.sc.unique, bulk, by = "external_gene_name")
dim(joint)
```

```{r}
genes.cibersort <- read_delim(file = here("results", "cibersortx_local", "analytic_gse75010", "kc_pr_tsang_sig_matrix.txt"))
cor.cibersort <- 
  joint %>% 
  filter(external_gene_name %in% genes.cibersort$NAME)
# 4596 overlapping and cibersortx marker genes
dim(cor.cibersort)[1]
```

```{r}
data.ciber <- (column_to_rownames(cor.cibersort, var = "external_gene_name"))
colnames(data.ciber)[1:7] <- paste0("Single-cell ", colnames(data.ciber[1:7]))
# Compute a correlation matrix
corr.ciber <- round(cor(data.ciber), 3)
corr.spearman.ciber <- round(cor(data.ciber, method = "spearman"), 3)
```

  
```{r}
data.cor <- (column_to_rownames(joint, var = "external_gene_name"))
# Compute a correlation matrix
corr <- round(cor(data.cor), 3)
corr.spearman <- round(cor(data.cor, method = "spearman"), 3)
```

```{r}
#dev.new()
#pdf(paste0(results_dir, "cell_type_props_corrplot_", Sys.Date(), ".pdf"))
corrplot(corr, method = "ellipse", order = "hclust", tl.cex = .7) #tl.cex control axis label size
corrplot(corr.spearman, method = "ellipse", order = "hclust", tl.cex = .7) #tl.cex control axis label size
corrplot(corr.spearman, method = "ellipse", order = "hclust", tl.cex = .7) #tl.cex control axis label size
#dev.off()

# Alternative options
#corrplot(corr, method = "ellipse", order = "hclust")
#corrplot(corr, method = "ellipse", order = "AOE")
```



```{r}
exprCorrelationPlot <- function(df, string) {
  
  # Store the string for a ggplot title
  title <- string
  
  # Take only the first word for matching purposes
  string <- word(string)
  
  # Subset the dataset to the two relevant columns
  pair <- df %>% dplyr::select(contains(string))
  
  # Calculate Pearson and Spearman correlation coefficients
  pearson <- cor(pair[,1], pair[,2], method = c("pearson")) %>% as.numeric %>% round(digits = 2)
  spearman <- cor(pair[,1], pair[,2], method = c("spearman")) %>% as.numeric %>% round(digits = 2)
  
  #print(paste0(pearson, " pearson correlation"))
  #print(paste0(spearman, " spearman correlation"))
  
  # Find the colnames that match the input
  match.indices <- grep(pattern = string, x = names(df), ignore.case = TRUE)
  var <- names(df)[match.indices]
  
  # Plot single-cell and bulk expression
  plot <- 
    ggplot(df) +
    aes(x = !!sym(var[1]), y = !!sym(var[2])) +
    geom_point(alpha = .3) +
    # Drop Pearson correlation
    labs(subtitle = paste0("\U03C1 = ", spearman)) + #, ", r = ", pearson)) +
    xlab("Single-cell Expression") + ylab("Sorted Expression") + ggtitle(title) + theme_bw(base_size = 20)
  
  # Save plots
  #ggsave(here("results", "bulk", "correlation", paste0(Sys.Date(), string, "_correlation.png")), width = 7, height = 7)
  
  print(plot)
  
  return(plot)
}
```

Which genes are all zero across all single-cell groups
```{r}
gene.sums <- data.ciber %>%
  rowwise() %>%
  mutate(gene.sum = sum(c(`Single-cell Cytotrophoblast`, `Single-cell Endothelial`, `Single-cell Extravillous Trophoblast`, `Single-cell Fibroblast`, `Single-cell Hofbauer Cell`, `Single-cell Leukocytes`, `Single-cell Syncytiotrophoblast`)))
summary(gene.sums)
quantile(gene.sums$gene.sum, probs = c(.01, .05, .1, .25))

ggplot(gene.sums, aes(x = gene.sum)) + geom_density() +  theme_bw()
```

Filter the lowest 10% expressed genes
```{r}
#data.ciber <- gene.sums %>% filter(gene.sum > quantile(gene.sums$gene.sum, probs = .1)) %>% dplyr::select(!gene.sum)
```

```{r}
contrast.labels <- c("Cytotrophoblast", "Endothelial", "Extravillous Trophoblast", "Fibroblast", "Hofbauer Cell", "Leukocyte", "Syncytiotrophoblast")

plots <- map(contrast.labels, ~ exprCorrelationPlot(data.ciber, .x))
```

Optimal file export settings are much different from display in R
```{r}
#panel <- ggarrange(plots[[1]], plots[[2]], plots[[3]], plots[[4]], plots[[5]], plots[[6]], plots[[7]], ncol = 4, nrow = 2, labels = "AUTO", font.label = list(size = 24, face = "bold", color = "black"))
#print(panel)
# Works okay with theme_bw(base_size = 20) and font.label size = 24 in ggarrange
#ggexport(panel, filename = here("results", "bulk", "correlation", paste0(Sys.Date(), "_correlation_multipanel.png")), width = 1920, height = 1080)
```


```{r}
exprCorrelation <- function(df, string) {
  
  # Store the string for a ggplot title
  title <- string
  
  # Take only the first word for matching purposes
  string <- word(string)
  
  # Subset the dataset to the two relevant columns
  pair <- df %>% dplyr::select(contains(string)) %>% as.matrix()
  
  # Calculate Pearson and Spearman correlation coefficients
  pearson <- cor(pair[,1], pair[,2], method = c("pearson")) %>% as.numeric %>% round(digits = 2)
  spearman <- cor(pair[,1], pair[,2], method = c("spearman")) %>% as.numeric %>% round(digits = 2)
  
  pearson.test <- cor.test(pair[,1], pair[,2], method = "pearson")
  p.pearson <- pearson.test$p.value %>% as.numeric %>% round(digits = 2)
  spearman.test <- cor.test(pair[,1], pair[,2], method = "pearson")
  p.spearman <- spearman.test$p.value %>% as.numeric %>% round(digits = 2)
  
  # Find the colnames that match the input
  match.indices <- grep(pattern = string, x = names(df), ignore.case = TRUE)
  var <- names(df)[match.indices]
  
  res <- data.frame(
    r = pearson,
    r.pvalue = p.pearson,
    rho = spearman,
    rho.pvalue = p.spearman
  )
  
  return(res)
}

contrast.labels <- c("Cytotrophoblast" , "Endothelial", "Extravillous Trophoblast", "Fibroblast", "Hofbauer Cell", "Leukocyte", "Syncytiotrophoblast")

#correlations <- map(contrast.labels, ~ exprCorrelation(joint, .x))
correlations <- map(contrast.labels, ~ exprCorrelation(data.ciber, .x))
names(correlations) <- contrast.labels

pearsons <- map(correlations, ~ .x$r) %>% unlist()
mean(pearsons)
sd(pearsons)

spearmans <- map(correlations, ~ .x$rho) %>% unlist()
mean(spearmans)
sd(spearmans)

# All correlation tests return p = 0 (report as p < .001 )
```

## Export for figures

```{r}
#dev.new()
#pdf(here("results", "bulk_sc_expression_correlation", paste0("bulk_sc_all_pearson_", Sys.Date(), ".pdf")))
corrplot.mixed(corr, upper = "ellipse", lower = "number", order = "hclust", tl.cex = .7, number.cex = 0.5, tl.pos = "lt")
#dev.off()
```

```{r}
#dev.new()
#pdf(here("results", "bulk_sc_expression_correlation", paste0("bulk_sc_all_spearman_", Sys.Date(), ".pdf")))
corrplot.mixed(corr.spearman, upper = "ellipse", lower = "number", order = "hclust", tl.cex = .7, number.cex = 0.5, tl.pos = "lt")
#dev.off()
```

```{r}
#dev.new()
#pdf(here("results", "bulk_sc_expression_correlation", paste0("bulk_sc_cibersort_pearson_", Sys.Date(), ".pdf")))
corrplot.mixed(corr.ciber, upper = "ellipse", lower = "number", order = "hclust", tl.cex = .7, number.cex = 0.5, tl.pos = "lt")
#dev.off()
```


```{r}
#dev.new()
#pdf(here("results", "bulk_sc_expression_correlation", paste0("bulk_sc_cibersort_spearman_", Sys.Date(), ".pdf")))
corrplot.mixed(corr.spearman.ciber, upper = "ellipse", lower = "number", order = "hclust", tl.cex = .7, number.cex = 0.5, tl.pos = "lt")
#corrplot.mixed(corr.spearman.ciber, p.mat = testRes$p, upper = "ellipse", lower = "number", order = "hclust", tl.cex = .7, number.cex = 0.5, tl.pos = "lt",  sig.level = c(0.001, 0.01, 0.05), pch.cex = 0.5, pch.col = 'grey20', insig = 'label_sig')
#dev.off()
```