---
title: "Preeclampsia Deconvolution Data"
author: "Kyle Campbell"
date: "10/12/2020"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(GEOquery)
library(Seurat)
knitr::opts_chunk$set(echo = TRUE)

data_dir <- "C:/Users/k_cam/Downloads/scRNA_hpc_qc/"
```

## Helpful pages
The CIBERSORTX article uses CPM for 3' Chromium 10X and MAS5 or RMA normalization with custom CDFv21 for microarrays. Plan to use CPM for scRNA-seq data.

This CIBERSORT article explains that microarray signature matrix can be used to deconvolve RNA-seq samples but recommends interpreting p-values with caution, but does not provide normalization intstructions.

"Importantly, all expression data should be non-negative, devoid of missing values, and represented in non-log linear space. For Affymetrix microarrays, a custom chip definition file (CDF) is recommended (see section 3.2.2) and should be normalized with MAS5 or RMA. Illumina Beadchip and single color Agilent arrays should be processed as described in the limma package. Standard RNA-Seq expression quantification metrics, such as fragments per kilobase per million (FPKM) and transcripts per kilobase million (TPM), are suitable for use with CIBERSORT.

The platform and methods used to generate data for the signature matrix ideally should be identical to that applied to analysis of the mixture samples. While SVR is robust to unknown cell populations, performance can be adversely affected by genes that are highly expressed in a relevant unknown cell population (e.g., in the malignant cells) but not by any immune components present in the signature matrix. A simple option implemented in CIBERSORT to limit this effect is to remove genes highly expressed in non-hematopoietic cells or tumor cells. If expression data is available from purified tumor cells for the malignancy to be studied, this can be used as a guideline to filter other confounding genes from the signature matrix."
https://www-ncbi-nlm-nih-gov.proxy.lib.umich.edu/pmc/articles/PMC5895181/

Microarray expression units:
https://www.researchgate.net/post/What_exactly_is_the_unit_of_measure_in_microarray_experiments
https://www.biostars.org/p/214013/
http://www.ub.edu/stat/docencia/bioinformatica/microarrays/ADM/slides/2_PreprocessingMicroarrayData-2-Preprocessing%20and%20Normalization.pdf

Should use raw files processed together for integrating datasets:
https://www.biostars.org/p/309365/

GEOQuery vignette:
https://www.bioconductor.org/packages/release/bioc/vignettes/GEOquery/inst/doc/GEOquery.html

## Preparing GSE75010 for CibersortX input
Pull GEO accession as GSE Matrix file
```{r}
gse <- getGEO("GSE75010", GSEMatrix = TRUE)
```

```{r explore}
# 157 samples that were generated using this paper
dim(gse$GSE75010_series_matrix.txt.gz@assayData$exprs)

# data preprocessing information stored in the gse data
gse$GSE75010_series_matrix.txt.gz@phenoData@data$data_processing[1]
gse$GSE75010_series_matrix.txt.gz@phenoData@data$data_processing.1[1]
```

Supplementary table "Complete_Dataset" from GSE75010 is a simple gene-by-sample matrix of processed, normalized, and log2 values from all 330 samples. Verfied given that the max expression value is 14.42505 in the matrix. CIBERSORTX will automatically antilog the values because the largest expression value is less than 50 units. Will use for mixture file in CIBERSORTX, S-mode deconvolution using our combined single-cell data to generate the signature matrix (since it's Chromium platform)

```{r}
expr <- read.csv(paste0(data_dir, "GSE75010_complete_dataset.csv.gz"))
expr.df <- as.data.frame(expr)
dim(expr.df)
expr.num <- expr.df[,-1]
max(expr.num)
```
Reformating the data for intput into CIBERSORTX
```{r}
empty_row <- character(331)
expr.df <- rbind(empty_row, expr.df)
expr.df[1,] <- colnames(expr.df)
expr.df[1,1] <- "GeneSymbol"
colnames(expr.df) <- NULL

write.table(expr.df, file = paste0(data_dir, "GSE75010_input_mixture_", Sys.Date(), ".txt"), sep = "\t", quote = F, row.names = F, col.names = F)
```

## Preparing Single-cell Pique-Regi/KC combined for CibersortX input

```{r}
# Read in data
seu <- readRDS(paste0(data_dir,"tnl_combined_mnn_labelled.rda"))
# Subset to the confidently labelled cells
seu <- subset(x = seu, idents = c("Cytotrophoblasts", "Fibroblasts", "npiCytotrophoblasts", "Hofbauer Cells", "Extravillous Trophoblasts", "Synctyiotrophoblast", "Naive T Cells", "CD14+ Monocytes", "B Cells", "Activated T Cells", "Natural Killer", "pDC", "Endothelial Cells", "FCGR3A+ Monocytes"))
```

```{r}
# Extract counts
counts <- as.matrix(seu@assays$RNA@counts)
# Extract per-cell cell type identities
types <- as.character(Idents(seu))
# Unload Seurat for extra memory
rm(seu)

# Add empty row and populate with per-cell cell type identities
empty_row <- character(dim(counts)[2])
counts <- rbind(empty_row, counts)
counts[1,] <- types

# Add empty column and populate with genes
empty_column <- character(dim(counts)[1])
counts <- cbind(empty_column, counts)
genes <- rownames(counts)
counts[,1] <- genes

# Reformat column 1
counts[1,1] <- "GeneSymbol"
colnames(counts) <- NULL
rownames(counts) <- NULL
counts <- as.data.frame(counts)
names(counts) <- NULL

write.table(x=counts, sep = '\t', quote = F, row.names = F, col.names = F, file = paste0(data_dir, "KC_Pique-Regi_sc_signature_matrix_input_", Sys.Date(), ".txt"))
```