---
title: "Limma Analysis of Cox Preeclampsia GSE75010 Study"
subtitle: "Kyle Campbell - January 2nd, 2021"
output: html_document
---

```{r setup, include=FALSE}
# Update R all installed R Packages
# update.packages(ask = FALSE)
library(corrplot)
library(factoextra)
library(here)
library(ggridges)
library(ggrepel)
library(grid)
library(gridExtra)
library(gprofiler2)   # For gene-set enrichment tests
library(knitr)
library(pheatmap)
library(RColorBrewer)
library(tidymodels)
library(tidyverse)

library(BiocManager)
library(Biobase)
library(limma)

# library(biomaRt)
# library(DESeq2)
# library(edgeR)
# library(EnhancedVolcano)
# library(GO.db)
# library(GEOquery)
# library(mygene)
# library(org.Hs.eg.db)
# library(topGO)

knitr::opts_chunk$set(echo=TRUE)

data_dir <- paste0(here("data"), "/")
results_dir <- paste0(here("results", "GSE75010"), "/")
```
Based on Bioconductor workflow http://www.bioconductor.org/packages/release/workflows/vignettes/maEndToEnd/inst/doc/MA-Workflow.html#14_a_pathway_enrichment_analysis_using_reactome and limma user guide https://bioconductor.org/packages/3.12/bioc/vignettes/limma/inst/doc/usersguide.pdf. https://hbctraining.github.io/Training-modules/Visualization_in_R/lessons/03_advanced_visualizations.html also used to guide Volcano plot construction.
## Load and format data
Load normalized expression data and cleaned metadata from script "GSE75010_preprocess.Rmd".
```{r}
expr <- read.csv(paste0(data_dir, "GSE75010_complete_dataset.csv.gz"))
meta <- readRDS(paste0(data_dir, "GSE75010_with_abundances2021-01-01.rda"))

#gse <- getGEO("GSE75010", GSEMatrix = TRUE)
```
Based on the phenoData, this array platform is single color.
```{r}
#colnames(gse$GSE75010_series_matrix.txt.gz@phenoData@data)
```
Reformat expression data to a matrix with genes as rownames.
```{r}
genes <- expr %>% dplyr::select(X)
rownames(expr) <- genes$X
expr <- expr %>% dplyr::select(-X)
expr <- as.matrix(expr)
```
Initialize an expression set object.
```{r}
d <- ExpressionSet(expr)
# Clean-up
rm (gse, genes, expr)
```
Add metadata.
```{r}
# Drop p-value, correlation, and RMSE, won't drop for now
# meta <- meta[,1:19]
pData(d) <- meta
# Clean-up
#rm(meta)
```
## PCA projection
PCA for low-dimensional representation.
```{r, eval=FALSE}
PCA <- prcomp(t(exprs(d)), scale = FALSE)
percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

pca.dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Study = 
                     pData(d)$batch,
                    Phenotype = 
                     pData(d)$phenotype)
ggplot(pca.dataGG, aes(PC1, PC2)) +
      geom_point(aes(shape = Study, colour = Phenotype)) +
  ggtitle("PCA plot of the calibrated, summarized data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  coord_fixed(ratio = sd_ratio) +
  scale_shape_manual(values = c(1:8))
#ggsave(paste0(results_dir, "GSE75010_gene_expr_PCA_", Sys.Date(), ".png"))
```

No apparent clustering by study source. Some separation in outcome by PC1.

## Defining and fitting the model

Custom function to run limma model, print results, and output Volcano plot and model results for further analysis.
```{r}
run_limma_model <-
  function(d,
           covariates,
           threshold,
           genes.of.interest,
           gg.title) {
    # Create the design formula and use to make the model matrix
    design.formula <-
      as.formula(paste0("~ ", paste(covariates, collapse = " + ")))
    design <- model.matrix(object = design.formula, data = d)
    
    #Fit the model.
    fit <- lmFit(d, design)
    fit <- eBayes(fit)
    
    #Extract model results.
    hits <- topTable(fit, coef = "phenotypePE", number = Inf)
    
    #Histogram of raw p-values
    hist(
      hits$P.Value,
      col = brewer.pal(3, name = "Set2")[1],
      main = "Pre-eclamptic patients vs. controls",
      xlab = "P-values"
    )
    #Histogram of FDR-adjusted p-values.
    hist(
      hits$adj.P.Val,
      col = brewer.pal(3, name = "Set2")[1],
      main = "Pre-eclamptic patients vs. controls",
      xlab = "FDR-adjusted p-values"
    )
    # Print summary of hits
    hits$sig <- hits$adj.P.Val < threshold
    print(paste0(
      length(which(hits$sig)),
      " significant hits at FDR-adjusted cutoff of ",
      threshold
    ))
    # Move gene symbols to column
    res <- rownames_to_column(hits, "gene_symbol")
    # Get the top downregulated hits
    downreg <- res %>%
      filter(adj.P.Val < .05) %>%
      filter(logFC < 0) %>%
      arrange(logFC)
    #Get the top upregulated hits.
    upreg <- res %>%
      filter(adj.P.Val < .05) %>%
      filter(logFC > 0) %>%
      arrange(desc(logFC))
    #Label genes of interest by name, the top 10 upregulated and downregulated hits and hand-chosen genes of interest.
    num.to.plot <- 10
    top.upreg <- upreg$gene_symbol[1:num.to.plot]
    top.downreg <- downreg$gene_symbol[1:num.to.plot]
    genes.to.plot <- c(top.upreg, top.downreg, genes.of.interest)
    res$label[(res$gene_symbol %in% genes.to.plot)] <- T
    # Volcano plot
    plot <- ggplot(res) +
      geom_point(aes(
        x = logFC,
        y = -log10(adj.P.Val),
        colour = sig
      )) +
      geom_text_repel(aes(
        x = logFC,
        y = -log10(adj.P.Val),
        label = ifelse(label, gene_symbol, "")
      )) +
      geom_hline(yintercept = -log10(threshold),
                 linetype = "dotted") +
      ggtitle(paste0("Pre-eclamptic cases vs controls - ", gg.title)) +
      xlab("log2 fold change") +
      ylab("-log10 adjusted p-value") +
      theme(
        legend.position = "none",
        plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25))
      )
    # Print plot to output
    print(plot)
    # Create list of volcano plot and models result for export
    return(list
           (volcano = plot,
             res = res))
  }
```

# Prep Phenotype data names
```{r}
p.data <- pData(d)
names(p.data) <- p.data %>% names %>%  make.names
#head(p.data)
pData(d) <- p.data
```

# Naive Model
```{r}
naive.model <- run_limma_model(d = d,
                               covariates = "phenotype",
                               threshold = .05,
                               genes.of.interest = NULL,
                               gg.title = "Naive")
```
# Demographic-adjusted model
```{r}
base.model <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch"),
                              threshold = .05,
                              genes.of.interest = NULL,
                              gg.title = "Base")
```

## Unadjusted Gene-set Enrichment

Prepare a ranked gene list (q < .05 & descending absolute logFC) for gProfiler g:GOSt gene-set enrichment input
```{r, eval = FALSE}
res <- base.model$res

# Get ranked list
ranked.list <- res %>%
  filter(adj.P.Val < .05) %>%
  arrange(desc(abs(logFC)))
# Subset to just gene names
unadj.ranked.genes <- ranked.list$gene_symbol
# Write to whitespace-delimited file for g:ProfileR input
#write(ranked.genes, paste0(data_dir, "gProfiler_input_unadjusted_model_", Sys.Date(), ".txt"))

# Printing allows easy copy/paste into g:ProfileR website if desired
#print(ranked.list$gene_symbol, quote = F)
```

## Defining and fitting the cell type proportion-adjusted model
Create the design matrix. Covariates will be case-control status, sex, gestational age in weeks, batch, and cell type proportions significantly differntly abundant from the cell type proportion beta regression models (previously naive T cell, activated T cell, Hofbauer, fibroblast, CD14+ Monocytes without assigning maternal/fetal status)

Using all cell types or even all cell types save Fetal B Cells and Fetal Naive CD4+ T cells leads to unstable model.
```{r}
# Get cell type names
cell.types <- colnames(pData(d))[6:32]
# Which cell types are all zero
summary(colSums(pData(d)[,6:32]))
cell.types <- cell.types[!cell.types %in% c("Fetal.B.Cells", "Fetal.Naive.CD4..T.Cells")]

# Set the cell type proportions to include based on significant results from the beta regression analysis.
cell.types.sig <- c("Maternal.Plasma.Cells", "Maternal.Natural.Killer.Cells", "Maternal.B.Cells", "Fetal.Naive.CD8..T.Cells", "Fetal.Mesenchymal.Stem.Cells", "Fetal.Memory.CD4..T.Cells", "Fetal.Hofbauer.Cells", "Fetal.Fibroblasts", "Fetal.Extravillous.Trophoblasts", "Fetal.Cytotrophoblasts")

type.model <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch", cell.types.sig),
                              threshold = .05,
                              genes.of.interest = "FLT1",
                              gg.title = "Sig. Cell Types")
```

Retrieve cell type proportion PCA results from beta regression .Rmd, merge with meta data, and reset the phenotype data in the ExpressionSet object.
```{r}
pca <-readRDS(paste0(data_dir, "GSE75010_cell_type_proportions_pca_2021-02-15.rda"))
pca.ind <- get_pca_ind(pca)
pca.coordinates <- as.data.frame(pca.ind$coord)
pca.coordinates$sample <- rownames(pca.coordinates)
meta.total <- left_join(meta, pca.coordinates)

# Reset the metadata of the ExpressionSet object, d.
pData(d) <- meta.total

# Set the covariates, including the outcome
PCs <- c("Dim.1", "Dim.2" , "Dim.3", "Dim.4", "Dim.5")
covariates <- c("phenotype", "sex", "ga", "batch", PCs)

pc.model <- run_limma_model(d = d,
                              covariates = covariates,
                              threshold = .05, genes.of.interest = "FLT1",
                              gg.title = "5 PCs of Cell Types")

View(pc.model$res)
```

```{r}
covariates <- c("phenotype", "sex", "ga", "batch", "Dim.3")
pc3.model <- run_limma_model(d = d,
                              covariates = covariates,
                              threshold = .05, genes.of.interest = "FLT1",
                              gg.title = "Cell type PC3 + Base")
```

Amalgamate results from the four models.
```{r}
naive.res <- naive.model$res
naive.res$model <- "naive"

base.res <- base.model$res
base.res$model <- "base"

type.res <- type.model$res
type.res$model <- "type"

pc.res <- pc.model$res
pc.res$model <- "pc"

pc3.res <- pc3.model$res
pc3.res$model <- "pc3"

res <- rbind(naive.res, base.res, type.res, pc.res, pc3.res)
```

Plotting all the above models on one Volcano plot. No cell type adjustment results in a highly significant differential expression analysis. Principal components approach dampens significance more than adjusting for cell types individually with fewer covariates. PC 3 drives the vast majority of attenuation.

```{r}
ggplot(res) +
  geom_point(aes(
    x = logFC,
    y = -log10(adj.P.Val),
    colour = model
  )) +
  geom_text_repel(aes(
    x = logFC,
    y = -log10(adj.P.Val),
    label = ifelse(label, gene_symbol, "")
  )) +
  geom_hline(yintercept = -log10(.05),
             linetype = "dotted") +
  ggtitle("Compare Limma Models") +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  theme(
    legend.position = "right"
  )

#ggsave(paste0(results_dir, "compare_limma_models_", Sys.Date(), ".png"))
```

## Distribution of PC3
Notes:
https://www.researchgate.net/post/What-is-the-meaning-of-negative-values-in-components-from-PCA-analysis simple explanation of PC loadings

Limma is linear.
Potential nonlinear relation between cell type composition and gene expression? Interaction? Aaron Lun post below describes interaction and more flexible spline model.

```{r}
df <- pData(d)
colnames(df) <- make.names(colnames(df))
```

PC3 has large difference in distribution between case-control status.
```{r}
# Unsure if rlang needed
# library(rlang)

predictor_by_outcome_smooth_hist <- function(data, predictor, outcome) {
  plot <-
    ggplot(data, aes(x = {{predictor}}, fill = {{outcome}})) + geom_density(alpha = 0.25) +
    ggtitle(paste0("Distribution by Case-Control Status")) + ylab("Density") + labs(fill = "model")
  print(plot)
}
predictor_by_outcome_smooth_hist(df, Dim.1, phenotype)
predictor_by_outcome_smooth_hist(df, Dim.2, phenotype)
predictor_by_outcome_smooth_hist(df, Dim.3, phenotype)
predictor_by_outcome_smooth_hist(df, Dim.4, phenotype)
predictor_by_outcome_smooth_hist(df, Dim.5, phenotype)
```

```{r}
model1 <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch", "Fetal.Extravillous.Trophoblasts"),
                              threshold = .05,
                              genes.of.interest = "FLT1",
                              gg.title = "Replicate PC3 - add EVT")
```

```{r}
model2 <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch", "Fetal.Extravillous.Trophoblasts", "Fetal.Hofbauer.Cells"),
                              threshold = .05,
                              genes.of.interest = "FLT1",
                              gg.title = "Replicate PC3 - add Hofbauer")
```

```{r}
model3 <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch", "Fetal.Extravillous.Trophoblasts", "Fetal.Hofbauer.Cells", "Fetal.Mesenchymal.Stem.Cells"),
                              threshold = .05,
                              genes.of.interest = "FLT1",
                              gg.title = "Replicate PC3 - add MSCs")
```

```{r}
model4 <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch", "Fetal.Extravillous.Trophoblasts", "Fetal.Hofbauer.Cells", "Fetal.Mesenchymal.Stem.Cells", "Maternal.FCGR3A..Monocytes"),
                              threshold = .05,
                              genes.of.interest = "FLT1",
                              gg.title = "Replicate PC3")
```


```{r}
model5 <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch", "Fetal.Extravillous.Trophoblasts", "Fetal.Hofbauer.Cells", "Fetal.Mesenchymal.Stem.Cells", "Maternal.FCGR3A..Monocytes", "Fetal.Syncytiotrophoblast"),
                              threshold = .05,
                              genes.of.interest = "FLT1",
                              gg.title = "Replicate PC3")
```

```{r}
model6 <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch", "Fetal.Extravillous.Trophoblasts", "Fetal.Hofbauer.Cells", "Fetal.Mesenchymal.Stem.Cells", "Maternal.FCGR3A..Monocytes", "Fetal.Syncytiotrophoblast", "Fetal.Proliferative.Cytotrophoblasts"),
                              threshold = .05,
                              genes.of.interest = "FLT1",
                              gg.title = "Replicate PC3")
```

```{r}
model <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch", "Fetal.Extravillous.Trophoblasts", "Fetal.Hofbauer.Cells", "Fetal.Mesenchymal.Stem.Cells", "Maternal.FCGR3A..Monocytes", "Fetal.Syncytiotrophoblast", "Fetal.Proliferative.Cytotrophoblasts", "Maternal.Naive.CD8..T.Cells"),
                              threshold = .05,
                              genes.of.interest = "FLT1",
                              gg.title = "Replicate PC3")
```

```{r}
model <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch", "Fetal.Extravillous.Trophoblasts", "Fetal.Hofbauer.Cells", "Fetal.Mesenchymal.Stem.Cells", "Maternal.FCGR3A..Monocytes", "Fetal.Syncytiotrophoblast", "Fetal.Proliferative.Cytotrophoblasts", "Maternal.Naive.CD8..T.Cells", "Fetal.Endothelial.Cells"),
                              threshold = .05,
                              genes.of.interest = "FLT1",
                              gg.title = "Replicate PC3")
```

```{r}
model <- run_limma_model(d = d,
                              covariates = c("phenotype", "sex", "ga", "batch", "Fetal.Extravillous.Trophoblasts", "Fetal.Hofbauer.Cells", "Fetal.Mesenchymal.Stem.Cells", "Maternal.FCGR3A..Monocytes", "Fetal.Syncytiotrophoblast", "Fetal.Proliferative.Cytotrophoblasts", "Maternal.Naive.CD8..T.Cells", "Fetal.Endothelial.Cells", "Fetal.Fibroblasts"),
                              threshold = .05,
                              genes.of.interest = "FLT1",
                              gg.title = "Replicate PC3")
```

## Cell type-specific TCA

Data prep for TCA
```{r}
library(matrixcalc)
library(TCA)
# X corresponds to the bulk data (genes x individuals)
# Code to sort X and W samples the same way
X <- as.data.frame(t(d@assayData$exprs))
X <- rownames_to_column(X, var = "sample")
X <- X %>%
  arrange(sample)
rownames(X) <- X$sample
X <- X[,-1]
X <- t(X)

# W corresponds to cell type composition estimates for individuals
W.prep <- meta[,c(1:32)]
W.prep <- W.prep %>%
  arrange(sample)
rownames(W.prep) <- W.prep$sample
W.prep <- W.prep[, -1]
colnames(W.prep) <- make.names(colnames(W.prep))

# Extract cell type proportions for W
W <- W.prep[, 5:31]
# Drop Fetal B Cells and Fetal Naive CD4 T Cells to fix non-definite error
keep <- colnames(W) %in% c("Fetal.Naive.CD4..T.Cells", "Fetal.B.Cells")
keep <- !keep
W <- W[, keep]

# Check that W and X sample orders are identical
summary(colnames(X) == rownames(W))

# Extract sex, phenotype, and ga as biological factors that may affect cell type-specific gene expression
C1 <- W.prep [ , c(1, 3, 4)]
C1 <- rownames_to_column(C1, var = "sample")

# Convert factor variables to dummies
C1 <- C1 %>% mutate(sex.male = ifelse(sex == "M", 1, 0))
C1 <- C1 %>% mutate(phenotype.case = ifelse(phenotype == "PE", 1, 0))
C1 <- C1 %>% select(sample, phenotype.case, sex.male, ga)
rownames(C1) <- C1$sample
C1 <- C1[,-1]

# Extract batch as a technical covariate that should affect bulk gene expression only
C2 <- W.prep [ , 2, drop = FALSE]
C2 <- model.matrix(data = C2, object = ~ 0 + batch)
```

All rowsums of W are equal to 1?
```{r}
sums <- rowSums(W)
summary(sums)
```

Model directionality: assume that outcome (Y) affects expression (X), X|Y or the opposite, Y|X. Here, phenotype (preeclampsia case-control) affects expression (X|Y). Gestational age or sex would be similar (predictors of expression). tca() function is used to statistically test X|Y.

The marginal conditional model is similar to a T-test in linear, testing each cell type separately. The marginal model is similar to an F-test, looking at the combined statistical evidence coming from all the cell types. Generally advisable to start with the marginal model and do the marginal conditional post-hoc if signifance is detected in the marginal model, though this may not always be the most powerful approach.
```{r, eval = FALSE}
tca <- tca(X = X,
           W = W,
           C1 = C1,
           C2 = C2)
```

## Explore correlation and interaction relationships

```{r}
expression_data <- as.data.frame(t(d@assayData$exprs))
flt1 <- expression_data %>% select(FLT1)
flt1 <- rownames_to_column(flt1, var = "sample")
df <- left_join(df, flt1)
df$FLT1 <- log2(df$FLT1)
```


```{r}
numeric.data <- df[, c(4,6:32, 36:41, 61)]
keep <- colSums(numeric.data)!=0
numeric.data <- numeric.data[, keep]
# Compute a correlation matrix
corr <- round(cor(numeric.data), 1)

#dev.new()
#pdf(paste0(results_dir, "numeric_data_corrplot_", Sys.Date(), ".pdf"))
#corrplot.mixed(corr, lower = "ellipse", tl.cex = .7, upper = "number") #tl.cex control axis label size
corrplot(corr, method = "ellipse", tl.cex = .7) #tl.cex control axis label size
#dev.off()

# Alternative options
#corrplot(corr, method = "ellipse", order = "hclust")
#corrplot(corr, method = "ellipse", order = "AOE")
```

```{r}
ggplot(df, aes(x = Dim.3, y = FLT1, color = phenotype)) + geom_point() + geom_smooth(method=lm)
ggplot(df, aes(x = Fetal.Extravillous.Trophoblasts, y = FLT1, color = phenotype)) + geom_point() + geom_smooth(method=lm)
ggplot(df, aes(x = Dim.3, y = Fetal.Extravillous.Trophoblasts, color = phenotype)) + geom_point() + geom_smooth(method=lm)
ggplot(df, aes(x = ga, y = Fetal.Extravillous.Trophoblasts, color = phenotype)) + geom_point() + geom_smooth(method=lm)
ggplot(df, aes(x = ga, y = FLT1, color = phenotype)) + geom_point() + geom_smooth(method=lm)
```

## Mediation
Using the mediation package approach
Add log2(FLT1) expression data to metadata frame
```{r}
library(mediation)
```

Mediator model (PC3 in this case)
```{r}
med.fit <- lm(Dim.3 ~ phenotype + sex + ga + batch, data = df)
print(tidy(med.fit))
```

Outcome model
```{r}
out.fit <- lm(FLT1 ~ Dim.3 + phenotype + sex + ga + batch, data = df)
print(tidy(out.fit))
```

```{r}
med.out <- mediate(med.fit, out.fit, treat = "phenotype", mediator = "Dim.3", robustSE = TRUE, sims = 100)
summary(med.out)
```


FLT1 attentuation after cell type proportion adjustment in Limma (before and after cell type adjustment, 5 PCs)
```{r}
base.model$res %>% filter(gene_symbol == "FLT1")
pc.model$res %>% filter(gene_symbol == "FLT1")
```

## Interaction

# Simple interaction model - PCs
No evidence for interaction with PC3 of cell type proportions.
```{r}
interaction.flt1 <- lm(log2(FLT1) ~ sex + ga + batch + Dim.3*phenotype, data = df)
print(tidy(interaction.flt1))

interaction.flt1 <- lm(log2(FLT1) ~ sex + ga + batch + Dim.1 + Dim.2 + Dim.3*phenotype + Dim.4 + Dim.5, data = df)
print(tidy(interaction.flt1))
```
# Limma interaction model - PCs
```{r}
design <- model.matrix(object = ~ sex + ga + batch + Dim.1 + Dim.2 + phenotype*Dim.3 + Dim.4 + Dim.5, data = d)
# Check the design coefficients
colnames(design)

#Fit the model.
fit <- lmFit(d, design)
fit <- eBayes(fit)
```

Preeclampsia effect
```{r}
#Extract model results.
interaction <- topTable(fit, coef = "phenotypePE", number = Inf)
interaction <- rownames_to_column(interaction, var = "gene_symbol")
res.flt1 <- interaction %>%
  filter(gene_symbol == "FLT1")
print(res.flt1)
```

PC3 continous effect
```{r}
interaction <- topTable(fit, coef = "Dim.3", number = Inf)
interaction <- rownames_to_column(interaction, var = "gene_symbol")
res.flt1 <- interaction %>%
  filter(gene_symbol == "FLT1")
print(res.flt1)
```

Preeclampsia/PC3 Interaction
```{r}
interaction <- topTable(fit, coef = "phenotypePE:Dim.3", number = Inf)
interaction <- rownames_to_column(interaction, var = "gene_symbol")
res.flt1 <- interaction %>%
  filter(gene_symbol == "FLT1")
print(res.flt1)
```

```{r}
#coefficients <- c("phenotypePE", "Dim.3", "phenotypePE:Dim.3", "sexM", "ga")

# Trying to map over coefficients without sucess
#results <- map(.x = coefficients, .f = limma::topTable(), fit = fit, coef = coefficients, number = Inf)

#covariates <- c("sex", "ga", "batch", "Dim.1", "Dim.2", "phenotype*Dim.3", "Dim.4", "Dim.5")
#limma.interaction <- run_limma_interaction_model(d = d,
#                              covariates = covariates,
#                              threshold = .05,
#                              genes.of.interest = "FLT1",
#                              gg.title = "Sig. Cell Types")
#
#limma.interaction$res %>% filter(gene_symbol == "FLT1")
```

```{r}
ggplot(res,aes(x=logFC, fill=model)) + geom_density(alpha=0.25) +
  ggtitle("Distribution of Gestational Age by Case-Control Status") +
  xlab("Gestational Age (wks)") + ylab("Density") + labs(fill = "model")
```

```{r}
ggplot(res, aes(x = logFC, y = model)) +
  geom_density_ridges(
    jittered_points = TRUE, scale = .95, rel_min_height = .01,
    point_shape = "|", point_size = 1, size = 0.25,
    position = position_points_jitter(height = 0)
  ) +
  ggtitle("logFC distribution by model") +
  theme_ridges(center = TRUE) +
  theme(legend.title = element_blank(), legend.position = c(.65, .2), legend.text = element_text(size = 10), axis.title.x=element_blank(), plot.title = element_text(size = 12), axis.text.y = element_text(size = 10), axis.title.y = element_blank())
```

```{r}
ggplot(res, aes(x = -log10(adj.P.Val), y = model)) +
  geom_density_ridges(
    jittered_points = TRUE, scale = .95, rel_min_height = .01,
    point_shape = "|", point_size = 1, size = 0.25,
    position = position_points_jitter(height = 0)
  ) +
  ggtitle("-log10(adj.P.Val) distribution by model") +
  theme_ridges(center = TRUE) +
  theme(legend.title = element_blank(), legend.position = c(.65, .2), legend.text = element_text(size = 10), axis.title.x=element_blank(), plot.title = element_text(size = 12), axis.text.y = element_text(size = 10), axis.title.y = element_blank())
```

Manually set axis limits and drawing side-by-side.
```{r, eval = FALSE}
p1 <- plot1 + coord_cartesian(xlim = c(-.45, 1.2), ylim = c(0, 5)) + ggtitle("Unadjusted")
#ggsave(paste0(data_dir, "GSE75010_dex_unadj_", Sys.Date(), ".png"), device = png())
p2 <- plot2 + coord_cartesian(xlim = c(-.45, 1.2), ylim = c(0, 5)) + ggtitle("Cell type-adjusted") +
  geom_hline(yintercept = -log10(.05), linetype="dotted")
#ggsave(paste0(data_dir, "GSE75010_dex_adj_", Sys.Date(), ".png"), device = png())
```
Draw and save.
```{r, eval = FALSE}
png(paste0(results_dir, "GSE75010_dex_2_panel", Sys.Date(), ".png"))
grid.newpage()
grid.draw(cbind(ggplotGrob(p1), ggplotGrob(p2)))
dev.off()
```

## Adjusted Gene-set Enrichment

Prepare a ranked gene list (q < .05 & descending absolute logFC) for gProfiler g:GOSt gene-set enrichment input
```{r, eval = FALSE}
# Get ranked list
ranked.list <- res %>%
  filter(adj.P.Val < .05) %>%
  arrange(desc(abs(logFC)))
# Subset to just gene names
adj.ranked.genes <- ranked.list$gene_symbol
# Write to whitespace-delimited file for g:ProfileR input
#write(ranked.genes, paste0(data_dir, "gProfiler_input_unadjusted_model_", Sys.Date(), ".txt"))

# Printing allows easy copy/paste into g:ProfileR website if desired
#print(ranked.list$gene_symbol, quote = F)
```

## g:Profiler multiquery comparing unadjusted and cell type-adjusted model

Perform the enrichment test using both ranked gene lists, excluding automatically generated terms, using only annotated genes as the gene universe background.
```{r, eval = FALSE}
gostres <- gost(query = list("Cell Type-adjusted" = adj.ranked.genes, "Unadjusted" = unadj.ranked.genes),
                organism = "hsapiens", ordered_query = TRUE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = TRUE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE)
```

Plot all results as Manhattan Plot
```{r, eval = FALSE}
gostplot(gostres, capped = FALSE, interactive = FALSE)
#ggsave(paste0(data_dir, "gProfiler/", "gost_plot_all_", Sys.Date(), ".png"), device = "png")
```

Limit to gene sets less than 200 genes (term_size).
```{r, eval = FALSE}
# Limit the results
gostres.small <- gostres
gostres.small$result <- gostres$result %>%
  filter(term_size < 201)
```

Plot all terms less than 200 genes
```{r, eval = FALSE}
p <- gostplot(gostres.small, capped = FALSE, interactive = FALSE)
p
#ggsave(paste0(data_dir, "gProfiler/", "gost_plot_term_size_limit_200_", Sys.Date(), ".png"), device = "png")

```

Identify terms to highlight
```{r, eval = FALSE}
# Extract the results dataframe
res <- gostres.small$result
#write_csv(x = res, file = paste0(data_dir, "gProfiler/", "GSE75010_dex_gProfiler_200_term_size_results ", Sys.Date(), ".csv"))

# Get terms to highlight

# all significant cell type-adjusted results
adjusted.hits <- 
  res %>% 
  filter(query == "Cell Type-adjusted") %>%
  filter(significant == TRUE)
adjusted.hits <- adjusted.hits$term_id

# Get top unadjusted hits
top.unadjusted.hits <- 
  res %>% 
  filter(query == "Unadjusted")
top.unadjusted.hits <- top.unadjusted.hits$term_id[1:5]

terms.to.highlight <- c (adjusted.hits, top.unadjusted.hits)

publish_gosttable(gostres.small, 
                         highlight_terms = terms.to.highlight,
                        use_colors = TRUE, 
                        show_columns = c("source", "term_name", "term_size"),
                        filename = NULL)
```

Plot Manhattan with highlighted terms
```{r, eval = FALSE}
pp <- publish_gostplot(p, highlight_terms = terms.to.highlight, filename = NULL )
pp
#ggsave(paste0(data_dir, "gProfiler/", "gost_highlighted_terms_", Sys.Date(), ".png"), device = "png")
```

```{r, eval = FALSE}
gostres <- gost(query = list("Cell Type-adjusted" = adj.ranked.genes, "Unadjusted" = unadj.ranked.genes),
                organism = "hsapiens", ordered_query = TRUE, 
                multi_query = TRUE, significant = TRUE, exclude_iea = TRUE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE)
gostres.link <- gost(query = list("Cell Type-adjusted" = adj.ranked.genes, "Unadjusted" = unadj.ranked.genes),
                organism = "hsapiens", ordered_query = TRUE, 
                multi_query = TRUE, significant = TRUE, exclude_iea = TRUE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = TRUE)
```

Limit to gene sets less than 200 genes (term_size).
```{r, eval = FALSE}
# Limit the results
gostres.small <- gostres
gostres.small$result <- gostres$result %>%
  filter(term_size < 201)
```

Plot all terms less than 200 genes
```{r, eval = FALSE}
p <- gostplot(gostres.small, capped = FALSE, interactive = FALSE)
p
```

```{r, eval = FALSE}
result <- gostres.small$result
idx <- result[result$term_id %in% terms.to.highlight, ]
indices <- rownames(idx)
```


Plot Manhattan with highlighted terms
```{r, eval = FALSE}
#publish_gostplot(p, highlight_terms = gostres$result[indices, ], width = NA, height = NA,
#                 filename = paste0(data_dir, "gProfiler/", "gost_highlighted_terms_", Sys.Date(), ".png"))

publish_gostplot(p, highlight_terms = terms.to.highlight, width = NA, height = NA,
                 filename = paste0(data_dir, "gProfiler/", "gost_highlighted_terms_", Sys.Date(), ".png"))

```


```{r, eval = FALSE}
ggsave(paste0(data_dir, "gProfiler/", "gost_highlighted_terms_", Sys.Date(), ".png"), device = "png")
```


```{r, eval = FALSE}
publish_gosttable(gostres.small, 
                         highlight_terms = terms.to.highlight,
                        use_colors = TRUE, 
                        show_columns = c("source", "term_name", "term_size"),
                        filename = NULL)
ggsave(paste0(data_dir, "gProfiler/", "gost_highlighted_terms_table_", Sys.Date(), ".png"), device = "png")
```

