---
title: "DESeq2 Analysis of Placenta Sorted Cell Types"
subtitle: "Kyle Campbell - March 14th, 2019"
output: html_document
---

```{r setup, include=FALSE}
# Update R all installed R Packages
# update.packages(ask = FALSE)

library(knitr)
library(pheatmap)
library(RColorBrewer)
library(tidyverse)

library(BiocManager)
library(biomaRt)
library(DESeq2)
library(EnhancedVolcano)
library(GO.db)
library(mygene)
library(org.Hs.eg.db)
library(topGO)

knitr::opts_chunk$set(echo=TRUE)

# May need to update root.dir Drive Letter
local_data_path <- "C:/Users/k_cam/Google Drive/Placenta_Cell_Types_Kyle/code/R code/local/placenta_sort/dex/data/analysis_datasets/"
```
## Entering Data, Preproccessing, and Model Fitting
```{r DESeq2_object}

#counts <- readRDS(file =   "./data/analysis_datasets/edgeR_collapsed_counts.Rda")
#meta <- readRDS(file =     "./data/analysis_datasets/edgeR_collapsed_metadata.Rda")
#gene_annotation <- readRDS("./data/analysis_datasets/edgeR_gene_annotation_with_symbols.Rda")

# Must make meta rownames identical to counts colnames
#rownames(meta) <- substr(rownames(meta), 8, 14)

# Creating the DESeq2 experiment object from count data and meta info
#dds <- DESeqDataSetFromMatrix(countData = counts,
#                              colData = meta,
#                              design = ~ 0 + sample.ID + cell.type)

# Adding comprehensive gene annotation to dds
#mcols(dds) <- DataFrame(mcols(dds), gene_annotation)

# Minimal pre-processing to reduce computational load - does NOT represent meaningful filtering
#keep <- rowSums(counts(dds)) >= 10
#dds <- dds[keep,]

# Fitting the model
#dds <- DESeq(dds)
```

```{r AddGeneAnnotation}

# Chop off the .xx from the ENSEMBL ID
#row.names(dds) <- substr(x = row.names(dds), start = 1, stop = 15)

#row.names(dds) <- mapIds(org.Hs.eg.db,
#                     keys=row.names(dds),
#                     column="SYMBOL",
#                     keytype="ENSEMBL",
#                     multiVals="first")
```

BiomaRt returns 58,069 of 58,381 gene features from DESeq2 results. Of the 58,069 results, 56,521 return unique external gene names and 58,024 return unique ENSEMBL gene IDs.

```{r biomaRt_DESeq2_results}
# Setting up biomaRt
# Assigning Mart of interest, human ensembl here
mart = useMart("ensembl",dataset="hsapiens_gene_ensembl")

# Querying biomaRt using gene names to get back ENSEMBL IDs
DESeq2GeneIDs <- getBM(attributes = c('ensembl_gene_id', 'external_gene_name'),
             filters = c('ensembl_gene_id'),
             values = rownames(dds),
             mart = mart)
genes.tbl <- as_tibble(DESeq2GeneIDs)
```

Save/load 
```{r}
#saveRDS(DESeq2GeneIDs_tbl, "biomaRt_DESeq2_ensembl_ids.rda")
#DESeq2GeneIDs_tbl <- as_tibble(readRDS("biomaRt_DESeq2_ensembl_ids.rda"))
```

```{r SaveLoad}
# Saved the fitted model object to avoid having to reestimate parameters
#saveRDS(dds, file = "./data/analysis_datasets/DESeq2_dds_noint_celltype_sampleid_nofiltering.rda")
#dds <- readRDS(file = "./data/analysis_datasets/DESeq2_dds_noint_celltype_sampleid_nofiltering.rda")

# Load the finished DESeq2 object - note this object has trimmed ensembl IDs
#dds <- readRDS("C:/Users/Kyle/Google Drive/Placenta_Cell_Types_Kyle/code/R code/local/placenta_sort/dex/data/analysis_datasets/DESeq2_dds_noint_celltype_sampleid_nofiltering.rda")

dds <- readRDS(paste0(local_data_path, "/DESeq2_dds_noint_celltype_sampleid_nofiltering.rda"))
```

```{r}
plotDispEsts( dds, ylim = c(1e-4, 1e2) )
```

Custom function to summarize DESeq2 significant results based on different p-value cutoffs. Note that the input DESeq2 results object will have been run at some alpha level for the purpose of independent filtering that may slightly change results.

``` {r}
## Arguments:
## DESeq2_Results is a DESeq2 results (res) object.
## Cell_Type_Identity is a string to describe the analysis that was done for the results. In this case, the cell type of interest that is being compared against other placental cell types.
## FDR_Threshold is the desired DESeq2 p-adj cutoff you would like the function to return as a filtered, tibble-format DESeq2 results object.
## Gene_Anno_Tbl is gene annotation information to append to results
## Returns:
## List of two results tibbles at FDR cutoff with genes dropped due to independent filitering, all zero counts, or extreme outliers (none of which are used for DESeq2 model) and the universe of genes tested (those not used in the model and duplicate gene annotations)
summarizeDESeq2Results <-
   function(DESeq2_Results, Cell_Type_Identity, FDR_Threshold, Genes_Anno_Tbl) {
     
     # Coerce DESeq2 Results to df
     res.df <- as.data.frame(DESeq2_Results)
     
     # Move gene ensembl IDs from row to table column
     res <- rownames_to_column(res.df, "ensembl_gene_id")
     
     # Add gene annotation and drop duplicate gene names
     res <- left_join(res, genes.tbl) %>%
       dplyr::select(external_gene_name, everything()) %>%
       relocate(ensembl_gene_id, .after = last_col()) %>%
       group_by(external_gene_name) %>%
       filter(!(n() > 1))
       as_tibble()
     
     # Identify genes dropped from DESeq2 analysis due to independent filtering
     stats <- summary(res$padj)
     print(
       paste0(
         "For ",
         Cell_Type_Identity,
         ", ",
         stats["NA's"],
         " features dropped due to independent filtering, all zero counts, or extreme outlier status in DESeq2"
       )
     )
     # drop genes dropped for DESeq2 analysis
     res <-
       res %>%
       filter(!is.na(padj))
     
     print(paste0(
       "For ",
       Cell_Type_Identity,
       ", ",
       dim(res)[1],
       " genes assessed for significance"
     ))
     # # Limit to only upregulated genes
     # res <-
     #   res %>%
     #   filter(log2FoldChange > 0)
     # print(paste0(
     #   "For ",
     #   Cell_Type_Identity,
     #   ", ",
     #   dim(res)[1],
     #   " upregulated genes assessed for significance"
     # ))
     # Assess p-value cutoffs from DESeq2 analysis
     for (pvalcutoff in c(0.05, 0.01, 0.005, 0.001)) {
       res.cut <-
         res %>%
         filter(!is.na(padj)) %>%
         filter(padj < pvalcutoff)# %>%
         #filter(log2FoldChange > 0)
       print(
         paste0(
           "For ",
           Cell_Type_Identity,
           ", ",
           dim(res.cut)[1],
           " genes pass at FDR-adjusted ",
           pvalcutoff
         )
       )
     }
     # Generate table to be returned
     result <- 
       res %>%
       filter(!is.na(padj)) %>%
       filter(padj < FDR_Threshold)# %>%
       #filter(log2FoldChange > 0)
     
     # Remind user which FDR-adjusted value they called for
     print(
       paste0(
         "Returning upregulated results at FDR-adjusted ",
         FDR_Threshold
       )
     )
     
     return <- list(result = result,
                    universe = res)

     return(return)
   }
```
## Comparing cell type-specific gene expression to the average expression level across other cell types

Our first contrast of interest is how cell type-specific expression differs vs. the average expression across all other cell types. [This thread](https://support.bioconductor.org/p/118090/) recommends fitting a no-intercept model to answer this question. [Another thread](https://support.bioconductor.org/p/86347/) contains a brief discussion of how to code it. These contrasts will exclude the composite tissue group. I wonder if excluding the composite samples from normalization and filtering is appropriate for this particular set of comparisions. I think that also depends on whether we use the composite samples in any analyses.

``` {r}
# For the DESeq2 results() call in this chunk:
# The first vector is our cell type of interest
# Create the contrast of interest
# The following vector contains comparison group
# listValues Specifies comparision against average expression across other groups

# Set alpha
alpha <- .01
# Pull contrast of interest
contrast <- results(
  dds,
  contrast = list( c("cell.typecytotrophoblast"), 
                   c("cell.typeextravilloustrophoblast",
                     "cell.typefibroblast",
                     "cell.typehofbauer",
                     "cell.typesyncytiotrophoblast")),
  listValues = c(1, -1/4),
  alpha = alpha
)

# Summarize and annotate contrast results
return <- summarizeDESeq2Results(
  DESeq2_Results = contrast,
  Cell_Type_Identity = "Cytotrophoblasts",
  FDR_Threshold = alpha,
  Genes_Anno_Tbl = genes.tbl)

res <- return$result
universe <- return$universe

universe$external_gene_name %in% res.tbl$external_gene_name
```

Label genes of interest by name, the top 10 upregulated and downregulated hits and hand-chosen extravillous trophoblast marker gene HLA-G.
```{r}
num.to.plot <- 10
genes.of.interest <- c("HLA-G")
top.upreg <- upreg$gene_symbol[1:num.to.plot]
top.downreg <- downreg$gene_symbol[1:num.to.plot]
genes.to.plot <- c(top.upreg, top.downreg, genes.of.interest)
res$label[(res$gene_symbol %in% genes.to.plot)] <- T
```
Volcano plot of results. The chunk above will determine which genes are labelled in the plot.
```{r}
# Volcano plot
plot1 <- ggplot(res) +
        geom_point(aes(x=logFC, y=-log10(adj.P.Val), colour=sig)) +
        geom_text_repel(aes(x = logFC, y = -log10(adj.P.Val),
                            label = ifelse(
                              label, gene_symbol, ""
                            ))) +
        geom_hline(yintercept = -log10(threshold), linetype="dotted") +
        ggtitle("Pre-eclamptic cases vs controls") +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))
plot1
```

``` {r cyto_vs_all}
# Get Cytotrophoblast results
 ct.res <- SummarizeDESeq2Results(cytovall_res, "Cytorophoblasts", .005)

# GO analysis
# Create the numeric vector FDR-adjusted p-values from DESeq2 DEX output, remove missing p-vals from DESEq2, and name list with genes for topGO input
ctGenes <- ct.res$padj
names(ctGenes) <- rownames(ct.res)
ctGenes <- na.omit(ctGenes)
str(ctGenes)

# Define a function for padj cutoff selection
topDiffGenes <- function(allScore) {
  return(allScore < 0.05)
}

# Constructing the topGOdata object for downstream analysis
ctGOdata <- new("topGOdata",
                ontology = "BP",
                allGenes = ctGenes,
                geneSelectionFun = topDiffGenes,
                nodeSize = 10,
                annot = annFUN.org,
                mapping = "org.Hs.eg.db",
                ID = "ensembl")

# Running the enrichment test using the weight01Score method because weight method accounts for GO topology and therefore accoutns for multiple testing. Score method to employ the nonparametric Kolmogorov-Smirnov test based on gene rank (also known as GSEA)
test.stat <- new(Class = "weight01Score", testStatistic = GOKSTest, name = "Weight01 KS tests")

# getSigGroups returns an object of class 'topGOresult', here named resultweight01KS to reflect test employed
resultweight01KS <- getSigGroups(ctGOdata, test.stat)

# Already performed in dataset
# # Chop off the .xx from the ENSEMBL ID
# row.names(res) <- substr(x = row.names(res), start = 1, stop = 15)
# 
# # Add gene symbols
# res$symbol <- mapIds(org.Hs.eg.db,
#                      keys=row.names(res),
#                      column="SYMBOL",
#                      keytype="ENSEMBL",
#                      multiVals="first")
plotMA(res)
summary(res)

res_df <- as.data.frame(res)
dim(res_df) #58381 genes
summary(res_df$pvalue) #8862 missing, with zero counts in all conditions
length(res_df$stat[res_df$stat==0]) #14208 that have test statistics of 0, independent filtered out?
58381 - 8862 - 14208 #35311 total genes used in this contrast

testd <- res_df %>%
  filter(padj < .05) %>%
  arrange(desc(baseMean), desc(log2FoldChange))

testd <- as_tibble(testd)
testd <- rename(testd, scGenes = symbol)

testleftjoin <- left_join(scGenes_df, testd, by = "scGenes")
testleftjoin %>%
  arrange(desc(baseMean), desc(log2FoldChange))
testleftjoin <- drop_na(testleftjoin)
write.csv(x = testleftjoin, file = "CTvsAll_alpha005_scOverlap.csv")

```
### Cytotrophoblast vs. All Volcano Plot
```{r cytovall_volcano, echo = FALSE}
    EnhancedVolcano(res,
        lab = res$symbol,
        x = "log2FoldChange",
        y = "padj",
        
        xlab = bquote(~Log[2]~ "fold change"),
        ylab = bquote(~-Log[10]~adjusted~italic(P)),
        
        pCutoff = .005,
        FCcutoff = 2.0,
        
        #transcriptPointSize = 1.5,
        #transcriptLabSize = 3.0,
        colAlpha = 1,
        
        title = "Cytotrophoblast vs. Others",
        
        legend=c("NS","Log2 FC","Adjusted p-value",
                 "Adjusted p-value & Log2 FC"),
        legendPosition = "bottom",
        legendLabSize = 10,
        legendIconSize = 3.0,
        
        drawConnectors = FALSE,
        widthConnectors = 0.2,
        colConnectors = "grey30"
        )
```

``` {r hb_vs_all}
# Create the contrast of interest
# The first vector is our cell type of interest
# The following vector contains comparison group
# listValues Specifies comparision against average expression across other groups
hbvall_res <- results(dds,
               contrast = list( c("cell.typehofbauer"),
                                c("cell.typeendothelial",
                                  "cell.typeextravilloustrophoblast",
                                  "cell.typefibroblast",
                                  "cell.typecytotrophoblast",
                                  "cell.typesyncytiotrophoblast")),
               listValues = c(1, -1/5),
               alpha = 0.005
               )
# Rename to res for easy replication of DESeq2 vignette code
res <- hbvall_res

# Chop off the .xx from the ENSEMBL ID
row.names(res) <- substr(x = row.names(res), start = 1, stop = 15)

# Add gene symbols
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
plotMA(res)
summary(res)

res_df <- as.data.frame(res)
dim(res_df) #58381 genes
summary(res_df$pvalue) #8862 missing, with zero counts in all conditions
length(res_df$stat[res_df$stat==0]) #12023 that have test statistics of 0, independent filtered out?
58381 - 8862 - 12023 #37496 total genes used in this contrast

testd <- res_df %>%
  filter(padj < .005) %>%
  arrange(desc(baseMean), desc(log2FoldChange))

testd <- as_tibble(testd)
testd <- rename(testd, scGenes = symbol)

testleftjoin <- left_join(scGenes_df, testd, by = "scGenes")
testleftjoin %>%
  arrange(desc(baseMean), desc(log2FoldChange))
testleftjoin <- drop_na(testleftjoin)
write.csv(x = testleftjoin, file = "HBvsAll_alpha005_scOverlap.csv")

```

```{r hbvall_volcano, echo = FALSE}
    EnhancedVolcano(res,
        lab = res$symbol,
        x = "log2FoldChange",
        y = "padj",
        
        xlab = bquote(~Log[2]~ "fold change"),
        ylab = bquote(~-Log[10]~adjusted~italic(P)),
        
        pCutoff = 0.05,
        FCcutoff = 2.0,
        
        transcriptPointSize = 1.5,
        transcriptLabSize = 3.0,
        colAlpha = 1,
        
        legend=c("NS","Log2 FC","Adjusted p-value",
                 "Adjusted p-value & Log2 FC"),
        legendPosition = "bottom",
        legendLabSize = 10,
        legendIconSize = 3.0,
        
        drawConnectors = FALSE,
        widthConnectors = 0.2,
        colConnectors = "grey30",
        
        title = "Hofbauer Cells vs. Others")
```

``` {r fb_vs_all}
# Create the contrast of interest
fbvall_res <- results(dds,
               contrast = list( c("cell.typefibroblast"), # The first vector is our cell type of interest
                                c("cell.typeendothelial",  # The following vector contains comparison group
                                  "cell.typeextravilloustrophoblast",
                                  "cell.typehofbauer",
                                  "cell.typecytotrophoblast",
                                  "cell.typesyncytiotrophoblast")),
               listValues = c(1, -1/5) # Specifies comparision against average expression across other groups
               )
# Rename to res for easy replication of DESeq2 vignette code
res <- fbvall_res

# Chop off the .xx from the ENSEMBL ID
x <- substr(x = row.names(res), start = 1, stop = 15)

# Add gene symbols
res$hgnc_symbol <- convertIDs( x, "ENSEMBL", "SYMBOL", org.Hs.eg.db)

plotMA(res)
summary(res)

results_df <- as.data.frame(res)

results_df %>%
  filter(hgnc_symbol=="COL1A1")



# Create the contrast of interest
fbvall_res <- results(dds,
               contrast = list( c("cell.typefibroblast"), # The first vector is our cell type of interest
                                c("cell.typeendothelial",      # The following vector contains comparison group
                                  "cell.typeextravilloustrophoblast",
                                  "cell.typehofbauer",
                                  "cell.typecytotrophoblast")),
               listValues = c(1, -1/4) # Specifies comparision against average expression across other groups
               )
# Rename to res for easy replication of DESeq2 vignette code
res <- fbvall_res

# Chop off the .xx from the ENSEMBL ID
x <- substr(x = row.names(res), start = 1, stop = 15)

# Add gene symbols
res$hgnc_symbol <- convertIDs( x, "ENSEMBL", "SYMBOL", org.Hs.eg.db)

plotMA(res)
summary(res)

results_df <- as.data.frame(res)

results_df %>%
  filter(hgnc_symbol=="COL1A1")
```

``` {r e}
counts_test <- counts
counts_test$hgnc_symbol <- convertIDs(x, "ENSEMBL", "SYMBOL", org.Hs.eg.db)
COL1A1 <- counts_test %>%
  filter(hgnc_symbol=="COL1A1")
COL1A1 <- COL1A1[,1:24]
col1a1 <- as.data.frame(COL1A1)
col1a1[2,] <- names(col1a1)

tes <- data.frame()#data.frame(reads=integer(col1a1[1,]), samples = factor(names(col1a1)))
tes$samples <- names(col1a1)
tes$reads <- col1a1[1,]

a <- ggplot(col1a1, aes(x = col1a1[2,], y = col1a1[1,]))
a + geom_col()

write.csv(col1a1, )
```


```{r fbvall_volcano, echo = FALSE}
    EnhancedVolcano(res,
        lab = res$hgnc_symbol,
        x = "log2FoldChange",
        y = "padj",
        
        selectLab = "ENSG00000108821",
        
        xlab = bquote(~Log[2]~ "fold change"),
        ylab = bquote(~-Log[10]~adjusted~italic(P)),
        
        pCutoff = .1,
        FCcutoff = 1.0,
        
        transcriptPointSize = 1.5,
        transcriptLabSize = 3.0,
        colAlpha = 1,
        
        legend=c("NS","Log2 FC","Adjusted p-value",
                 "Adjusted p-value & Log2 FC"),
        legendPosition = "bottom",
        legendLabSize = 10,
        legendIconSize = 3.0,
        
        drawConnectors = FALSE,
        widthConnectors = 0.2,
        colConnectors = "grey30",
        
        title = "Fibroblast Cells vs. Others")
```
## Data Quality Assessment by Clustering and Visualization

We'll begin by transforming the data to the log-fold-2 scale using the same shrinkage transformation employed in the DEX pipeline.

```{r transformation}
rld <- rlog(dds, blind = TRUE)
#saveRDS(rld, file = paste0(local_data_path, "DESeq2_rld.rda"))
rld <- readRDS(file = paste0(local_data_path, "DESeq2_rld.rda"))
```
103911 - leukocyte is an extreme outlier on blind PCA analysis, will remove and rerun. PCA looks similar with 103911 - leukocyte removed
```{r}
dds_103911 <- dds[,colnames(dds) != "103911"]
dds_103911 <- DESeq(dds_103911)
rld <-rlog(dds_103911, blind = TRUE)
```
### Hierarchical Clustering Heatmap
```{r hc_heatmap}
# Calculate sample distances
sampleDists <- dist(t(assay(rld)))

# Hierarchical heatmap clustering - DESeq2 vignette
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$cell.type, rld$sample.ID, rld$is.Paired, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```
### PCA Plot
```{r rld_pca}
# Customized PCA ggplot
pcaData <- plotPCA(rld, intgroup=c("cell.type", "sample.ID"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=cell.type, shape=sample.ID)) +
  geom_point(size=3.5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
#ggsave(sprintf("%sSorted_PCA.png", local_output_path), device = "png")
```