---
title: " Maternal Cell type assignment - correcting by biorep for batch"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::knit_engines$set(python = reticulate::eng_python)

#install.packages("package_name",  repos="http://cran.us.r-project.org") # Only this mirror worked for multiple packages
library(here)
library(knitr)
library(tidyverse)
library(devtools)
library(Seurat)
#remotes::install_github('satijalab/seurat-wrappers')
library(SeuratWrappers)
library(cowplot)
library(rgl)
library(EnvStats)
library(reticulate)
library(clustree)
library(Matrix)

# Bioconductor packages
#remotes::install_github('chris-mcginnis-ucsf/DoubletFinder') #Install DoubletFinder and depends
library(batchelor)
library(edgeR)
library(scater)
library(scran)
library(scDblFinder)
# Set seed since this pipeline includes stochastic algorithms
set.seed(1)
# Set default working directory
#data_dir <- "C:/Users/k_cam/Google Drive/Placenta_Cell_Types/RNA/placenta_rna/data/"
data_dir <- paste0(here("data"), "/")
#results_dir <- "C:/Users/k_cam/Google Drive/Placenta_Cell_Types/RNA/placenta_rna/results/seurat/"
results_dir <- paste0(here("results", "seurat"), "/")
```
# Single-cell RNA-seq analysis of KC and TNL Pique-Regi placental villous tissue samples
### Important References
Minimal QC recommended: Current best practices in single‐cell RNA‐seq analysis: a tutorial
Orchestrating Single-cell Analysis
### Compiling Issues
This document had trouble compiling via knitr and pandoc. The following thread fixed the issue (https://github.com/rstudio/rstudio/issues/3661).

## Load Data
```{r}
seu <- readRDS(paste0(data_dir, "fetal_assigned_res.0.3_2020-12-16.rda"))
```

```{r}
DimPlot(seu, group.by = 'seurat_clusters', label = T, pt.size = .25)
```
Split into maternal versus fetal subsets.
```{r}
split <- SplitObject(seu, split.by = 'fetal')
rm(seu) # Clean memory
fetal <- split$Fetal
maternal <- split$Maternal
rm(split) # Clean memory
```

```{r}
DimPlot(fetal) + ggtitle("Fetal")
DimPlot(maternal) + ggtitle("Maternal")
```
Custom function to run fastMNN pipe with 
```{r}
pipe.fast.mnn <- function(seu, batch) {
  seu <- RunFastMNN(object.list = SplitObject(seu, split.by = batch), verbose = T)
  #print(seu@tools$RunFastMNN@metadata$merge.info$lost.var)
  seu <- RunUMAP(seu, reduction = "mnn", dims = 1:30)
  seu <- FindNeighbors(seu, reduction = "mnn", dims = 1:30)
  seu <- FindClusters(seu, res = 0.2)
  return(seu)
}
```

```{r}
maternal <- pipe.fast.mnn(maternal, "biorep")
DimPlot(maternal, group.by = "seurat_clusters", label = T, pt.size =.25)
```

```{r}
maternal.batch <- pipe.fast.mnn(maternal, "batch")
DimPlot(maternal.batch, group.by = "seurat_clusters", label = T, pt.size =.25)
```

```{r}
# Function that accepts Seurat  that has been processed up to the clustering step, clusters at desired resolutions (vector), adds cluster identities at different resolutions, and returns Seurat object with resolution cluster identities
Seurat_clustree <- function (seuratObject, resolutions) {
  
  for(hyperparameter in resolutions) {
    print(hyperparameter)
    prefix <- paste0("res.", hyperparameter)
    print(prefix)
    seuratObject <- FindClusters(object = seuratObject, resolution = hyperparameter)
    seuratObject <- AddMetaData(object = seuratObject, metadata = seuratObject$seurat_clusters, col.name = prefix)
  }
  return(seuratObject)
}

resolutions <- seq(from = 0.1, to = 0.8, by = .1)
maternal <- Seurat_clustree(maternal, resolutions)
```
Iterating over 0.2 to 0.7 clustering resolution, .3 looks stable. Additional divisions at .5
```{r clustree_graph}
clustree(maternal, prefix = "res.", node_colour = "sc3_stability") + theme(legend.position = "bottom") + guides(edge_alpha = F)
```

```{r}
DimPlot(maternal, group.by = 'res.0.1', label = T, pt.size = .25)
DimPlot(maternal, group.by = 'res.0.2', label = T, pt.size = .25)
DimPlot(maternal, group.by = 'res.0.3', label = T, pt.size = .25)
DimPlot(maternal, group.by = 'res.0.4', label = T, pt.size = .25)
DimPlot(maternal, group.by = 'res.0.5', label = T, pt.size = .25)
DimPlot(maternal, group.by = 'res.0.6', label = T, pt.size = .25)
DimPlot(maternal, group.by = 'res.0.7', label = T, pt.size = .25)
DimPlot(maternal, group.by = 'res.0.8', label = T, pt.size = .25)
# Set default to 0.7 for now
maternal <- FindClusters(maternal, res = 0.7)
```

```{r}
DimPlot(maternal, group.by = 'seurat_clusters', label = T, repel = T, pt.size = .25, label.size=  4) + NoLegend()
```
Cluster 6 is a plasma cell based on differential expression lookup - lots of immunoglobulins overxpressed compared to B cells.
```{r, eval = F}
b.cell.markers <- FindMarkers(maternal, ident.1 = "6", ident.2 = "2")
```

```{r, eval = F}
maternal.markers.res.0.4 <- FindAllMarkers(maternal)
```

Initial findDoubletClusters by cluster doesn't readily identify any clusters. Cluster 11 is a little low on DE.
```{r}
maternal.sce <- as.SingleCellExperiment(maternal)
dbl <- findDoubletClusters(maternal.sce, clusters = maternal.sce$seurat_clusters)
dbl.df <- as.data.frame(dbl)
#saveRDS(dbl.df, file = paste0(data_dir, "maternal.res.0.2.subcluster.labelled.initial.doublet.finder", Sys.Date(), ".rda"))
```

Try simulation approach to finding doublets. Clusters 10 and 11 stand out on VlnPlot.
```{r}
dbl.dens <- computeDoubletDensity(maternal.sce)
maternal$dbl.dens <- dbl.dens
FeaturePlot(maternal, features = "dbl.dens")
VlnPlot(maternal, features = "dbl.dens")
```
Key gene expression markers
```{r}
# Monocyte (and potential CD1C+ Dendritic Cell subset)
FeaturePlot(object = maternal,
            features = c("CD1C", "CD14"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = maternal,
            features = c("CD1C", "CLEC9A"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
# FCGR3A+ Monocytes
FeaturePlot(object = maternal,
            features = c("FCGR3A", "MS4A7"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
# B Cells
FeaturePlot(object = maternal,
            features = c("CD79A"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)

# pDC ref: "Single-cell RNA-seq reveals new types of human blood dendritic cells, monocytes and progenitors"
# Looks like pDC are actually off the tail of 
FeaturePlot(object = maternal,
            features = c("HLA-DRA"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = maternal,
            features = c("CLEC4C", "IL3RA"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = maternal,
            features = c("GZMB", "SERPINF1"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = maternal,
            features = c("IL3RA"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
# DC 1, bizarrely, no CLEC9A expression, which is supposed to clearly delineate a THBD+ subset, but shows up in the small cluster thought to be pDC
FeaturePlot(object = maternal,
            features = c("THBD", "CLEC9A"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
# RBCs
FeaturePlot(object = maternal,
            features = c("HBB", "HBG2"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
```

```{r}
t.cells <- subset(maternal, idents = c("2", "3", "4", "6", "7", "9", "10"))
t.cells <- pipe.fast.mnn(t.cells, batch = "biorep")
```

0.4 looks stable, another split at 0.7
```{r}
t.cells <- Seurat_clustree(t.cells, seq(0.1, 0.8, by = 0.1))
clustree(t.cells, prefix = "res.", node_colour = "sc3_stability") + theme(legend.position = "bottom") + guides(edge_alpha = F)
```

```{r}
DimPlot(t.cells, group.by = 'res.0.1', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.2', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.3', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.4', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.5', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.6', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.7', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.8', label = T, pt.size = .25)
```

Clusters 6 and 7, the same clusters from above look like doublet clusters
```{r}
FeaturePlot(t.cells, features = "dbl.dens", max.cutoff = 7)
VlnPlot(t.cells, features = "dbl.dens")
```

Cluster 5 is of low quality relative to the rest and DEX overrepresents ribosomal, mitochondrial, and stress response genes. Will drop that along with the doublet clusters 6 and 7.
```{r}
VlnPlot(t.cells, features = "subsets_Mito_percent")
```


```{r}
t.cells <- FindClusters(t.cells, resolution = 0.8)
doublets.tcells <- WhichCells(t.cells, ident = c("5", "6", "7"))
t.cells <- subset(t.cells, cells = doublets.tcells, invert = T)
DimPlot(t.cells, group.by = 'seurat_clusters', label = T, pt.size = .25)
```

```{r}
t.cells <- pipe.fast.mnn(t.cells, "biorep")
DimPlot(t.cells)
```

Resolution 0.2 avoid the complexity of CD8+ Cytotoxic cells.
```{r}
t.cells <- Seurat_clustree(t.cells, seq(0.1, 0.8, by = 0.1))
clustree(t.cells, prefix = "res.", node_colour = "sc3_stability") + theme(legend.position = "bottom") + guides(edge_alpha = F)
```

```{r}
DimPlot(t.cells, group.by = 'res.0.1', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.2', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.3', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.4', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.5', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.6', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.7', label = T, pt.size = .25)
DimPlot(t.cells, group.by = 'res.0.8', label = T, pt.size = .25)
```

Set default to 0.4
```{r}
t.cells <- FindClusters(t.cells, resolution = 0.4)
```

```{r, eval = F}
t.cells.markers <- FindAllMarkers(t.cells)
```

Differentiating CD4 from CD8 T cells
https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html Seurat cell type assignment
```{r}
# Naive CD4+
FeaturePlot(object = t.cells,
            features = c("IL7R", "CCR7"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = t.cells,
            features = c("CD4", "CCR7"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = t.cells,
            features = c("CD8A", "CCR7"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
# T Cell naive markers, not really used
FeaturePlot(object = fetal,
            features = c("SELL", "LRRN3"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
# Memory CD4+
FeaturePlot(object = t.cells,
            features = c("IL7R", "S100A4"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
# T regs do not come out clearly, not used
FeaturePlot(object = t.cells,
            features = c("IL2RA", "FOXP3"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
# TRM/TEM markers found in some CD8+ T cells, not used
FeaturePlot(object = t.cells,
            features = c("CXCR6"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = t.cells,
            features = c("ITGA1"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
# NCAM1 = CD56, NK specific marker, but can also be expressed by T cells; KLRB1 is NK1.1 (CD161)
FeaturePlot(object = t.cells,
            features = c("NCAM1", "KLRB1"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
# Expressed in NKs, "NK cell–intrinsic FcεRIγ limits CD8+ T-cell expansion and thereby turns an acute into a chronic viral infection"
FeaturePlot(object = t.cells,
            features = c("FCER1G"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)


# NK subtypes GZMB vs. GZMK
FeaturePlot(object = t.cells,
            features = c("GZMB", "GZMK", "CD3D", "KLRB1", "NKG7", "FGFBP2"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = t.cells,
            features = c("CD8A", "KLRB1"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = t.cells,
            features = c("CD3E", "KLRB1"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = t.cells,
            features = c("CD3D", "NCAM1"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
FeaturePlot(object = t.cells,
            features = c("CD3D", "KLRB1"),
            cols = c("lightgrey", "blue"),
            pt.size = .1)
```

"Single-cell transcriptomic landscape of nucleated cells in umbilical cord blood" KLRB1+ in the T cell CD8+ Cytotoxic is different from the publication. Otherwise, it seems that cluster 5 is NKT due to CD3E and NK markers and clusters 4 vs. 6 distinguish NK subtypes GZMB+ vs GZMK+.
```{r}
VlnPlot(t.cells, features = "CD3E")
VlnPlot(t.cells, features = "CD4")
VlnPlot(t.cells, features = "CD8A")
VlnPlot(t.cells, features = "KLRB1")
VlnPlot(t.cells, features = c("GZMB", "GZMK"))
VlnPlot(t.cells, features = "CD34")
```


```{r}
t.cells <- RenameIdents(t.cells,
                      "0" = "CD8+ Cytotoxic T Cells",
                      "1" = "Naive CD8+ T Cells",
                      "2" = "Natural Killer Cells",
                      "3" = "Naive CD4+ T Cells")
DimPlot(t.cells, group.by = 'ident', label = T, repel = T, pt.size = .25) + NoLegend()
# saveRDS(t.cells, file = paste0(data_dir, "t_cells_subcluster", Sys.Date(), ".rda"))
```

```{r}
doublet.dens.plot <- FeaturePlot(maternal, features = "dbl.dens")
```

```{r}
#b.cell.doublets <- CellSelector(doublet.dens.plot)
```

```{r}
cluster.doublet <- WhichCells(maternal, idents = "8")
```

```{r, eval = F}
maternal.cells.to.drop <- c(cluster.doublet, doublets.tcells, b.cell.doublets)
```

Overwrite coarse labels with T cell subcluster labels
```{r}
maternal <- SetIdent(maternal, cells = WhichCells(t.cells), value = Idents(t.cells))
```

```{r, eval = F}
maternal <- subset(maternal, cells = maternal.cells.to.drop, invert = T)
```

Rename maternal clusters
```{r}
maternal <- RenameIdents(maternal,
                      "2" = "B Cells",
                      "6" = "Plasma Cells",
                      "4" = "FCGR3A+ Monocytes", 
                      "1" = "CD14+ Monocytes")
# Stash identities for later use if needed
maternal <- StashIdent(maternal, "maternal.res.0.2.cell.type.labelled")
```

```{r}
doublet.density.maternal <- FeaturePlot(maternal, features = "dbl.dens")
#tcell.leftovers <- CellSelector(doublet.density.maternal)
#mono.tcell.doublets <- CellSelector(doublet.density.maternal)
#bcell.tcell.doubets <- CellSelector(doublet.density.maternal)
```

```{r, eval = F}
additional.doublets.to.drop <- c(tcell.leftovers, mono.tcell.doublets, bcell.tcell.doubets)
maternal <- subset(maternal, cells = additional.doublets.to.drop, invert = T)
```

Overwrite coarse cell type labels with subclustering labels.
```{r}
maternal <- StashIdent(maternal, "maternal.res.0.4.subclustered.labelled")
```

```{r}
#saveRDS(maternal, paste0(data_dir,"cleaned_maternal_seurat", Sys.Date(), ".rda"))
# Read cleaned data in
maternal <- readRDS(paste0(data_dir, "cleaned_maternal_seurat2020-12-22.rda"))
```

```{r, eval = F}
all.dropped.cells <- c(maternal.cells.to.drop, additional.doublets.to.drop)
#saveRDS(all.dropped.cells, paste0(data_dir, "maternal_dropped_cells_", Sys.Date(), ".rda"))
```

Rename cell type clusters with Maternal label
```{r}
idents <- Idents(maternal)
cell.type <- paste0("Maternal ", idents)
Idents(maternal) <- cell.type
maternal <- StashIdent(maternal, "cell.type")
#saveRDS(named.maternal, paste0(data_dir, "named_cleaned_maternal_seurat_2020-12-22.rda"))
named.maternal <- readRDS(paste0(data_dir, "named_cleaned_maternal_seurat_2020-12-22.rda"))
```

```{r}
DimPlot(named.maternal, label = T, repel = T, pt.size = .25, label.size = 4) + NoLegend()
```

```{r}
DimPlot(maternal, label = T, repel = T, pt.size = .25, label.size = 4) + NoLegend()
```

```{r}
#ggsave(filename = paste0(data_dir, "maternal_cleaned_", Sys.Date(), ".png"), device = "png")
```

## Combine fetal and maternal datasets
```{r}
# Read in and merge fetal and maternal datasets
fetal <- readRDS(paste0(data_dir,"named_cleaned_fetal_seurat_2020-12-22.rda"))
maternal <- readRDS(paste0(data_dir,"named_cleaned_maternal_seurat_2020-12-22.rda"))
seu <- merge(fetal, maternal)
rm(fetal)    # Memory clean-up
rm(maternal) # Memory clean-up
umap.fast.mnn <- function(seu, batch) {
  seu <- RunFastMNN(object.list = SplitObject(seu, split.by = batch), verbose = T)
  #print(seu@tools$RunFastMNN@metadata$merge.info$lost.var)
  seu <- RunUMAP(seu, reduction = "mnn", dims = 1:30)
  return(seu)
}
seu <- umap.fast.mnn(seu, "biorep")
```

```{r}
DimPlot(seu, label = T, repel = T, pt.size = .25, label.size = 2.5) + NoLegend()
#ggsave(paste0(results_dir, "Maternal_Fetal_Combined_UMAP_", Sys.Date(), ".png"))
#saveRDS(seu, paste0(data_dir,"cleaned_combined_seurat_2020-12-22.rda"))
seu <- readRDS(paste0(data_dir,"cleaned_combined_seurat_2020-12-22.rda"))
```

QC metrics by cluster
```{r}
VlnPlot(seu, features = "subsets_Mito_percent", pt.size = .1) + NoLegend() + 
  theme(axis.text.x = element_text(size = 8)
        ) +
  ggtitle("Percentage of Reads Mapping to Mitochondrial Genes by Cluster")
ggsave(paste0(results_dir, "qc_cluster_mito_", Sys.Date(), ".png"))

VlnPlot(seu, features = "nCount_RNA", pt.size = .1) + NoLegend() + 
  theme(axis.text.x = element_text(size = 8)
        ) +
  ggtitle("Total RNA Molecules Per Cell by Cluster")
ggsave(paste0(results_dir, "qc_cluster_total_RNA_", Sys.Date(), ".png"))

VlnPlot(seu, features = "nFeature_RNA", pt.size = .1) + NoLegend() + 
  theme(axis.text.x = element_text(size = 8)
        ) +
  ggtitle("Detected Genes Per Cell by Cluster")
#ggsave(paste0(results_dir, "qc_cluster_total_genes_", Sys.Date(), ".png"))
```

```{r}
list <- SplitObject(seu, split.by = "fetal")
DimPlot(list$Fetal, label = T, repel = T, pt.size = .25, label.size = 2.5) + NoLegend()
#ggsave(paste0(results_dir, "Fetal_Combined_UMAP_", Sys.Date(), ".png"))
DimPlot(list$Maternal, label = T, repel = T, pt.size = .25, label.size = 2.5) + NoLegend()
#ggsave(paste0(results_dir, "Maternal_Combined_UMAP_", Sys.Date(), ".png"))
```
