---
title: "seurat_dex"
author: "Kyle Campbell"
date: "1/4/2021"
output: html_document
---

```{r setup, include=FALSE}
library(here)
library(ggrepel)
library(gprofiler2)
library(openxlsx)
library(psych)
library(Seurat)
library(tidyverse)

knitr::opts_chunk$set(echo = TRUE)
```

## Imporant References
"Bias, robustness and scalability in single-cell differential expression analysis"
"Current best practices in single‐cell RNA‐seq analysis: a tutorial"
Straightforward approach is to simply use the default Wilcoxon Rank Sum test in Seurat or MAST, which takes into dropout and provides a GLM-like model to adjust for covariates (limma-voom also works)

Load cleaned Seurat data
```{r, eval = F}
seu <- readRDS(here("data", "analytic", "2022-05-01_all_merged_pruned_batch_mnn.rda"))
```

```{r, eval = F}
#all.markers <- FindAllMarkers(seu)
#saveRDS(object = all.markers, file = here("data", "all_markers_cleaned_combined_tsang_seurat_2022-05-01.rda"))
```

```{r, eval = F}
DimPlot(seu, repel = T, label = T) + NoLegend()
```


Load dex results
```{r}
# Initial Tsang analysis
#markers <- readRDS(here("data", "all_markers_cleaned_combined_tsang_seurat_2022-03-31.rda"))
# Analytic final Tsang resubmit analysis
markers <- readRDS(file = here("data", "all_markers_cleaned_combined_tsang_seurat_2022-05-01.rda"))
```

```{r}
markers %>% filter(cluster %in% "Fetal Naive CD8+ T Cells")
markers %>% filter(cluster %in% "Fetal CD8+ Activated T Cells")
markers %>% filter(cluster %in% "Fetal Naive CD4+ T Cells")
```

```{r}
markers %>% group_by(cluster) %>% tally()
```

Set p_val_adj_threshold, q < .05, Bonferroni-adjusted in Seurat
```{r}
p_val_adj_threshold <- .05
```

Save marker gene results
```{r}
# All results
#write.csv(markers, here("results", "seurat", "all_markers_cleaned_combined_seurat_2020-12-22.csv"))
```

```{r}
colnames(markers)[2] <- "avg_logFC"

all.cluster.markers <- markers %>%
  # Only upregulated genes
  filter(avg_logFC > 0) %>%
  filter(p_val_adj < p_val_adj_threshold) %>%
  group_by(cluster) %>%
  arrange(cluster) %>%
  #arrange(desc(avg_logFC)) %>%
  #slice_head(n=10) %>%  # Formatting after this line
  mutate(p_val = if_else(p_val == 0, 2.23e-308, p_val)) %>%
  mutate(p_val_adj = if_else(p_val_adj == 0, 2.23e-308, p_val_adj)) %>%
  dplyr::select(cluster, gene, avg_logFC, pct.1, pct.2, p_val, p_val_adj) %>%
  mutate(p_val = format(p_val, scientific = TRUE, digits = 3)) %>%
  mutate(p_val_adj = format(p_val_adj, scientific = TRUE, digits = 3)) %>%
  mutate(avg_logFC = format(avg_logFC, digits = 3)) %>%
  mutate('p-value' = if_else(p_val == 2.23e-308, "< 2.23E-308", as.character(p_val))) %>%
  mutate('adjusted p-value' = if_else(p_val_adj == 2.23e-308, "< 2.23E-308", as.character(p_val_adj))) %>%
  dplyr::select(-c(p_val, p_val_adj)) %>%
  # Somewhere in this code block, avg_logFC is converted to a non-numeric data type
  mutate(avg_logFC = as.numeric(avg_logFC))

# Label column headers
all.cluster.markers.colnames <- c("Cell type cluster", "Gene", "Average log2 Fold-change", "Percentage cluster cells expressing gene", "Percentage other cells expressing gene", "P-value", "Bonferroni-adjusted p-value")
colnames(all.cluster.markers) <- all.cluster.markers.colnames
cell.type.levels <- c("Fetal B Cells", "Fetal CD14+ Monocytes", "Fetal CD8+ Activated T Cells", "Fetal Cytotrophoblasts", "Fetal Endothelial Cells", "Fetal Extravillous Trophoblasts", "Fetal Fibroblasts", "Fetal GZMB+ Natural Killer", "Fetal GZMK+ Natural Killer", "Fetal Hofbauer Cells", "Fetal Memory CD4+ T Cells", "Fetal Mesenchymal Stem Cells", "Fetal Naive CD4+ T Cells", "Fetal Naive CD8+ T Cells", "Fetal Natural Killer T Cells", "Fetal Nucleated Red Blood Cells", "Fetal Plasmacytoid Dendritic Cells", "Fetal Proliferative Cytotrophoblasts", "Fetal Syncytiotrophoblast", "Maternal B Cells", "Maternal CD14+ Monocytes", "Maternal CD8+ Activated T Cells", "Maternal FCGR3A+ Monocytes", "Maternal Naive CD4+ T Cells", "Maternal Naive CD8+ T Cells", "Maternal Natural Killer Cells", "Maternal Plasma Cells")
all.cluster.markers$`Cell type cluster` <- factor(all.cluster.markers$`Cell type cluster`, cell.type.levels)

cluster.markers <- all.cluster.markers %>% group_split()
names(cluster.markers) <- cell.type.levels
names(cluster.markers)[names(cluster.markers) == "Fetal Plasmacytoid Dendritic Cells"] <- "Fetal Plasmacytoid Dendritic"
names(cluster.markers)[names(cluster.markers) == "Fetal Proliferative Cytotrophoblasts"] <- "Fetal Prolif. Cytotrophoblasts"

# Write to multiple sheets
write.xlsx(as.list(cluster.markers), file = here("results", "seurat", paste0(Sys.Date(), "_findallmakers_res.xlsx")))

# Write to .csv
#write.csv(all.cluster.markers, here("results", "seurat", paste0("all_markers_cleaned_combined_seurat_", Sys.Date(), ".csv")), row.names = FALSE)
```

Number differentially expressed genes per cluster, q < .05
```{r}
num <- markers %>%
  group_by(cluster) %>%
  tally()
num

num.de <- markers %>%
  filter(p_val_adj < p_val_adj_threshold) %>%
  group_by(cluster) %>%
  tally()
num.de

pct.sig <- num.de/(num)
```

Number overexpressed genes per cluster
```{r}
num.up <- markers %>%
  filter(p_val_adj < p_val_adj_threshold) %>%
  filter(avg_logFC > 0) %>%
  group_by(cluster) %>%
  tally()
num.up
describe(num.up)
```

```{r}
# Function to select the gene column in each sublist
marker_Volcano <- function(res, names, num.to.plot, genes.of.interest, threshold, output_dir) {
  
  # Indicator variable as to whether significant
  res$sig <- (res$p_val_adj < threshold)
  
  # Get the top downregulated hits
  downreg <- res %>%
    filter(p_val_adj < threshold) %>%
    filter(avg_logFC < 0) %>%
    arrange(avg_logFC)
  #Get the top upregulated hits.
  upreg <- res %>%
    filter(p_val_adj < threshold) %>%
    filter(avg_logFC > 0) %>%
    arrange(desc(avg_logFC))
  
  top.upreg <- upreg$gene[1:num.to.plot]
  top.downreg <- downreg$gene[1:num.to.plot]
  genes.to.plot <- c(top.upreg, top.downreg, genes.of.interest)
  res$label <- F
  res$label[(res$gene %in% genes.to.plot)] <- T
  
  p <- ggplot(res) +
    geom_point(aes(
      x = avg_logFC,
      y = -log10(p_val_adj),
      colour = sig
    )) +
    geom_text_repel(aes(
      x = avg_logFC,
      y = -log10(p_val_adj),
      label = ifelse(label, gene, "")
    )) +
    geom_hline(yintercept = -log10(threshold),
               linetype = "dotted") +
    ggtitle(names) +
    xlab("log2 fold change") +
    ylab("-log10 adjusted p-value") +
    theme(
      legend.position = "none",
      plot.title = element_text(size = rel(1.5), hjust = 0.5),
      axis.title = element_text(size = rel(1.25))
    )
  print(p)
  ggsave(plot = p, filename = paste0(output_dir, names, "_", Sys.Date(), ".png"), device = "png")
}
```

Select differentially genes per cluster
```{r}
# Group results by cluster and split by cluster
top <- markers %>%
  group_by(cluster) %>%
  group_split()

# Get cluster names
cluster.names <- levels(markers$cluster)

# Apply function over lists using map
map2(.x = top,                  # List of tibbles containing dex results
     .y = cluster.names,        # List of names, used to label plots
     .f = marker_Volcano,       # Function to plot Volcano plots
     num.to.plot = 10,          # Top x upregulated and downregulated gene labels to plot, ranked by logFC
     genes.of.interest = NULL,  # Vector of genes to manually label 
     threshold = .05,           # Bonferroni-adjusted p_val cutoff
     output_dir = here("results", "seurat", "markerVolcanos/") # Save directory
     )
```

## g:Profiler multiquery comparing cell types

Select overexpressed genes per cluster, ranked by avg_logFC
```{r}
# Filter genes by cluster and split by cluster
top <- markers %>%
  filter(avg_logFC > 0) %>%
  filter(p_val_adj < p_val_adj_threshold) %>%
  group_by(cluster) %>%
  arrange(desc(avg_logFC)) %>%
  dplyr::select(cluster, gene) %>%
  group_split()

# Get cluster names
cluster.names <- levels(markers$cluster)

# Function to select the gene column in each sublist
get_genes <- function(data) {
  data %>%
    dplyr::select(gene)
}

# Apply function over lists using map
gene.lists <- map(top, get_genes)
# Remove a level hierarchy from the list
list.gProfiler <- gene.lists %>% flatten()
# Apply Cluster names
names(list.gProfiler) <- cluster.names
```

Perform the enrichment test using all ranked gene lists, excluding automatically generated terms, using only annotated genes as the gene universe background.
```{r}
gostres <- readRDS(here("results", "seurat", "seurat_all_markers_gost_results_2022-05-08.rda"))
#gostres <- readRDS(here("results", "seurat", "seurat_all_markers_gost_results_2021-01-05.rda"))
#gostres <- gost(query = list.gProfiler,
#                organism = "hsapiens", ordered_query = TRUE, 
#                multi_query = FALSE, significant = TRUE, exclude_iea = TRUE, 
#                measure_underrepresentation = FALSE, evcodes = FALSE, 
#                user_threshold = 0.05, correction_method = "g_SCS", 
#                domain_scope = "annotated", custom_bg = NULL, 
#                numeric_ns = "", sources = NULL, as_short_link = FALSE)
#saveRDS(gostres, here("results", "seurat", paste0("seurat_all_markers_gost_results_", Sys.Date(), ".rda")))
```

Exporting results; unsurprisingly, "Fetal Plasmacytoid Dendritic Cells" had no significant enrichment pathways and Fetal Nucleated Red Blood cells had no enriched pathways when term sizes are limited
```{r}
# Drop parent column list for .csv save of all results
gostres.table <- gostres$result
gostres.table$parents <- NULL

gostres.table <- 
  gostres.table %>%
  relocate(term_name, .after = significant)
#write.csv(gostres.table, here("results", "seurat", paste0("seurat_all_markers_gost_results_table_", Sys.Date(), ".csv")), row.names = FALSE)

# Limit term sizes to <1000
gostres.small <- gostres
#gostres.small$result <- gostres$result %>% filter(term_size < 1000)
gostres.small.table$query %>% factor %>% levels

# Drop parent column list for .csv save of filtered GO:BP results
gostres.small.table <- gostres.small$result
gostres.small.table$parents <- NULL

gostres.small.table <- gostres.small.table %>%
  dplyr::filter(source == "GO:BP") %>%
  dplyr::relocate(term_name, .after=query) %>%
  dplyr::select(!significant) %>%
  group_by(query)

#gostres.small.table.top.10 <- gostres.small.table %>%
  #filter(term_size < 1500) %>%
  #arrange(p_value) %>%
  #group_by(query) #%>%
  #slice_head(n=20)

# View a couple cell types individually
#gostres.small.table %>% filter(query %in% c("Fetal Cytotrophoblasts", "Fetal Proliferative Cytotrophoblasts")) %>% View
#gostres.small.table %>% filter(query == "Fetal Hofbauer Cells") %>% View
  
#write.csv(gostres.small.table, here("results", "seurat", "seurat_all_markers_gost_results_table_top10_BP_filtered_", Sys.Date(), ".csv"))
```

Loop over all single-cell clusters, print Manhattan plots, and save.
```{r}
cell.type.levels.loop <- cell.type.levels[-c(17)]
for(idx in 1:length(cell.type.levels.loop)) {
  g <- gostres.small
  g$result <- gostres.small$result %>% filter(query %in% cell.type.levels.loop[idx])
  print(gostplot(g, capped = FALSE, interactive = FALSE))
  #ggsave(here("data", "gProfiler", "sc_clusters", paste0(cluster.names[idx], "_gost_plot_200_term_size_", Sys.Date(), ".png")), device = "png")
}
```

```{r}
cluster.ontologies <- gostres.small.table %>%
  dplyr::select(query, term_name, p_value, term_size, query_size, intersection_size, term_id)
colnames(cluster.ontologies) <- c("Cell type cluster", "Term name", "Adjusted P-value", "Term size", "Query size", "Intersection size", "Term ID")
  cluster.ontologies <- cluster.ontologies %>%
  group_by(`Cell type cluster`) %>%
  group_split()
names(cluster.ontologies) <- cell.type.levels.loop

#names(cluster.ontologies)[names(cluster.ontologies) == "Fetal Plasmacytoid Dendritic Cells"] <- "Fetal Plasmacytoid Dendritic"
names(cluster.ontologies)[names(cluster.ontologies) == "Fetal Proliferative Cytotrophoblasts"] <- "Fetal Prolif. Cytotrophoblasts"

# Write to multiple sheets
#write.xlsx(as.list(cluster.ontologies), file = here("results", "seurat", paste0(Sys.Date(), "_findallmarkers_gprofiler2_ontologies_res.xlsx")))
```

