---
title: "scRNA-seq import and QC of placental villous tissue"
author: "Kyle Campbell"
date: "2/14/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::knit_engines$set(python = reticulate::eng_python)

#install.packages("package_name",  repos="http://cran.us.r-project.org") # Only this mirror worked for multiple packages
library(here)
library(knitr)
library(tidyverse)
library(devtools)
library(Seurat)
#remotes::install_github('satijalab/seurat-wrappers')
library(SeuratWrappers)
library(cowplot)
library(rgl)
library(EnvStats)
library(reticulate)
library(clustree)
library(Matrix)

# Bioconductor packages
library(batchelor)
library(edgeR)
library(scater)
# Set seed since this pipeline includes stochastic algorithms
set.seed(1)
```
# Single-cell RNA-seq analysis of KC and TNL Pique-Regi placental villous tissue samples
### Important References
Minimal QC recommended: Current best practices in single‐cell RNA‐seq analysis: a tutorial
Orchestrating Single-cell Analysis
### Compiling Issues
This document had trouble compiling via knitr and pandoc. The following thread fixed the issue (https://github.com/rstudio/rstudio/issues/3661).

## Load QC'd Data
```{r}
seu <- readRDS(here("data", "kcprfinal_tsang_qc_merged_2022-02-15.rda"))
# Subset to normal samples only
#seu <- subset(x = seu,
#              subset = biorep  %in% c("tsang_pe1", "tsang_pe2", "tsang_pe3", "tsang_pe4"),
#              invert = T)
#seu.analyzed <- readRDS(here("data", "cleaned_combined_seurat_2020-12-22.rda"))
```

## PCA and clustering

```{r Normalization_and_Feature_Selection}
# Should normalize and find variable features in each dataset separately
seu <- NormalizeData(object = seu)
seu <- FindVariableFeatures(object = seu)
#all.genes <- rownames(seu)
# Increase memory limit for the following commands
memory.limit(size=56000)
seu <- ScaleData(seu)#, features = all.genes)
seu <- RunPCA(seu, features = VariableFeatures(object = seu))
```

```{r Dataset_Dimensionality}
ElbowPlot(object = seu, ndims = 50, reduction = "pca")
```

```{r}
DimPlot(seu, reduction = "pca", group.by = 'orig.ident', pt.size = 0.25)
```

```{r}
seu <- RunTSNE(seu, dims.use = 1:50, do.fast = T, dim.embed = 2)
seu <- FindNeighbors(object=seu, dims=1:50, reduction='pca')
seu <- RunUMAP(object = seu, dims = 1:50)
seu <- FindClusters(object = seu, resolution = 0.2) #default is 0.8; 0.2 looks stable in the kc dataset
```


```{r}
table(seu@meta.data$seurat_clusters)
table(seu@meta.data$seurat_clusters,seu$orig.ident)

# tSNE plot by sample cluster
DimPlot(object = seu,pt.size=0.7,reduction='tsne',group.by='seurat_clusters', label = T)
DimPlot(object = seu,pt.size=0.7,reduction='tsne',group.by='orig.ident')
DimPlot(object = seu,pt.size=0.01,reduction='tsne',group.by='batch')
DimPlot(object = seu,pt.size=0.01,reduction='tsne',group.by='final_clusters', label = T, repel = T) + NoLegend()

# UMAP plot by sample cluster
DimPlot(object = seu,pt.size=0.7,reduction='umap',group.by='ident', label = T)
DimPlot(object = seu,pt.size=0.01,reduction='umap',group.by='orig.ident')
DimPlot(object = seu,pt.size=0.01, reduction='umap',group.by='batch')
DimPlot(object = seu,pt.size=0.01,reduction='umap',group.by='final_clusters', label = T, repel = T) + NoLegend()
```

## Assigning maternal/fetal status and sex
Plotting several sex-specific transcripts, it is clear that KC42, PR478, and PR484 are all male fetuses. Markers come from "Robust and tissue-independent gender-specific transcript biomarkers.", published using peripheral blood as sample source (RPS4Y1, EIF1AY, DDX3Y, KDM5D and XIST; XIST being upregulated in females compared to males). \n
Similarly, tsang_n1, _n2, pe1, and pe2 are all male
```{r}
VlnPlot(object= seu,
        features = c("XIST", "RPS4Y1", "EIF1AY", "DDX3Y", "KDM5D"),
        group.by = 'orig.ident')
```

Grouping by biological replicate and then splitting by freemux assignment (0/1), plotting the three most robustly expressed genes in the dataset, we clearly see that 0 (the more abundant of the two individuals) corresponds to the fetus in PR478 and PR484, and 1 corresponds to the fetus in KC42 (the less abundant of the two individuals). It can be inferred this pattern holds for PR481 and KC40, but will verify with placental gene expression below. In the male Tsang samples, it is clear that 0 is the fetus.
```{r}
VlnPlot(object= seu,
        features = c("XIST", "RPS4Y1", "EIF1AY"),
        group.by = 'biorep',
        split.by = 'freemuxlet.assignments',
        split.plot = T)
```
Yes, the pattern holds, clearly visible in KRT7 in PR481, 0 is still the fetus in all the PR biological replicates. In other other genes, it looks like kc40.1 freemux 0 is the fetus. 
```{r}
VlnPlot(object= seu,
        features = c("KRT7"),
        group.by = 'orig.ident',
        split.by = 'freemuxlet.assignments',
        split.plot = T)
VlnPlot(object= seu,
        features = c("COL1A1"),
        group.by = 'orig.ident',
        split.by = 'freemuxlet.assignments',
        split.plot = T)
VlnPlot(object= seu,
        features = c("CD163"),
        group.by = 'orig.ident',
        split.by = 'freemuxlet.assignments',
        split.plot = T)
VlnPlot(object= seu,
        features = c("KDR"),
        group.by = 'orig.ident',
        split.by = 'freemuxlet.assignments',
        split.plot = T)
```

```{r}
fetal.genes <- AverageExpression(seu, features = c("XIST", "RPS4Y1", "EIF1AY", "KRT7", "CD163", "COL1A1", "KDR", "PSG2", "PSG4", "CD79A", "CD8A"), group.by = c("orig.ident", "freemuxlet.assignments"))
fetal.genes <- t(fetal.genes$RNA)
```


```{r}
VlnPlot(object= seu,
        features = c("total"),
        group.by = 'orig.ident',
        split.by = 'freemuxlet.assignments',
        split.plot = T)
VlnPlot(object= seu,
        features = c("detected"),
        group.by = 'orig.ident',
        split.by = 'freemuxlet.assignments',
        split.plot = T)
VlnPlot(object= seu,
        features = c("subsets_Mito_percent"),
        group.by = 'orig.ident',
        split.by = 'freemuxlet.assignments',
        split.plot = T)
```

From above, KC42, freemux 1 is the fetus. It looks like in almost every KC sample except kc.40.1, freemux 1 is the fetus.
Based on these findings, PR freemux 0 corresponds to fetus, KC freemux 0 corresponds to maternal except for KC40.1, Tsang freemux 0 corresponds to fetal except for n3c and n4c
```{r}
batch <- SplitObject(seu, split.by = "batch")
kc <- batch$kc
pr <- batch$pr
tsang <- batch$tsang
rm(batch)
DimPlot(kc, group.by = 'freemuxlet.assignments', split.by = 'orig.ident', pt.size = .25)
DimPlot(pr, group.by = 'freemuxlet.assignments', split.by = 'orig.ident', pt.size = .25)
DimPlot(tsang, group.by = 'freemuxlet.assignments', split.by = 'orig.ident', pt.size = .25)
```

```{r}
rm(kc)
rm(pr)
rm(tsang)
```
Add maternal/fetal assignment and sex
```{r}
meta <- seu@meta.data
barcode <- rownames(meta)
df <- as_tibble(cbind(barcode, meta))
meta.new <- df %>%
  mutate(fetal = case_when(
    grepl("pr", orig.ident) ~  # for all PR samples, there are more fetal than maternal cells
      if_else(freemuxlet.assignments == 0,
              TRUE,
              FALSE),
    grepl("kc", orig.ident) ~     
             if_else(
               orig.ident == "kc.40.1", # Of the KC samples, only kc.40.1 has more fetal than maternal cells
               if_else(freemuxlet.assignments == 0,
                       TRUE,
                       FALSE),
               if_else(freemuxlet.assignments == 1,
                       TRUE,
                       FALSE)
             ),
    grepl("tsang", orig.ident) ~
                   if_else(
               orig.ident %in% c("tsang_n3c", "tsang_n4c"), # Of the Tsang samples, only n3c and n4c have more maternal than fetal cells
               if_else(freemuxlet.assignments == 1,
                       TRUE,
                       FALSE),
               if_else(freemuxlet.assignments == 0,
                       TRUE,
                       FALSE)
             )
  )) %>%
  mutate(sex =
           if_else(
             fetal == F,
             "female",
             if_else(
               biorep %in% c("kc.42", "pr.481", "pr.484", "tsang_n1", "tsang_n2", "tsang_pe1", "tsang_pe2"),
               "male",
               "female")
           )
  )
meta.new$fetal <- factor(meta.new$fetal,
                         levels = c(FALSE, TRUE),
                         labels = c("Maternal", "Fetal"))
meta.new$sex <- factor(meta.new$sex,
                         levels = c("female", "male"),
                         labels = c("Female", "Male"))
```

```{r}
seu <- AddMetaData(object = seu, metadata = meta.new$fetal, col.name = 'fetal')
seu <- AddMetaData(object = seu, metadata = meta.new$sex, col.name = 'sex')
```

Double-check fetal and sex assignments. Y-linked RPS4Y1 looks normal, but XIST expression looks way too high in male Tsang samples. Possible miscoding of fetal cells
```{r}
#seu.fetal <- subset(seu, subset = fetal == "Fetal")
```

```{r}
AverageExpression(object = seu,
                  features = "RPS4Y1",
                  group.by = c("orig.ident", "freemuxlet.assignments"))$RNA %>% t %>% View
AverageExpression(object = seu,
                  features = "XIST",
                  group.by = c("orig.ident", "fetal"))$RNA %>% t %>% View
AverageExpression(object = seu,
                  features = "COL1A1",
                  group.by = c("orig.ident", "fetal"))$RNA %>% t %>% View
```

```{r}
batch <- SplitObject(seu, split.by = "batch")
kc <- batch$kc
pr <- batch$pr
tsang <- batch$tsang
rm(batch)
DimPlot(kc, group.by = 'fetal', split.by = 'orig.ident', pt.size = .25)
DimPlot(pr, group.by = 'fetal', split.by = 'orig.ident', pt.size = .25)
DimPlot(tsang, group.by = 'fetal', split.by = 'orig.ident', pt.size = .25)
```

```{r}
rm(kc)
rm(pr)
rm(tsang)
```

```{r}
DimPlot(seu, group.by = 'fetal', pt.size = .25)
DimPlot(seu, group.by = 'biorep', split.by = 'fetal', pt.size = .25)
DimPlot(seu, group.by = 'sex', pt.size = .25)
```

## MNN Correction and Initial t
Clustering
MNN batch correction pipeline, batch correct for biological replicate
```{r}
# Mutual Nearest Neighbor Reduction
# RunFastMNN, splitting by biological replicate (i.e., KC40 vs. KC42 vs. each tnl sample vs. each normal tsang sample)
seu <- RunFastMNN(object.list = SplitObject(seu, split.by = "biorep"))
DimPlot(object = seu,pt.size=0.7,reduction='mnn', group.by='orig.ident')
DimPlot(object = seu,pt.size=0.7,reduction='mnn', group.by='biorep')

# Re-run UMAP and cluster pipeline after MNN correction
seu <- RunUMAP(seu, reduction = "mnn", dims = 1:50)
seu <- FindNeighbors(seu, reduction = "mnn", dims = 1:50)
seu <- FindClusters(seu, resolution = 0.2)
```

```{r}
# 'biorep' is metadata describing biological replicates
DimPlot(seu, group.by = 'biorep')
# 'seurat_clusters' is metadata describing default clustering'
DimPlot(seu, group.by = 'seurat_clusters', label = T, pt.size = .25)
# 'SampleName' describes KC40A (1A), KC40B (1B), KC42A (2A), KC42B (2B)
DimPlot(seu, group.by = 'orig.ident')
DimPlot(seu, group.by = 'freemuxlet.assignments')
```

MNN batch correction pipeline, batch correct for each study
```{r}
# Mutual Nearest Neighbor Reduction
# RunFastMNN, splitting by biological replicate (i.e., KC40 vs. KC42 vs. each tnl sample vs. each normal tsang sample)
seu <- RunFastMNN(object.list = SplitObject(seu, split.by = "batch"))
DimPlot(object = seu,pt.size=0.7,reduction='mnn', group.by='orig.ident')
DimPlot(object = seu,pt.size=0.7,reduction='mnn', group.by='batch')

# Re-run UMAP and cluster pipeline after MNN correction
seu <- RunUMAP(seu, reduction = "mnn", dims = 1:50)
seu <- FindNeighbors(seu, reduction = "mnn", dims = 1:50)
seu <- FindClusters(seu, resolution = 0.2)
```

```{r}
DimPlot(object = seu,pt.size=0.7,reduction='mnn', group.by='orig.ident')
DimPlot(object = seu,pt.size=0.7,reduction='mnn', group.by='biorep')
DimPlot(object = seu,pt.size=0.7,reduction='mnn', group.by='batch')
DimPlot(object = seu,pt.size=0.7,reduction='mnn', group.by='seurat_clusters')
```


## Clustering optimization

```{r}
# Function that accepts Seurat  that has been processed up to the clustering step, clusters at desired resolutions (vector), adds cluster identities at different resolutions, and returns Seurat object with resolution cluster identities
Seurat_clustree <- function (seuratObject, resolutions) {
  
  for(hyperparameter in resolutions) {
    print(hyperparameter)
    prefix <- paste0("res.", hyperparameter)
    print(prefix)
    seuratObject <- FindClusters(object = seuratObject, resolution = hyperparameter)
    seuratObject <- AddMetaData(object = seuratObject, metadata = seuratObject$seurat_clusters, col.name = prefix)
  }
  return(seuratObject)
}

resolutions <- seq(from = 0.2, to = 0.7, by = .1)
seu <- Seurat_clustree(seu, resolutions)
```

Iterating over 0.2 to 0.7 clustering resolution, .3 looks stable. Additional divisions at .5
```{r clustree_graph}
clustree(seu, prefix = "res.", node_colour = "sc3_stability") + theme(legend.position = "bottom") + guides(edge_alpha = F)
```
res of 0.3 splits the large T-cell containing cluster into 4 groups instead of 2 and splits the monocyte group into 2 groups.
```{r}
# 'seurat_clusters' is metadata describing default clustering'
DimPlot(seu, group.by = 'res.0.3', label = T, pt.size = .25)
DimPlot(seu, group.by = 'res.0.2', label = T, pt.size = .25)
```

```{r}
# Reset default cluster resolution
seu <- FindClusters(seu, resolution = 0.3)
#saveRDS(seu, paste0(data_dir, "fetal_assigned_res.0.3_",Sys.Date(), ".rda"))
seu.res.0.3 <- FindAllMarkers(seu)
```