---
title: "Fetal Cell type assignment - correcting by batch (study source)"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::knit_engines$set(python = reticulate::eng_python)

#install.packages("package_name",  repos="http://cran.us.r-project.org") # Only this mirror worked for multiple packages
library(here)
library(knitr)
library(tidyverse)
library(devtools)
library(Seurat)
#remotes::install_github('satijalab/seurat-wrappers')
library(SeuratWrappers)
library(cowplot)
library(rgl)
library(EnvStats)
library(reticulate)
library(clustree)
library(Matrix)

# Bioconductor packages
#remotes::install_github('chris-mcginnis-ucsf/DoubletFinder') #Install DoubletFinder and depends
library(batchelor)
library(edgeR)
library(scater)
library(scDblFinder)
library(scran)
library(SingleR)
# Set seed since this pipeline includes stochastic algorithms
set.seed(1)

options(ggrepel.max.overlaps = Inf)
```
# Single-cell RNA-seq analysis of KC and TNL Pique-Regi placental villous tissue samples
### Important References
Minimal QC recommended: Current best practices in single‐cell RNA‐seq analysis: a tutorial
Orchestrating Single-cell Analysis
### Compiling Issues
This document had trouble compiling via knitr and pandoc. The following thread fixed the issue (https://github.com/rstudio/rstudio/issues/3661).

Custom function to run fastMNN pipe with 
```{r}
pipe.fast.mnn <- function(seu, batch) {
  seu <- RunFastMNN(object.list = SplitObject(seu, split.by = batch), verbose = T)
  seu@tools$RunFastMNN$merge.info$lost.var %>% print
  seu <- RunUMAP(seu, reduction = "mnn", dims = 1:30)
  seu <- FindNeighbors(seu, reduction = "mnn", dims = 1:30)
  seu <- FindClusters(seu, res = 0.2)
  return(seu)
}
```

## Load Normal Placenta Data
This data includes the original Tsang analysis QC thresholds merged with the original kc/pr samples analysis
```{r}
seu <- readRDS(here("data", "2022-02-20_seu_normal_mnn.rda"))
#seu <- pipe.fast.mnn(seu, "batch")
```

```{r}
DimPlot(seu, group.by = 'cell.type', label = T, repel = T, pt.size = .25) + NoLegend() + ggtitle("Normal Samples - MNN, Batch")
```

```{r}
#ggsave(here("results", "seurat", paste0(Sys.Date(), "_normal_samples_seu_mnn.png")))
```

## Split fetal and maternal

Split into maternal versus fetal subsets.
```{r}
split <- SplitObject(seu, split.by = 'fetal')
rm(seu) # Clean memory
fetal <- split$Fetal
maternal <- split$Maternal
rm(split) # Clean memory
```

```{r}
DimPlot(fetal) + ggtitle("Fetal")
DimPlot(maternal) + ggtitle("Maternal")
```

```{r}
rm(maternal)
```

## MNN and rerunning UMAP on fetal subset and initial clustering


```{r}
fetal <- pipe.fast.mnn(fetal, "batch")
```

```{r}
DimPlot(fetal, group.by = "seurat_clusters", label = T, pt.size =.25) + NoLegend()
DimPlot(fetal, group.by = 'cell.type', label = T, repel = T, pt.size = .25) + NoLegend()
```

## Subset to normal Tsang samples

```{r}
tsang <- subset(fetal, subset = batch == "tsang")
```

```{r}
#tsang <- pipe.fast.mnn(tsang, "biorep")
tsang <- Seurat::RunUMAP(object = tsang, reduction = "mnn", dims = 1:50)
```

```{r}
DimPlot(tsang, group.by = "seurat_clusters", label = T, pt.size =.25) + NoLegend()
```

### Doublet simulation
https://www.nature.com/articles/ncomms14049#MOESM828 ; original 10x paper
Should I up the doublet rate for 10x v1 chemistry? The most important hyperparameter for doublet calling? Doubled it
```{r}
set.seed(1)
tsang.sce <- as.SingleCellExperiment(tsang)
tsang.sce$cluster <- fastcluster(tsang.sce)
tsang.sce$cluster <- as.factor(tsang.sce$cluster)
#plotUMAP(tsang.sce, colour_by="cluster", label = T)
tsang.dbl <- scDblFinder(tsang.sce, samples = tsang$orig.ident, clusters = "cluster", dbr.sd = 0.15*2)
```

```{r}
plotUMAP(tsang.dbl, colour_by="scDblFinder.score")
```

```{r}
table(tsang.dbl$scDblFinder.class)
```

```{r}
table(tsang.dbl$scDblFinder.class, tsang.dbl$cluster)
```


```{r}
plotUMAP(tsang.dbl, colour_by="scDblFinder.class")
```

```{r}
tsang$scDblFinder.score <- tsang.dbl$scDblFinder.score
tsang$scDblFinder.class <- tsang.dbl$scDblFinder.class
tsang$cluster <- tsang.dbl$cluster
```

```{r}
FeaturePlot(tsang, features = "scDblFinder.score")
DimPlot(tsang, group.by = "scDblFinder.class")
DimPlot(tsang, group.by = "cluster", label = T, repel = T)
dbl.score.plot <- FeaturePlot(tsang, features = "scDblFinder.score")
```

### Manually lassoing of nearby doublets

```{r, eval = F}
dbl.3 <- CellSelector(dbl.score.plot)
```


```{r, eval = F}
dbl.12.8 <- CellSelector(dbl.score.plot)
```

```{r, eval = F}
dbl.8.10 <- CellSelector(dbl.score.plot)
```

```{r, eval = F}
dbl.13 <- CellSelector(dbl.score.plot)
```

```{r, eval = F}
dbl.10.9 <- CellSelector(dbl.score.plot)
```

```{r}
#dbl.cell.selector <- c(dbl.10.9, dbl.12.8, dbl.13, dbl.3, dbl.8.10)
#saveRDS(dbl.cell.selector, file = here("data", paste0("tsang_dbl_cell_selector_ids_", Sys.Date(), ".rda")))
dbl.cell.selector <- readRDS(here("data", "tsang_dbl_cell_selector_ids_2022-02-28.rda"))
```

319 additional doublets based on manual lassoing
```{r}
scDblFinder.cell.ids <- rownames(tsang@meta.data)[tsang$scDblFinder.class == "doublet"]
length(scDblFinder.cell.ids)
dbl.cell.selector %in% scDblFinder.cell.ids %>% summary
```

```{r}
dbl.ids.to.drop <- union(scDblFinder.cell.ids, dbl.cell.selector)
length(dbl.ids.to.drop)
```

```{r}
tsang.dbl.filtered <- subset(x = tsang, cells = dbl.ids.to.drop, invert = T)
```

## Algorithmic cell type assignment of Tsang samples using KC and PR carefully annotated samples with SingleR
Subset to already annotated KC/PR fetals cells and convert to SingleCellExperiment and log-normalize
```{r}
fetal.kc.pr.seu <- subset(fetal, subset = batch %in% c("kc", "pr"))
fetal.kc.pr.sce <- as.SingleCellExperiment(fetal.kc.pr.seu)
fetal.kc.pr.sce <- logNormCounts(fetal.kc.pr.sce)
```

Convert SeuratObject to SingleCellExperiment and log-normalize
```{r}
tsang.sce <- logNormCounts(as.SingleCellExperiment(tsang.dbl.filtered))
```

Run SingleR with "wilcox" de.method because single-cell test and reference datasets
```{r}
pred <- SingleR(test = tsang.sce,
                ref = fetal.kc.pr.sce,
                labels = fetal.kc.pr.sce$cell.type,
                de.method = "wilcox")
```

```{r}
table(pred$labels)
```

```{r}
plotScoreHeatmap(pred)
```

```{r}
plotDeltaDistribution(pred, ncol = 3)
```

```{r}
summary(is.na(pred$pruned.labels))
```

```{r}
tsang.dbl.filtered$labels <- pred$labels
tsang.dbl.filtered$pruned.labels <- pred$pruned.labels
tsang.dbl.filtered$cell.type <- pred$pruned.labels
tsang.dbl.filtered$pruned <- is.na(tsang.dbl.filtered$pruned.labels)
```

```{r}
DimPlot(tsang.dbl.filtered, group.by = "cell.type", label = T, repel = T) + NoLegend()
DimPlot(tsang.dbl.filtered, group.by = "pruned")
FeaturePlot(tsang.dbl.filtered, features = "scDblFinder.score")
```

### DROP 268 outliers from SingleR notation
TSANG.QC

Get IDs of dropped cells
```{r}
# Get SingleR pruned IDs
pruned.ids <- rownames(tsang@meta.data)[tsang@meta.data$pruned == T]
# Putative doublets are assigned doublets, lassoed doublets, and SingleR pruned IDs
tsang.putative.doublets <- union(pruned.ids, dbl.ids.to.drop)
#saveRDS(tsang.putative.doublets, here("data", "analytic", "post_cell_type", paste0(Sys.Date(), ".tsang.putative.doublets")))
# Get analytic initial Tsang analysis
tsang.initial.seu <- subset(seu.initial.analysis, subset = batch == "tsang")
# What are the low feature (< 200 genes) cells ids from the initial analysis
tsang.low.feature.ids <- rownames(tsang.initial.seu@meta.data)[tsang.initial.seu$nFeature_RNA < 200]
# None of the low feature cells fall into the putative doublets; so just drop the low feature cells from the initial Tsang analysis
tsang.putative.doublets %in% tsang.low.feature.ids %>% summary

# Report the following putative doublets by sample
tsang$putative.doublets <- rownames(tsang@meta.data) %in% tsang.putative.doublets
table(tsang$putative.doublets, tsang$orig.ident)
```

```{r}
table(tsang.dbl.filtered$pruned)
tsang.pruned <- subset(tsang, subset = pruned == T, invert = T)
```

```{r}
merged <- merge(fetal.kc.pr.seu, tsang.pruned)
```

```{r}
merged <- pipe.fast.mnn(merged, "batch")
DimPlot(merged, group.by = "cell.type", label = T, repel = T) + NoLegend() + ggtitle("Fetal pruned and merged")
```

```{r}
#ggsave(here("results", paste0(Sys.Date(), "_singler_seu_pruned_merged_umap.png")))
```

```{r}
#saveRDS(merged, here("data", paste0("fetal_pruned_and_merged_", Sys.Date(), ".rda")))
```

# Updated resubmit
TODO
## Tsang resubmit - how many doublets above were removed with more stringent QC
PR and Tsang have more strict thresholds, compare drop nFeature_RNA < 200 with expected removal based on updated QC pipeline, doublet and SingleR being excluded as putative doublets
Doublets for KC/PR and maybe keep manual lassoing; where do number of doublets removed come from in kc/pr samples - done
SingleR on unannotated KC cells - done, no pruned

```{r}
#Load the initial Tsang analytic dataset
seu.ini <- readRDS(here("data", "analytic", "2022-04-18_all_merged_pruned_batch_mnn.rda"))

# The minimum nFeature_RNA for all of batch KC is 624 genes
seu.initial.analysis@meta.data %>% filter(batch == "kc") %>% pull(nCount_RNA) %>% min()
```

Load the WIP resubmit analysis
```{r}
seu <- readRDS(here("data", "analytic", "post_anno", "2022-04-25_resubmit_annotated_seu_normal_mnn.rda"))
seu <- subset(seu, subset = fetal == "Fetal")
# 38,689 fetal cells
#seu

# Split by v1 vs. v2 10x chemistry
seu.kc.pr <- subset(seu, subset = batch %in% c("kc", "pr"))
seu.kc.pr <- pipe.fast.mnn(seu.kc.pr, batch = "batch")
```

### Doublet simulation
https://www.nature.com/articles/ncomms14049#MOESM828 ; original 10x paper
Double the doublet rate for 10x v1 chemistry, the most important hyperparameter for doublet calling
```{r}
set.seed(1)
sce <- as.SingleCellExperiment(seu.kc.pr)
sce$cluster <- fastcluster(sce)
sce$cluster <- as.factor(sce$cluster)
sce.dbl <- scDblFinder(sce, samples = sce$orig.ident, clusters = "cluster", dbr.sd = 0.15)
```

```{r}
plotUMAP(sce.dbl, colour_by="scDblFinder.score")
```

```{r}
table(sce.dbl$scDblFinder.class)
```

```{r}
table(sce.dbl$scDblFinder.class, sce.dbl$cluster)
```

```{r}
table(sce.dbl$scDblFinder.class, sce.dbl$orig.ident)
```

```{r}
plotUMAP(sce.dbl, colour_by="scDblFinder.class")
```

```{r}
seu.kc.pr$scDblFinder.score <- sce.dbl$scDblFinder.score
seu.kc.pr$scDblFinder.class <- sce.dbl$scDblFinder.class
seu.kc.pr$cluster <- sce.dbl$cluster
```

```{r}
FeaturePlot(seu.kc.pr, features = "scDblFinder.score")
DimPlot(seu.kc.pr, group.by = "scDblFinder.class")
DimPlot(seu.kc.pr, group.by = "cluster", label = T, repel = T)
dbl.score.plot <- FeaturePlot(seu.kc.pr, features = "scDblFinder.score")
```

scDblFinder performs well on KC/PR, captures most of the manually annotated doublets, add additional 24 manually annotated doublets from nRBC/pDC rare populations initial kc/pr analysis
```{r}
kc.pr.sc.dbl.finder.ids <- seu.kc.pr@meta.data %>% filter(scDblFinder.class == "doublet") %>% rownames
# Pull manually annotated doublets from initial publication submission
manual.intial.analysis.doublets <- c(readRDS(here("data", "ct_mesenchymal_doublet_cluster_ids_2020-12-22.rda")), readRDS(here("data", "cell_selector_dropped_pdc_rbc_clusters_2020-12-22.rda")))

# Identify the additional manually added doublets to be added
manual.intial.analysis.doublets %in% kc.pr.sc.dbl.finder.ids %>% summary
additional.kc.pr.doublets <- manual.intial.analysis.doublets[!(manual.intial.analysis.doublets %in% kc.pr.sc.dbl.finder.ids)]
kc.pr.doublet.ids <- c(kc.pr.sc.dbl.finder.ids, additional.kc.pr.doublets)
#saveRDS(kc.pr.doublet.ids, here("data", "analytic", "post_cell_type", paste0(Sys.Date(), "_kc_pr_putative_doublet_ids.rda")))

# Add doublet label metadata
seu.kc.pr$doublet <- rownames(seu.kc.pr@meta.data) %in% kc.pr.doublet.ids
table(seu.kc.pr$orig.ident, seu.kc.pr$doublet)

# Identify if additional cell add due to slightly relaxed QC requirement
seu.kc.pr$annotated <- rownames(seu.kc.pr@meta.data) %in% rownames(seu.ini@meta.data)

table(seu.kc.pr$doublet, seu.kc.pr$annotated)

# Get keep IDs (non-doublets); does not include SingleR considerations; will need to update pruned as also being doublets
seu.kc.pr.doublet.ids <- seu.kc.pr@meta.data %>% filter(doublet == F) %>% rownames
```

```{r}
DimPlot(seu.kc.pr, group.by = "scDblFinder.class", cells.highlight = manual.intial.analysis.doublets)
DimPlot(seu.kc.pr, group.by = "scDblFinder.class")
FeaturePlot(seu.kc.pr, features = ("COL1A1"))
```

## Algorithmic cell type assignment of updated QC using initial Tsang analysis carefully annotated samples with SingleR

```{r}
# Subset to unannotated, non-doublets
seu.kc.pr.dbl.filtered <- subset(seu.kc.pr, subset = (doublet == F & annotated == F))
# Only 37 cells are not annotated and not doublets with the updated QC thresholds of lib.size = 500, genes.detected = 200
seu.kc.pr.dbl.filtered
```

```{r}
sce.ini <- as.SingleCellExperiment(subset(seu.ini, subset = fetal == "Fetal" & batch %in% c("kc", "pr")))
sce.ini <- logNormCounts(sce.ini)
```

Convert SeuratObject to SingleCellExperiment and log-normalize, subset to only non-annotated cells
```{r}
seu <- subset(seu.kc.pr.dbl.filtered)
sce <- logNormCounts(as.SingleCellExperiment(seu))
```

Run SingleR with "wilcox" de.method because single-cell test and reference datasets
```{r}
pred <- SingleR(test = sce,
                ref = sce.ini,
                labels = sce.ini$cell.type,
                de.method = "wilcox")
```

```{r}
table(pred$labels)
```

```{r}
plotScoreHeatmap(pred)
```

```{r}
plotDeltaDistribution(pred, ncol = 3)
```

```{r}
summary(is.na(pred$pruned.labels))
```

```{r}
seu$labels <- pred$labels
seu$pruned.labels <- pred$pruned.labels
seu$cell.type <- pred$pruned.labels
seu$pruned <- is.na(seu$pruned.labels)
```

```{r}
DimPlot(seu, group.by = "cell.type", label = T, repel = T) + NoLegend()
DimPlot(seu, group.by = "pruned")
table(seu$cell.type, seu$orig.ident)
```

### DROP additional 0 outliers from SingleR notation

```{r}
table(seu$pruned)
seu.pruned <- subset(seu, subset = pruned == T, invert = T)
```

Newly annotated fetal KC cells ready to add to analytic dataset
```{r}
seu.kc.add <- seu.pruned
#saveRDS(seu.kc.add, here("data", "analytic", "post_cell_type", paste0(Sys.Date(), "_seu_kc_add.Rda")))
```

### KC/PR Doublet drop summary
KC PR QC
```{r}
table(seu.kc.pr$doublet, seu.kc.pr$orig.ident)
```

```{r}
merged <- merge(seu.ini, seu.kc.add)
```

```{r}
merged <- pipe.fast.mnn(merged, "batch")
DimPlot(merged, group.by = "cell.type", label = F, repel = T, pt.size = 0.25) + ggtitle("Fetal pruned and merged")
FeaturePlot(merged, features = "scDblFinder.score")
FeaturePlot(seu.ini, features = "scDblFinder.score")
DimPlot(seu.ini, group.by = "cell.type")
```

