---
title: "DESeq2 Analysis of Placenta Sorted Cell Types"
subtitle: "Kyle Campbell - March 14th, 2019"
output: html_document
---

```{r setup, include=FALSE}
# Update R all installed R Packages
# update.packages(ask = FALSE)
library(ashr)
library(ggplot2)
library(ggrepel)
library(gprofiler2)
library(here)
library(knitr)
library(pheatmap)
library(RColorBrewer)
library(tidyverse)

library(BiocManager)
library(biomaRt)
library(DESeq2)
library(GO.db)
library(mygene)
library(org.Hs.eg.db)
#library(topGO)

knitr::opts_chunk$set(echo=TRUE)

# May need to update root.dir Drive Letter
#local_data_path <- "C:/Users/k_cam/Google Drive/Placenta_Cell_Types_Kyle/code/R code/local/placenta_sort/dex/data/analysis_datasets/"
```
## Entering Data, Preproccessing, and Model Fitting
```{r DESeq2_object}
# Older file management system
#counts <- readRDS(file =   "./data/analysis_datasets/edgeR_collapsed_counts.Rda")
#meta <- readRDS(file =     "./data/analysis_datasets/edgeR_collapsed_metadata.Rda")
#gene_annotation <- readRDS("./data/analysis_datasets/edgeR_gene_annotation_with_symbols.Rda")

#counts <- readRDS(file =  here("data", "bulk", "edgeR_collapsed_counts.Rda"))
#meta <- readRDS(file = here("data", "bulk", "edgeR_collapsed_metadata.Rda"))
#gene_annotation <- readRDS(here("data", "bulk", "edgeR_gene_annotation_with_symbols.Rda"))

# Must make meta rownames identical to counts colnames
#rownames(meta) <- substr(rownames(meta), 8, 14)

# Creating the DESeq2 experiment object from count data and meta info
#dds <- DESeqDataSetFromMatrix(countData = counts,
#                              colData = meta,
#                              design = ~ 0 + sample.ID + cell.type)

# Adding comprehensive gene annotation to dds
#mcols(dds) <- DataFrame(mcols(dds), gene_annotation)

# NOT using at this point
# Minimal pre-processing to reduce computational load - does NOT represent meaningful filtering
#keep <- rowSums(counts(dds)) >= 10
#dds <- dds[keep,]

# Fitting the model
#dds <- DESeq(dds)
```

Droppping .xx from ENSG IDs for symbol annotation: https://www.biostars.org/p/293965/
Explanation of ENSG vs. ENST: https://www.biostars.org/p/199073/

```{r AddGeneAnnotation}
# Chop off the .xx from the ENSEMBL ID
#row.names(dds) <- substr(x = row.names(dds), start = 1, stop = 15)

#row.names(dds) <- mapIds(org.Hs.eg.db,
#                     keys=row.names(dds),
#                     column="SYMBOL",
#                     keytype="ENSEMBL",
#                     multiVals="first")
```

BiomaRt returns 58,069 of 58,381 gene features from DESeq2 results. Of the 58,069 results, 56,521 return unique external gene names and 58,024 return unique ENSEMBL gene IDs.

```{r biomaRt_DESeq2_results}
# Setting up biomaRt
# Assigning Mart of interest, human ensembl here
#mart = useMart("ensembl",dataset="hsapiens_gene_ensembl")

# Querying biomaRt using gene names to get back ENSEMBL IDs
#DESeq2GeneIDs <- getBM(attributes = c('ensembl_gene_id', 'external_gene_name'),
#             filters = c('ensembl_gene_id'),
#             values = rownames(dds),
#             mart = mart)
#genes.tbl <- as_tibble(DESeq2GeneIDs)

#saveRDS(genes.tbl, here("data", "bulk", "deseq2_gene_symbols.rda"))
# Load gene ids
genes.tbl <- readRDS(here("data", "bulk", "deseq2_gene_symbols.rda"))
```

Save/load 
```{r}
#saveRDS(DESeq2GeneIDs_tbl, "biomaRt_DESeq2_ensembl_ids.rda")
#DESeq2GeneIDs_tbl <- as_tibble(readRDS("biomaRt_DESeq2_ensembl_ids.rda"))
```

```{r SaveLoad}
# Saved the fitted model object to avoid having to reestimate parameters
#saveRDS(dds, file = "./data/analysis_datasets/DESeq2_dds_noint_celltype_sampleid_nofiltering.rda")
#dds <- readRDS(file = "./data/analysis_datasets/DESeq2_dds_noint_celltype_sampleid_nofiltering.rda")

# Load the finished DESeq2 object - note this object has trimmed ensembl IDs
#dds <- readRDS("C:/Users/Kyle/Google Drive/Placenta_Cell_Types_Kyle/code/R code/local/placenta_sort/dex/data/analysis_datasets/DESeq2_dds_noint_celltype_sampleid_nofiltering.rda")

#dds <- readRDS(paste0(local_data_path, "/DESeq2_dds_noint_celltype_sampleid_nofiltering.rda"))

dds <- readRDS(here("data", "bulk", "DESeq2_dds_noint_celltype_sampleid_nofiltering.rda"))
```

Ideas for how to deal with multiple Ensembl IDs mapping to a single gene symbol
https://www.researchgate.net/post/How-to-deal-with-multiple-ensemble-IDs-mapping-to-one-gene-symbol-in-a-RNA-Seq-dataset

## Comparing cell type-specific gene expression to the average expression level across other cell types

Our first contrast of interest is how cell type-specific expression differs vs. the average expression across all other cell types. [This thread](https://support.bioconductor.org/p/118090/) recommends fitting a no-intercept model to answer this question. [Another thread](https://support.bioconductor.org/p/86347/) contains a brief discussion of how to code it. These contrasts will exclude the composite tissue group. I wonder if excluding the composite samples from normalization and filtering is appropriate for this particular set of comparisions. I think that also depends on whether we use the composite samples in any analyses.

Modernized code (3/2021) and outstanding questions:
1. Include composite samples in model for gene variance moderation or drop?
  a. "If I have multiple groups, should I run all together or split into pairs of groups?" DESeq2 vignette recommends to do so unless within-group variability varies differently than other groups, the high variance can disrupt the genewise dispersion estimates in the other groups.
  b. Emprically test running with and without
2. Include endothelial in the cell type comparison group?
  a. Empirically test
    i. Stable model either way (fewer hits than CT)
3. Handling Ensembl ID/HGNC symbol multimapping
  a. previously handled at annotating results stage (when DESeq2 vignette does it)
    i. Currently, independent filtering handles most multimapping reads due to low counts or low variance, requiring no filtering
    ii. Less than 100 multimapping genes per contrast remain even after independent filtering and are dropped

Note: in DESeq2 alpha is the FDR-adjusted significance cutoff

Bimodal histograms are a little weird down below. Several threads (https://support.bioconductor.org/p/102313/, https://www.biostars.org/p/331711/, https://www.biostars.org/p/312792/), including Michael Love suggest pre-filtering as low counts may be causing the peaks on the p-value histogram (recall that a well-calibrated statistical test has a uniform p-value distribution with a spike at low p-values if there's a significant difference). 

Pre-filtering: normalized counts greater than 10 per gene and detected in at least three samples, as recommended in the DESeq2 paper to address bimodal p-value distribution.

```{r}
binary <- as.matrix((counts(dds) > 0) + 0)
keep.three <- rowSums(binary) >= 3
dds <- dds[keep.three]
```

```{r}
keep.min <- rowSums(counts(dds, normalized = TRUE)) >= 10
dds <- dds[keep.min, ]
```

``` {r}
run_deseq2 <- function(target, dds) {
  
  # Print current target
  print(target)
  
  # Set alpha
  alpha <- .05
  
  # List coefficients
    # cell.types <-
    # c(
    #   "cell.typecytotrophoblast",
    #   "cell.typeendothelial",
    #   "cell.typeextravilloustrophoblast",
    #   "cell.typefibroblast",
    #   "cell.typehofbauer",
    #   "cell.typeleukocyte",
    #   "cell.typesyncytiotrophoblast"
    # )
  
  # cell.types <-
  #   c(
  #     "cell.typecytotrophoblast",
  #     "cell.typeextravilloustrophoblast",
  #     "cell.typefibroblast",
  #     "cell.typehofbauer",
  #     "cell.typeleukocyte",
  #     "cell.typesyncytiotrophoblast"
  #   )
  
    # cell.types <-
    # c(
    #   "cell.typecytotrophoblast",
    #   "cell.typeextravilloustrophoblast",
    #   "cell.typefibroblast",
    #   "cell.typehofbauer",
    #   "cell.typesyncytiotrophoblast"
    # )
  
      cell.types <-
    c(
      "cell.typecytotrophoblast",
      "cell.typefibroblast",
      "cell.typehofbauer",
      "cell.typesyncytiotrophoblast"
    )
  
  # Select contrast coefficients based on cell type input
  coef <- cell.types[grepl(target, cell.types)]
  inv.coef <- cell.types[!grepl(target, cell.types)]
  
  # For the DESeq2 results() call in this chunk:
  # The first vector is our cell type of interest
  # Create the contrast of interest
  # The following vector contains comparison group
  # listValues Specifies comparision against average expression across other groups
  # Pull contrast of interest
  results <- results(
    dds,
    contrast = list(coef,
                    inv.coef),
    listValues = c(1,-1 / 3),
    alpha = alpha
  )
  
  # Apply logFoldChange shrinkage
  resLFC <- lfcShrink(dds, contrast = list(coef,
                  inv.coef), type="ashr")
  
  # Summarize DESeq2 filtering
  summary <- summary(results)
  print(
    paste0(
      summary(is.na(results$padj))[3],
      " features dropped due to independent filtering, all zero counts, or extreme outlier status in DESeq2"
    )
  )
  
  # Coerce DESeq2 Results to df
  res.df <- as.data.frame(results)
  
  # Move gene ensembl IDs from row to table column
  res <- rownames_to_column(res.df, "ensembl_gene_id")
  
  # Remove dropped genes from results
  res <-
    res %>%
    filter(!is.na(padj))
  
  # Add gene annotation and drop duplicate gene names
  res <- left_join(res, genes.tbl) %>%
    dplyr::select(external_gene_name, everything()) %>%
    relocate(ensembl_gene_id, .after = last_col())
  
  # Print the small number of duplicated genes after DESeq2 auto filtering. Most are SNORNAs and many are vague Y_RNA, rRNA, metazoa RNA, etc.
  duplicated.genes <-
    res %>%
    group_by(external_gene_name) %>%
    tally() %>%
    filter(n > 1)
  print(paste0(dim(duplicated.genes)[1], " duplicated gene symbols dropped"))
  
  # Drop duplicated genes after DESeq2 auto filtering
  res <-
    res %>%
    group_by(external_gene_name) %>%
    filter(!(n() > 1)) %>%
    as_tibble()
  
  # Rename gene symbol column from "external_gene_name" to "gene_symbol"
  res <-
    res %>%
    mutate(gene_symbol = external_gene_name) %>%
    dplyr::select(-external_gene_name)
  
  res <-
    res %>%
    dplyr::select(gene_symbol,
                  baseMean,
                  log2FoldChange,
                  lfcSE,
                  stat,
                  pvalue,
                  padj,
                  ensembl_gene_id)
  
  # Get and arrange upreg
  upreg <-
    res %>%
    filter(padj < alpha) %>%
    filter(log2FoldChange > 0) %>%
    arrange(desc(log2FoldChange))
  
  # Get and arragne downreg
  downreg <-
    res %>%
    filter(padj < alpha) %>%
    filter(log2FoldChange < 0) %>%
    arrange(desc(abs(log2FoldChange)))
  
  # Number of top downregulated and upregulated genes to label on Volcano plot
  num.to.plot <- 10
  
  # Option to manually label a vector of genes
  genes.of.interest <- NULL
  
  # Create label dummy to decide if gene labelled on Volcano plot
  top.upreg <- upreg$gene_symbol[1:num.to.plot]
  top.downreg <- downreg$gene_symbol[1:num.to.plot]
  genes.to.plot <- c(top.upreg, top.downreg, genes.of.interest)
  res$label[(res$gene_symbol %in% genes.to.plot)] <- T
  
  # Add significance indicator variable for plotting
  res <-
    res %>%
    mutate(sig = ifelse(padj < alpha,
                        T,
                        F))
  # Volcano plot
  plot <- ggplot(res) +
    geom_point(aes(
      x = log2FoldChange,
      y = -log10(padj),
      colour = sig
    )) +
    geom_text_repel(aes(
      x = log2FoldChange,
      y = -log10(padj),
      label = ifelse(label, gene_symbol, "")
    ),
    max.overlaps = Inf) +
    geom_hline(yintercept = -log10(alpha), linetype = "dotted") +
    ggtitle("Contrast of Interest") +
    xlab("log2 fold change") +
    ylab("-log10 adjusted p-value") +
    theme(
      legend.position = "none",
      plot.title = element_text(size = rel(1.5), hjust = 0.5),
      axis.title = element_text(size = rel(1.25))
    ) +
    labs(caption = paste0(
      dim(res)[1],
      " total genes (",
      summary(is.na(results$padj))[3],
      " dropped from contrast)"
    ))
  
  # Create a label dataframe for geom_label annotation
  label.df <- data.frame(
    label = paste0(dim(upreg)[1], " upregulated genes\n", paste0(dim(downreg)[1], " downregulated genes")),
    x = Inf,
    y = Inf,
    hjustvar = 1,
    vjustvar = 1
  )
  
  # Add geom_label annotation
  plot <-
    plot + geom_label(data = label.df,
                      aes(
                        x = x,
                        y = y,
                        label = label,
                        hjust = hjustvar,
                        vjust = vjustvar
                      ))
  plot
  
  returnList <- list("res" = res,
                     "volcano" = plot,
                     "duplicated.genes" = duplicated.genes,
                     "results" = results,
                     "shrunk" = resLFC)
  return(returnList)
}

# cell.types <-
#     list(
#       "cytotrophoblast",
#       "endothelial",
#       "extravilloustrophoblast",
#       "fibroblast",
#       "hofbauer",
#       "leukocyte",
#       "syncytiotrophoblast"
#     )

# cell.types <-
#     list(
#       "cytotrophoblast",
#       "extravilloustrophoblast",
#       "fibroblast",
#       "hofbauer",
#       "leukocyte",
#       "syncytiotrophoblast"
#     )

    # cell.types <-
    # c(
    #   "cell.typecytotrophoblast",
    #   "cell.typeextravilloustrophoblast",
    #   "cell.typefibroblast",
    #   "cell.typehofbauer",
    #   "cell.typesyncytiotrophoblast"
    # )
    
          cell.types <-
    c(
      "cell.typecytotrophoblast",
      "cell.typefibroblast",
      "cell.typehofbauer",
      "cell.typesyncytiotrophoblast"
    )

mapped <- lapply(cell.types, run_deseq2, dds = dds)
names(mapped) <- cell.types
```

MA Plots, p-adj < .1 is highlighted by default
```{r}
cell.types
mas <- map(mapped, ~ DESeq2::plotMA((.x[["results"]])))
```

ashr-shrunk MA Plots. Not yet sure what the purpose of shrinkage is beyond "visualization and gene ranking" as explained in the DESeq2 vignette
```{r}
cell.types
mas <- map(mapped, ~ DESeq2::plotMA((.x[["shrunk"]])))
```

Making presentable Volcanos plots
```{r}
# Input ggplot to get x and y limits for graph
get_coord_limits <- function(ggplot) {
  
  # Get build options
  build <- ggplot_build(ggplot)
  # Get x-limits
  x <- build$layout$panel_scales_x[[1]]$range$range
  # Get y-limits
  y <- build$layout$panel_scales_y[[1]]$range$range
  
  # Return x and y limits as list
  return(list(
    "x" = x,
    "y" = y
  ))
}

# Pull limits from each graph
limits <- map(mapped, ~ get_coord_limits(.x[["volcano"]]))

# Pull x-limits and convert to list of vectors
x_limits <- map(limits, ~ unlist(.x["x"]))
# Convert x-limits to vector
x_limits <- unlist(x_limits)
# Get x-min
x_min <- min(x_limits)
# Get x-max
x_max <- max(x_limits)

# Pull y-limits and convert to list of vectors
y_limits <- map(limits, ~ unlist(.x["y"]))
# Convert y-limits to vector
y_limits <- unlist(y_limits)
# Get y-min
y_min <- min(y_limits)
# Get y-max
# y-max for ST is 1e-60, going to use 2nd largest value instead for graph readibility
y_sorted <- sort(y_limits)
y_max <- y_sorted[length(y_sorted)-3] %>% as.numeric
#y_max <- max(y_limits)

names(mapped)

contrast.labels <- c("Cytotrophoblast" , "Fibroblast", "Hofbauer Cell", "Syncytiotrophoblast")

#contrast.labels <- c("Cytotrophoblast" , "Extravillous Trophoblast", "Fibroblast", "Hofbauer Cell", "Syncytiotrophoblast")

#contrast.labels <- c("Cytotrophoblast" , "Extravillous Trophoblast", "Fibroblast", "Hofbauer Cell", "Leukocyte", "Syncytiotrophoblast")

#contrast.labels <- c("Cytotrophoblast" , "Endothelial", "Extravillous Trophoblast", "Fibroblast", "Hofbauer Cell", "Leukocyte", "Syncytiotrophoblast")

# Get presentable volcanos
# Add contrast title and set x and y limits without removing data points
volcanos <- map2(mapped, contrast.labels,
                 ~ .x[["volcano"]] + ggtitle(.y) + coord_cartesian(xlim = c(x_min, x_max), ylim = c(y_min, y_max)))

# Custom function to save ggplots
save_volcano_plot <- function(plot, filename) {
  print(plot)
  ggsave(
    here("results", "bulk", "volcanos", paste0(filename, "_volcano_", Sys.Date(), ".png"))
  )
}
# Save volcanos
#map2(volcanos, names(mapped), ~ save_volcano_plot(.x, .y))
```

```{r}
map(volcanos, print)
```

```{r}
# Pull dex results
results <- map(mapped, ~ .x[["res"]])
```

PrerankGSEA does not resolve ties. Code to make GSEA input and diagnostic plots to ensure that there are not excessive ties. GSEA results seem hard to interpret so far. GSEA should always include all genes tested in differential expression. Perhaps subsetting to the most enriched pathways will be easier to interpret. In RNA-seq, GSEA may benefit from very lowly expressed genes that may be artifactual and expressed in any samples.
```{r}
make_gsea_input <- function(res, title) {

  return <- res %>%
    arrange(desc(stat)) %>%
    dplyr::select(gene_symbol, stat)
  
  # Pull the stat associated with the largest p-value less than .05 for boundaries
  stat.lim <- res %>%
    filter(padj < .05) %>%
    arrange(desc(padj)) %>%
    slice_head(n = 1) %>%
    pull(stat)
  
  # Pull the associated padj
  padj.lim <- res %>%
    filter(padj < .05) %>%
    arrange(desc(padj)) %>%
    slice_head(n = 1) %>%
    pull(padj)
  
  # Get number of stat duplicates
  dup.stats.counts <- res %>%
  group_by(stat) %>%
  filter(n() > 1) %>%
  tally()
  
  # Graph stat duplicates
  print(ggplot(dup.stats.counts, aes(x = stat)) + geom_histogram() + ggtitle(paste0("Duplicate Test Statistics - ", title)) +
    geom_vline(xintercept = abs(stat.lim)) + geom_vline(xintercept = -abs(stat.lim)) +
    geom_label(
      data = NULL,
      x = abs(stat.lim),
      y = 50,
      label = paste0("padj = ", round(padj.lim, 2))
    ))
  
  # Save plots
  # Unsure why this code is producing an error about not being able to open .png device
  # ggsave(
  #   here("results", "GSEA", " sorted", "dup_stats_plots", paste0(title, "_dup_gsea_input_stats_", Sys.Date(), ".png"))
  # )
  
  # Write  .rnk file for GSEA input
  write_tsv(return, here("results", "GSEA", "sorted", paste0(title, "_no_lk_en_evt_prefilter_gsea_input_stats_", Sys.Date(), ".rnk")), col_names = FALSE)
  
  return(return)
}

inputs <- map2(results, contrast.labels, ~ make_gsea_input(res = .x, title = .y))
```

```{r}
map2(results, contrast.labels, ~ ggplot(data = .x, mapping = aes(x = pvalue)) + geom_histogram() + ggtitle(.y))
```

```{r}
gost_deseq2 <- function(res, alpha) {
  
  # Filter results and rank by test statistic
  res <-
    res %>%
    filter(padj < alpha) %>%
    arrange(desc(stat)) %>%
    filter(log2FoldChange > 0)
  gost.input <- res$gene_symbol
  
  # gost from ggprofiler2 with ordered query and manually curated annotation background
  gostres <- gost(query = gost.input,
               organism = "hsapiens", ordered_query = TRUE,
               multi_query = FALSE, significant = FALSE, exclude_iea = TRUE,
               measure_underrepresentation = FALSE, evcodes = FALSE,
               user_threshold = alpha, correction_method = "g_SCS",
               domain_scope = "annotated", custom_bg = NULL,
               numeric_ns = "", sources = NULL, as_short_link = FALSE)
  # graph
  gostplot <- gostplot(gostres, capped = FALSE, interactive = FALSE)
  print(gostplot)
  
  results <- list(
    "gostres" = gostres,
    "gostplot" = gostplot
  )
  
  return(results)
}

gost.mapped <- map(results, ~ gost_deseq2(.x, alpha = .05))
```

``` {r cyto_vs_all, eval = F}
# Get Cytotrophoblast results
 ct.res <- SummarizeDESeq2Results(cytovall_res, "Cytorophoblasts", .005)

# GO analysis
# Create the numeric vector FDR-adjusted p-values from DESeq2 DEX output, remove missing p-vals from DESEq2, and name list with genes for topGO input
ctGenes <- ct.res$padj
names(ctGenes) <- rownames(ct.res)
ctGenes <- na.omit(ctGenes)
str(ctGenes)

# Define a function for padj cutoff selection
topDiffGenes <- function(allScore) {
  return(allScore < 0.05)
}

# Constructing the topGOdata object for downstream analysis
ctGOdata <- new("topGOdata",
                ontology = "BP",
                allGenes = ctGenes,
                geneSelectionFun = topDiffGenes,
                nodeSize = 10,
                annot = annFUN.org,
                mapping = "org.Hs.eg.db",
                ID = "ensembl")

# Running the enrichment test using the weight01Score method because weight method accounts for GO topology and therefore accoutns for multiple testing. Score method to employ the nonparametric Kolmogorov-Smirnov test based on gene rank (also known as GSEA)
test.stat <- new(Class = "weight01Score", testStatistic = GOKSTest, name = "Weight01 KS tests")

# getSigGroups returns an object of class 'topGOresult', here named resultweight01KS to reflect test employed
resultweight01KS <- getSigGroups(ctGOdata, test.stat)

# Already performed in dataset
# # Chop off the .xx from the ENSEMBL ID
# row.names(res) <- substr(x = row.names(res), start = 1, stop = 15)
# 
# # Add gene symbols
# res$symbol <- mapIds(org.Hs.eg.db,
#                      keys=row.names(res),
#                      column="SYMBOL",
#                      keytype="ENSEMBL",
#                      multiVals="first")
plotMA(res)
summary(res)

res_df <- as.data.frame(res)
dim(res_df) #58381 genes
summary(res_df$pvalue) #8862 missing, with zero counts in all conditions
length(res_df$stat[res_df$stat==0]) #14208 that have test statistics of 0, independent filtered out?
58381 - 8862 - 14208 #35311 total genes used in this contrast

testd <- res_df %>%
  filter(padj < .05) %>%
  arrange(desc(baseMean), desc(log2FoldChange))

testd <- as_tibble(testd)
testd <- rename(testd, scGenes = symbol)

testleftjoin <- left_join(scGenes_df, testd, by = "scGenes")
testleftjoin %>%
  arrange(desc(baseMean), desc(log2FoldChange))
testleftjoin <- drop_na(testleftjoin)
#write.csv(x = testleftjoin, file = "CTvsAll_alpha005_scOverlap.csv")

```

``` {r e, eval = F}
counts_test <- counts
counts_test$hgnc_symbol <- convertIDs(x, "ENSEMBL", "SYMBOL", org.Hs.eg.db)
COL1A1 <- counts_test %>%
  filter(hgnc_symbol=="COL1A1")
COL1A1 <- COL1A1[,1:24]
col1a1 <- as.data.frame(COL1A1)
col1a1[2,] <- names(col1a1)

tes <- data.frame()#data.frame(reads=integer(col1a1[1,]), samples = factor(names(col1a1)))
tes$samples <- names(col1a1)
tes$reads <- col1a1[1,]

a <- ggplot(col1a1, aes(x = col1a1[2,], y = col1a1[1,]))
a + geom_col()

#write.csv(col1a1, )
```

## Data Quality Assessment by Clustering and Visualization

We'll begin by transforming the data to the log-fold-2 scale using the same shrinkage transformation employed in the DEX pipeline.

```{r transformation}
#rld <- rlog(dds, blind = TRUE)
#saveRDS(rld, file = paste0(local_data_path, "DESeq2_rld.rda"))
#saveRDS(rld, here("data", "bulk", paste0(Sys.Date(), "DESeq2_rld.rda")))
#rld <- readRDS(file = paste0(local_data_path, "DESeq2_rld.rda"))
rld <- readRDS(here("data", "bulk", "2021-03-26DESeq2_rld.rda"))
```
103911 - leukocyte is an extreme outlier on blind PCA analysis, will remove and rerun. PCA looks similar with 103911 - leukocyte removed
```{r}
#dds_103911 <- dds[,colnames(dds) != "103911"]
#dds_103911 <- DESeq(dds_103911)
#rld <-rlog(dds_103911, blind = TRUE)
```
### Hierarchical Clustering Heatmap
```{r hc_heatmap}
# Calculate sample distances
sampleDists <- dist(t(assay(rld)))

# Hierarchical heatmap clustering - DESeq2 vignette
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$cell.type, rld$sample.ID, rld$is.Paired, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```
### PCA Plot
```{r rld_pca}
# Customized PCA ggplot
pcaData <- plotPCA(rld, intgroup=c("cell.type", "sample.ID"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=cell.type, shape=sample.ID)) +
  geom_point(size=3.5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
#ggsave(sprintf("%sSorted_PCA.png", local_output_path), device = "png")
```